{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 4 Assignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Description\" data-toc-modified-id=\"Description-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Description</a></span><ul class=\"toc-item\"><li><span><a href=\"#readme.txt\" data-toc-modified-id=\"readme.txt-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>readme.txt</a></span></li><li><span><a href=\"#Story\" data-toc-modified-id=\"Story-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Story</a></span></li><li><span><a href=\"#File-Structure\" data-toc-modified-id=\"File-Structure-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>File Structure</a></span></li><li><span><a href=\"#copyExomes.sh\" data-toc-modified-id=\"copyExomes.sh-1.4\"><span class=\"toc-item-num\">1.4&nbsp;&nbsp;</span>copyExomes.sh</a></span></li><li><span><a href=\"#createCrisprReady.sh\" data-toc-modified-id=\"createCrisprReady.sh-1.5\"><span class=\"toc-item-num\">1.5&nbsp;&nbsp;</span>createCrisprReady.sh</a></span></li><li><span><a href=\"#identifyCrisprSite.sh\" data-toc-modified-id=\"identifyCrisprSite.sh-1.6\"><span class=\"toc-item-num\">1.6&nbsp;&nbsp;</span>identifyCrisprSite.sh</a></span></li><li><span><a href=\"#editGenome.sh\" data-toc-modified-id=\"editGenome.sh-1.7\"><span class=\"toc-item-num\">1.7&nbsp;&nbsp;</span>editGenome.sh</a></span></li><li><span><a href=\"#exomeReport.py\" data-toc-modified-id=\"exomeReport.py-1.8\"><span class=\"toc-item-num\">1.8&nbsp;&nbsp;</span>exomeReport.py</a></span></li></ul></li><li><span><a href=\"#Readme.txt\" data-toc-modified-id=\"Readme.txt-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Readme.txt</a></span></li><li><span><a href=\"#copyExomes.sh\" data-toc-modified-id=\"copyExomes.sh-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>copyExomes.sh</a></span><ul class=\"toc-item\"><li><span><a href=\"#Structure-Decision\" data-toc-modified-id=\"Structure-Decision-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>Structure Decision</a></span></li><li><span><a href=\"#Programing-Decision\" data-toc-modified-id=\"Programing-Decision-3.2\"><span class=\"toc-item-num\">3.2&nbsp;&nbsp;</span>Programing Decision</a></span></li><li><span><a href=\"#Read-in-clinical_data-with-pandas\" data-toc-modified-id=\"Read-in-clinical_data-with-pandas-3.3\"><span class=\"toc-item-num\">3.3&nbsp;&nbsp;</span>Read in clinical_data with pandas</a></span></li><li><span><a href=\"#Filter-dataframe-by-diameter-size\" data-toc-modified-id=\"Filter-dataframe-by-diameter-size-3.4\"><span class=\"toc-item-num\">3.4&nbsp;&nbsp;</span>Filter dataframe by diameter size</a></span></li><li><span><a href=\"#Make-exomes-directory-and-copy-over-filtered-files\" data-toc-modified-id=\"Make-exomes-directory-and-copy-over-filtered-files-3.5\"><span class=\"toc-item-num\">3.5&nbsp;&nbsp;</span>Make exomes directory and copy over filtered files</a></span></li><li><span><a href=\"#Create-motif_manager.py\" data-toc-modified-id=\"Create-motif_manager.py-3.6\"><span class=\"toc-item-num\">3.6&nbsp;&nbsp;</span>Create motif_manager.py</a></span></li><li><span><a href=\"#Task-1.4-Complete\" data-toc-modified-id=\"Task-1.4-Complete-3.7\"><span class=\"toc-item-num\">3.7&nbsp;&nbsp;</span>Task 1.4 Complete</a></span></li></ul></li><li><span><a href=\"#createCrisprReady.sh\" data-toc-modified-id=\"createCrisprReady.sh-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>createCrisprReady.sh</a></span><ul class=\"toc-item\"><li><span><a href=\"#Check-size-length-of-.fasta-files\" data-toc-modified-id=\"Check-size-length-of-.fasta-files-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>Check size length of .fasta files</a></span></li><li><span><a href=\"#Build-Animal-Dataframe\" data-toc-modified-id=\"Build-Animal-Dataframe-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Build Animal Dataframe</a></span></li><li><span><a href=\"#Build-dataframe-counter-for-each-animal\" data-toc-modified-id=\"Build-dataframe-counter-for-each-animal-4.3\"><span class=\"toc-item-num\">4.3&nbsp;&nbsp;</span>Build dataframe counter for each animal</a></span></li><li><span><a href=\"#Top-Ten-Motifs\" data-toc-modified-id=\"Top-Ten-Motifs-4.4\"><span class=\"toc-item-num\">4.4&nbsp;&nbsp;</span>Top Ten Motifs</a></span></li><li><span><a href=\"#Top-motifs-for-each-animal\" data-toc-modified-id=\"Top-motifs-for-each-animal-4.5\"><span class=\"toc-item-num\">4.5&nbsp;&nbsp;</span>Top motifs for each animal</a></span></li><li><span><a href=\"#Count-Report\" data-toc-modified-id=\"Count-Report-4.6\"><span class=\"toc-item-num\">4.6&nbsp;&nbsp;</span>Count Report</a></span></li><li><span><a href=\"#Create-motif_manager.py\" data-toc-modified-id=\"Create-motif_manager.py-4.7\"><span class=\"toc-item-num\">4.7&nbsp;&nbsp;</span>Create motif_manager.py</a></span></li></ul></li><li><span><a href=\"#identifyCrisprSite.sh\" data-toc-modified-id=\"identifyCrisprSite.sh-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>identifyCrisprSite.sh</a></span></li><li><span><a href=\"#editGenome.sh\" data-toc-modified-id=\"editGenome.sh-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>editGenome.sh</a></span></li><li><span><a href=\"#exomeReport.py\" data-toc-modified-id=\"exomeReport.py-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>exomeReport.py</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Description"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<strong>Due Tuesday Week 4 </strong><br>\n",
    "Beginning Python scripting <br>\n",
    "Key concepts covered: Apply Bash and Python practices to solving problems"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### readme.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<strong>Important! Your assignment must contain a readme.txt file that indicates the exact command used to execute your script on the AWS server. It should indicate what inputs are used, and indicate what outputs are produced.\n",
    "Use relative paths so that if I copied your week4 folder to any other directory, it will still execute properly.</strong>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Story"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dr. Banner leads a team of researchers that have each discovered a unique eukaryotic organism. They all seem to be related to each other and the team has tried to sequence their specimens. Homologous genes between each eukaryote were labeled with the same gene name. Each of organisms have some degree of gamma ray resistance and Dr. Banner constructed a list of motifs that may be associated with resistance. He wants to use only the most abundant motif and knockout the genes that contain these motifs using CRISPR.  He first needs to identify a list of suitable CRISPR gene candidates and submit a report. \n",
    "\n",
    "Inputs:\n",
    "\n",
    "/home/rbif/week4/clinical_data.txt <br>\n",
    "/home/rbif/week4/motif_list.txt <br>\n",
    "/home/rbif/week4/exomes/*.fasta <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File Structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "notes: Task completed. Tim, I am not sure if you want me to make a bash script that does this with something like:\n",
    "```\n",
    "mkdir week4\n",
    "rsync . /home/ruggm/BioInformatics/001_RBIF-100-1DL/week4\n",
    "```\n",
    "or just make it here. I am going to assume that I have done this successfully, but let me know if you want me to make a little build_me script."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<strong>Task 1.3 completed</strong>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### copyExomes.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in the clinical data file and identify the samples that have a diameter between 20 and 30 mm long (inclusive) and have had their genomes sequenced. Copy the identified exomes using the sample code names to a new directory called exomes.<br><br><strong>Task 1.4 completed</strong>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### createCrisprReady.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the motif_list.txt file, identify the 3 highest occurring motifs in each exome inside the exomes folder. \n",
    "Output the headers and corresponding sequences to a new file called exomename_topmotifs.fasta.  **This means to only select for the genes that have at least one of the three top motifs and output to a new file\n",
    "Rename the headers of the exomename_topmotifs.fasta files to include the motif sequences and their count. (Example: >gene1  ===> >gene1 ATGC_1 AAAA_38)\n",
    "\n",
    "Note: The sequence does not need to contain all 3 motifs, it just needs to have at least 1 of them.\n",
    "“exomename_topmotifs.fasta” should contain the code name of the sample. For example, fox_topmotifs.fasta."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### identifyCrisprSite.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each gene inside the exomename_topmotifs.fasta files, identify a suitable CRISPR site.\n",
    "Find sequences that contain “NGG”, where “N” can be any base, that has at least 20 basepairs upstream.\n",
    "Example of upstream: ATGAACGTCTGTAAGAACTGCGGATCTGTCA (Everything left of CGG is upstream of the DNA) \n",
    "Output suitable candidates (headers and sequences) to a new file called exomenames_precrispr.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### editGenome.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using those files, write a script that will insert the letter A right before the NGG site. Output to a new file called exomename_postcrispr.fasta \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### exomeReport.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a single report that summarizes the findings. It should be a text file that lists the name of the discoverer of the organism, the diameter, the code name and the environment it came from. -Additionally, include how many genes the organism has in common with all of the other organisms, and which genes are unique to that organism.\n",
    "Note: You are only comparing that organism against the other organisms of the same cohort (20-30mm in diameter). ie You are summarizing the motif-containing exomes with just the CRISPR-ready genes, in otherwords using the exomenames_precrispr.fasta (or exomenames_postcrispr.fasta since the headers should be identical between those two files)\n",
    "There may not be any genes in common in that entire cohort.\n",
    "\n",
    "Example:\n",
    "Organism FOX, discovered by DISCOVERER, has a diameter of DIAMETER, and has the NUMBER of genes in common with the cohort, and the following genes unique to only itself: gene1, gene4, gene823"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Readme.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile Readme.txt\n",
    "\n",
    "Welcome to week4 homework! To run the script, step by step type: \n",
    "\n",
    "    ./copyExomes\n",
    "    ./createCrisprReady.sh\n",
    "    ./identifyCrisprSite.sh\n",
    "    ./editGenome.sh\n",
    "\n",
    "into terminal or type ./runAll.sh. If you use the runAll.sh pipe, it will\n",
    "delete old reports. \n",
    "\n",
    "About the report...I am not sure if you mean shared genes or motifs. It \n",
    "seems all genes are unique, but I will look at it tomorrow morning."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## copyExomes.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Structure Decision"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a little confusing to me Tim, we already have a file \n",
    "exomes. Do you want me to over write this folder and delete the \n",
    "Older .fasta files? \n",
    "\n",
    "Seems weird to overwrite everything like that and destroy\n",
    "The old .fasta files. So I am going to move the exomes folder to a \n",
    "exomes_given and make a new directory as asked."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Programing Decision"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It seems like I have two choices to make all of this work. I can make a bunch of little scripts that play together in a modular fashion or I can maybe a larger python script that is called in different bash scripts. For such a little project, it seems strained to make a bunch of little scripts to fry all together. Therefore, I am going to build one larger python script that will do most of the heavy lifting and call all of the parts of that script from different parts of the bash shell. <br><br>\n",
    "It looks like you are giving me a lot of latitude here, so I am going to use pandas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in clinical_data with pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def copy_exomes():\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    return df\n",
    "\n",
    "df = copy_exomes()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This looks rather clean, moving to filer by the requested diameters."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter dataframe by diameter size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "mover.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make exomes directory and copy over filtered files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(\"exomes\"):\n",
    "    os.mkdir(\"exomes\")\n",
    "else:\n",
    "    shutil.rmtree(\"exomes\")\n",
    "    os.mkdir(\"exomes\")\n",
    "    \n",
    "for i in mover[\"code_name\"].tolist():\n",
    "    print(\"copying {}.fasta file\".format(i))\n",
    "    shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                 \"exomes/{}.fasta\".format(i))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create motif_manager.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "\n",
    "def copy_exomes():\n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    print(\"The following entries are of interest:\")\n",
    "    print(mover)\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        print(\"copying {}.fasta file\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile copyExomes.sh\n",
    "#!/bin/bash\n",
    "\n",
    "python -c 'import motif_manager; motif_manager.copy_exomes()'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 1.4 Complete"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## createCrisprReady.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check size length of .fasta files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for fasta in os.listdir(\"exomes\"):\n",
    "    with open('exomes/{}'.format(fasta)) as f:\n",
    "        lines = [line.rstrip() for line in f]\n",
    "        print(len(lines))\n",
    "        \n",
    "with open(\"motif_list.txt\") as f:\n",
    "    motifs = [line.rstrip() for line in f]\n",
    "    \n",
    "print(motifs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is very helpful that all of these lists have the same size. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build Animal Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "animal_house = []\n",
    "animal_list = {}\n",
    "for fasta in os.listdir(\"exomes\"):\n",
    "    with open('exomes/{}'.format(fasta)) as f:\n",
    "        lines = [line.rstrip() for line in f]\n",
    "    animal_list[fasta] = lines\n",
    "    entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "    entry[\"name\"] = fasta[:-6]\n",
    "    animal_house.append(entry)\n",
    "\n",
    "        \n",
    "with open(\"motif_list.txt\") as f:\n",
    "    motifs = [line.rstrip() for line in f]\n",
    "        \n",
    "animal_house = pd.DataFrame(animal_house)\n",
    "animal_house.set_index('name', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "animal_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It looks like we have two different lines of attack here. First, I could make a tensor of all of the different motifs and then collapse the tensor by count. Also, I could crush down genes inside each animal and then explore the values. The tensor option seems the most fun, but I think we will stick to a plain Jane adding a column count inside of the dataframe."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build dataframe counter for each animal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(animal_house.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "animal_data = {}\n",
    "count_list = []\n",
    "top_per_animal = {}\n",
    "gene_dict = {}\n",
    "for animal in list(animal_house.index):\n",
    "    animal_data[animal] = animal_house.loc[[animal]].T\n",
    "    for motif in motifs:\n",
    "        animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "                \n",
    "    # Suming up columns and rows\n",
    "    animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "    \n",
    "    # There is a quirk in Pandas where the Text column is added, will drop\n",
    "    animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "    \n",
    "    # Casting to ints\n",
    "    animal_data[animal] = animal_data[animal].astype(int)\n",
    "    \n",
    "    \n",
    "    count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T)\n",
    "    top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "    \n",
    "    df = animal_data[animal][top_per_animal[animal]]\n",
    "    df = df[(df.T != 0).any()]\n",
    "    \n",
    "    gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "count_values = pd.concat(count_list, axis=1, sort = True)\n",
    "\n",
    "#Total sum per row: \n",
    "count_values.loc[:,'sum'] = count_values.sum(axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Top Ten Motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_values[\"sum\"].nlargest(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_values.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "count_values[\"sum\"].nlargest(10).plot.bar(x=\"name\");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Top motifs for each animal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_per_animal[\"lamb\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for animal in top_per_animal:\n",
    "    print(animal)\n",
    "    print(top_per_animal[animal])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we can see that the same three motifs do in fact dominate all\n",
    "of the various animals."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count Report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = animal_data[animal][top_per_animal[animal]]\n",
    "df = df[(df.T != 0).any()]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I do not really understand the reason why anyone would want this report type, versus something like a json or dataframe, but here you go. <br><br>\n",
    "I am taking the statement <b>The sequence does not need to contain all 3 motifs</b> to mean that I <b>can</b> include them. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I am running out of time, so I am going to build this all at once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "report.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Send the report dictionary to .txt files, not sure where you want them\n",
    "# Sending to different directory \"exomes_top\"\n",
    "\n",
    "# Make Directory Structure\n",
    "if not os.path.exists(\"exomes_top\"):\n",
    "    os.mkdir(\"exomes_top\")\n",
    "else:\n",
    "    shutil.rmtree(\"exomes_top\")\n",
    "    os.mkdir(\"exomes_top\")\n",
    "\n",
    "for file in report:\n",
    "    with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "        for line in report[file]:\n",
    "            crispy.write(\"{}\\n\".format(line))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create motif_manager.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "def copy_exomes():\n",
    " \n",
    "    now = datetime.datetime.now()\n",
    "    \n",
    "    # Making Report as we go\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nAnalysis Run Started:\\n\" + now.strftime(\"%c\") + \"\\n----------\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    \n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"The following entries are of interest:\\n\")\n",
    "        f.write(\"{}\".format(mover) + \"\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"copying {}.fasta file\\n\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))\n",
    "        \n",
    "def create_crispy():\n",
    "    animal_apartment = []\n",
    "    animal_list = {}\n",
    "    gene_dict = {}\n",
    "\n",
    "    #First build dataframe animals in exome dir\n",
    "    for fasta in os.listdir(\"exomes\"):\n",
    "        with open('exomes/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        animal_list[fasta[:-6]] = lines\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "        entry[\"name\"] = fasta[:-6]\n",
    "        animal_apartment.append(entry)\n",
    "\n",
    "    animal_house = pd.DataFrame(animal_apartment)\n",
    "    animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "    #Iterate through motif_list\n",
    "    with open(\"motif_list.txt\") as f:\n",
    "        motifs = [line.rstrip() for line in f]\n",
    "\n",
    "    # Build counting dataframe for each animal\n",
    "    animal_data = {}\n",
    "    count_list = []\n",
    "    top_per_animal = {}\n",
    "\n",
    "    for animal in list(animal_house.index):\n",
    "        animal_data[animal] = animal_house.loc[[animal]].T\n",
    "        for motif in motifs:\n",
    "            animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "        # Suming up columns and rows\n",
    "        animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "        # There is a quirk in Pandas where the Text column is added, will drop\n",
    "        animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "        # Casting to ints\n",
    "        animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "        # Add the sum into a count_column list\n",
    "        count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "        # Populate top 3 lists\n",
    "        top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "        # Look at only the populated data lists\n",
    "        df = animal_data[animal][top_per_animal[animal]]\n",
    "        df = df[(df.T != 0).any()]\n",
    "\n",
    "        # Make animal dict to convert back to text file\n",
    "        gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "    # Join all of the columns into one dataframe\n",
    "    count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "    #Total sum per row: \n",
    "    count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "    report_dict = {}\n",
    "\n",
    "    # Build a report dictionary \n",
    "    for animal in gene_dict:\n",
    "        report_dict[animal] = {}\n",
    "        for gene in gene_dict[animal]:\n",
    "            d = gene_dict[animal][gene]\n",
    "\n",
    "            #Flatten Dictionary\n",
    "            report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "    # Build the final report with proper formating\n",
    "    report = {}\n",
    "    for animal in animal_list:\n",
    "        text_list = []\n",
    "        for line in animal_list[animal]:\n",
    "            if line in report_dict[animal]:\n",
    "                text_list.append(line + \" \" + report_dict[animal][line])\n",
    "                text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "        report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "    # Send the report dictionary to .txt files, not sure where you want them\n",
    "    # Sending to different directory \"exomes_top\"\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_top\"):\n",
    "        os.mkdir(\"exomes_top\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_top\")\n",
    "        os.mkdir(\"exomes_top\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in report:\n",
    "        with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "            for line in report[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nTop Ten Motifs\\n\")\n",
    "        f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "        f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "        for animal in top_per_animal:\n",
    "            f.write(\"\\n{}\\n\".format(animal))\n",
    "            f.write(\"{}\".format(top_per_animal[animal]))\n",
    "        f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile createCrisprReady.sh\n",
    "#!/bin/bash\n",
    "\n",
    "python -c 'import motif_manager; motif_manager.create_crispy()'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## identifyCrisprSite.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I fell way behind on the time, so to catch up I am just going to script this right out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "def copy_exomes():\n",
    " \n",
    "    now = datetime.datetime.now()\n",
    "    \n",
    "    # Making Report as we go\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nAnalysis Run Started:\\n\" + now.strftime(\"%c\") + \"\\n----------\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    \n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"The following entries are of interest:\\n\")\n",
    "        f.write(\"{}\".format(mover) + \"\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"copying {}.fasta file\\n\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))\n",
    "        \n",
    "def create_crispy():\n",
    "    animal_apartment = []\n",
    "    animal_list = {}\n",
    "    gene_dict = {}\n",
    "\n",
    "    #First build dataframe animals in exome dir\n",
    "    for fasta in os.listdir(\"exomes\"):\n",
    "        with open('exomes/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        animal_list[fasta[:-6]] = lines\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "        entry[\"name\"] = fasta[:-6]\n",
    "        animal_apartment.append(entry)\n",
    "\n",
    "    animal_house = pd.DataFrame(animal_apartment)\n",
    "    animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "    #Iterate through motif_list\n",
    "    with open(\"motif_list.txt\") as f:\n",
    "        motifs = [line.rstrip() for line in f]\n",
    "\n",
    "    # Build counting dataframe for each animal\n",
    "    animal_data = {}\n",
    "    count_list = []\n",
    "    top_per_animal = {}\n",
    "\n",
    "    for animal in list(animal_house.index):\n",
    "        animal_data[animal] = animal_house.loc[[animal]].T\n",
    "        for motif in motifs:\n",
    "            animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "        # Suming up columns and rows\n",
    "        animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "        # There is a quirk in Pandas where the Text column is added, will drop\n",
    "        animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "        # Casting to ints\n",
    "        animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "        # Add the sum into a count_column list\n",
    "        count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "        # Populate top 3 lists\n",
    "        top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "        # Look at only the populated data lists\n",
    "        df = animal_data[animal][top_per_animal[animal]]\n",
    "        df = df[(df.T != 0).any()]\n",
    "\n",
    "        # Make animal dict to convert back to text file\n",
    "        gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "    # Join all of the columns into one dataframe\n",
    "    count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "    #Total sum per row: \n",
    "    count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "    report_dict = {}\n",
    "\n",
    "    # Build a report dictionary \n",
    "    for animal in gene_dict:\n",
    "        report_dict[animal] = {}\n",
    "        for gene in gene_dict[animal]:\n",
    "            d = gene_dict[animal][gene]\n",
    "\n",
    "            #Flatten Dictionary\n",
    "            report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "    # Build the final report with proper formating\n",
    "    report = {}\n",
    "    for animal in animal_list:\n",
    "        text_list = []\n",
    "        for line in animal_list[animal]:\n",
    "            if line in report_dict[animal]:\n",
    "                text_list.append(line + \" \" + report_dict[animal][line])\n",
    "                text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "        report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "    # Send the report dictionary to .txt files, not sure where you want them\n",
    "    # Sending to different directory \"exomes_top\"\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_top\"):\n",
    "        os.mkdir(\"exomes_top\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_top\")\n",
    "        os.mkdir(\"exomes_top\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in report:\n",
    "        with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "            for line in report[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nTop Ten Motifs\\n\")\n",
    "        f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "        f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "        for animal in top_per_animal:\n",
    "            f.write(\"\\n{}\\n\".format(animal))\n",
    "            f.write(\"{}\".format(top_per_animal[animal]))\n",
    "        f.write(\"\\n\")\n",
    "        f.close()\n",
    "        \n",
    "def locate():\n",
    "    \n",
    "    top_house = {}\n",
    "    soggy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_top\"):\n",
    "        with open('exomes_top/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        top_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in top_house:\n",
    "        soggy_list = []\n",
    "        have_gg = 0\n",
    "        no_gg = 0\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"\\nSearching {} for NGG\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in top_house[animal]:\n",
    "            candidate = top_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                soggy_list.append(gene)\n",
    "                soggy_list.append(candidate)\n",
    "                have_gg += 1\n",
    "\n",
    "                # Reporter\n",
    "                with open(\"report.txt\", \"a+\") as f:\n",
    "                    f.write(\"Found NGG at {}\\n\".format(gene.split(\" \")[0]))\n",
    "                    f.close()\n",
    "\n",
    "            else:\n",
    "                no_gg += 1\n",
    "\n",
    "            # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"On {}, found {} NGG genes out of {} candidates\\n\".format(animal,\n",
    "                                                                              have_gg,\n",
    "                                                                              have_gg + no_gg))\n",
    "            f.write(\"-----------------\\n\")\n",
    "            f.close()\n",
    "        soggy_dict[animal] = soggy_list\n",
    "\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_precrispr\"):\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_precrispr\")\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in soggy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_precrispr.fasta\"\n",
    "        with open(\"exomes_precrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in soggy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile identifyCrisprSite.sh\n",
    "#!/bin/bash\n",
    "\n",
    "python -c 'import motif_manager; motif_manager.locate()'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## editGenome.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Two quick assumptions, are we adding an A to multiple sites? I am going to assume yes. Next, are we inserting before the GG (NAGG) or before NGG (ANGG). I am assuming NAGG. <br><br>\n",
    "\n",
    "Probably could have done this all at once, but no big deal to do it in two steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "import Levenshtein as lev\n",
    "\n",
    "def copy_exomes():\n",
    " \n",
    "    now = datetime.datetime.now()\n",
    "    \n",
    "    # Making Report as we go\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nAnalysis Run Started:\\n\" + now.strftime(\"%c\") + \"\\n----------\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    \n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"The following entries are of interest:\\n\")\n",
    "        f.write(\"{}\".format(mover) + \"\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"copying {}.fasta file\\n\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))\n",
    "        \n",
    "def create_crispy():\n",
    "    animal_apartment = []\n",
    "    animal_list = {}\n",
    "    gene_dict = {}\n",
    "\n",
    "    #First build dataframe animals in exome dir\n",
    "    for fasta in os.listdir(\"exomes\"):\n",
    "        with open('exomes/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        animal_list[fasta[:-6]] = lines\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "        entry[\"name\"] = fasta[:-6]\n",
    "        animal_apartment.append(entry)\n",
    "\n",
    "    animal_house = pd.DataFrame(animal_apartment)\n",
    "    animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "    #Iterate through motif_list\n",
    "    with open(\"motif_list.txt\") as f:\n",
    "        motifs = [line.rstrip() for line in f]\n",
    "\n",
    "    # Build counting dataframe for each animal\n",
    "    animal_data = {}\n",
    "    count_list = []\n",
    "    top_per_animal = {}\n",
    "\n",
    "    for animal in list(animal_house.index):\n",
    "        animal_data[animal] = animal_house.loc[[animal]].T\n",
    "        for motif in motifs:\n",
    "            animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "        # Suming up columns and rows\n",
    "        animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "        # There is a quirk in Pandas where the Text column is added, will drop\n",
    "        animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "        # Casting to ints\n",
    "        animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "        # Add the sum into a count_column list\n",
    "        count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "        # Populate top 3 lists\n",
    "        top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "        # Look at only the populated data lists\n",
    "        df = animal_data[animal][top_per_animal[animal]]\n",
    "        df = df[(df.T != 0).any()]\n",
    "\n",
    "        # Make animal dict to convert back to text file\n",
    "        gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "    # Join all of the columns into one dataframe\n",
    "    count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "    #Total sum per row: \n",
    "    count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "    report_dict = {}\n",
    "\n",
    "    # Build a report dictionary \n",
    "    for animal in gene_dict:\n",
    "        report_dict[animal] = {}\n",
    "        for gene in gene_dict[animal]:\n",
    "            d = gene_dict[animal][gene]\n",
    "\n",
    "            #Flatten Dictionary\n",
    "            report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "    # Build the final report with proper formating\n",
    "    report = {}\n",
    "    for animal in animal_list:\n",
    "        text_list = []\n",
    "        for line in animal_list[animal]:\n",
    "            if line in report_dict[animal]:\n",
    "                text_list.append(line + \" \" + report_dict[animal][line])\n",
    "                text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "        report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "    # Send the report dictionary to .txt files, not sure where you want them\n",
    "    # Sending to different directory \"exomes_top\"\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_top\"):\n",
    "        os.mkdir(\"exomes_top\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_top\")\n",
    "        os.mkdir(\"exomes_top\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in report:\n",
    "        with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "            for line in report[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "    similar = animal_house.T.to_dict()\n",
    "    count_dict = {}\n",
    "    gene_list = []\n",
    "\n",
    "    # First make a list of all genes\n",
    "    for animal in similar:\n",
    "        for gene in similar[animal]:\n",
    "            gene_list.append(similar[animal][gene])\n",
    "\n",
    "    # Search for duplicates\n",
    "    for animal in similar:\n",
    "        count_dict[animal] = {}\n",
    "        for gene in similar[animal]:\n",
    "            count = gene_list.count(similar[animal][gene]) - 1\n",
    "            count_dict[animal][gene] = count\n",
    "            if count > 0: print(\"shared gene!\")\n",
    "\n",
    "    shared_genes = pd.DataFrame(count_dict)\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nShared Genes\\n\")\n",
    "        f.write(\"{}\".format(shared_genes[shared_genes != 0].dropna()))\n",
    "        f.write(\"\\n\\nTop Ten Motifs\\n\")\n",
    "        f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "        f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "        for animal in top_per_animal:\n",
    "            f.write(\"\\n{}\\n\".format(animal))\n",
    "            f.write(\"{}\".format(top_per_animal[animal]))\n",
    "        f.write(\"\\n\")\n",
    "        f.close()\n",
    "        \n",
    "def locate():\n",
    "    \n",
    "    top_house = {}\n",
    "    soggy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_top\"):\n",
    "        with open('exomes_top/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        top_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in top_house:\n",
    "        soggy_list = []\n",
    "        have_gg = 0\n",
    "        no_gg = 0\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"\\nSearching {} for NGG\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in top_house[animal]:\n",
    "            candidate = top_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                soggy_list.append(gene)\n",
    "                soggy_list.append(candidate)\n",
    "                have_gg += 1\n",
    "\n",
    "                # Reporter\n",
    "                with open(\"report.txt\", \"a+\") as f:\n",
    "                    f.write(\"Found NGG at {}\\n\".format(gene.split(\" \")[0]))\n",
    "                    f.close()\n",
    "\n",
    "            else:\n",
    "                no_gg += 1\n",
    "\n",
    "            # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"On {}, found {} NGG genes out of {} candidates\\n\".format(animal,\n",
    "                                                                              have_gg,\n",
    "                                                                              have_gg + no_gg))\n",
    "            f.write(\"-----------------\\n\")\n",
    "            f.close()\n",
    "        soggy_dict[animal] = soggy_list\n",
    "\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_precrispr\"):\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_precrispr\")\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in soggy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_precrispr.fasta\"\n",
    "        with open(\"exomes_precrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in soggy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "                \n",
    "def edit_genome():\n",
    "    big_house = {}\n",
    "    crispy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_precrispr\"):\n",
    "        with open('exomes_precrispr/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        big_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in big_house:\n",
    "        crispy_list = []\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"Inserting A at NGG sites on file {}\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in big_house[animal]:\n",
    "            candidate = big_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "            for ind in indices:\n",
    "                candidate = candidate[:ind] + \"A\" + candidate[ind:]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                crispy_list.append(gene)\n",
    "                crispy_list.append(candidate)\n",
    "\n",
    "        crispy_dict[animal] = crispy_list\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_postcrispr\"):\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_postcrispr\")\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in crispy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_postcrispr.fasta\"\n",
    "        with open(\"exomes_postcrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in crispy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile editGenome.sh\n",
    "#!/bin/bash\n",
    "\n",
    "python -c 'import motif_manager; motif_manager.edit_genome()'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile runAll.sh\n",
    "#!/bin/bash\n",
    "\n",
    "echo \"removing old reports\"\n",
    "rm report.txt\n",
    "./copyExomes.sh\n",
    "./createCrisprReady.sh\n",
    "./identifyCrisprSite.sh\n",
    "./editGenome.sh\n",
    "echo \"completed\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## exomeReport.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "import Levenshtein as lev\n",
    "\n",
    "def copy_exomes():\n",
    " \n",
    "    now = datetime.datetime.now()\n",
    "    \n",
    "    # Making Report as we go\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nAnalysis Run Started:\\n\" + now.strftime(\"%c\") + \"\\n----------\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    \n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"The following entries are of interest:\\n\")\n",
    "        f.write(\"{}\".format(mover) + \"\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"copying {}.fasta file\\n\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))\n",
    "        \n",
    "def create_crispy():\n",
    "    animal_apartment = []\n",
    "    animal_list = {}\n",
    "    gene_dict = {}\n",
    "\n",
    "    #First build dataframe animals in exome dir\n",
    "    for fasta in os.listdir(\"exomes\"):\n",
    "        with open('exomes/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        animal_list[fasta[:-6]] = lines\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "        entry[\"name\"] = fasta[:-6]\n",
    "        animal_apartment.append(entry)\n",
    "\n",
    "    animal_house = pd.DataFrame(animal_apartment)\n",
    "    animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "    #Iterate through motif_list\n",
    "    with open(\"motif_list.txt\") as f:\n",
    "        motifs = [line.rstrip() for line in f]\n",
    "\n",
    "    # Build counting dataframe for each animal\n",
    "    animal_data = {}\n",
    "    count_list = []\n",
    "    top_per_animal = {}\n",
    "\n",
    "    for animal in list(animal_house.index):\n",
    "        animal_data[animal] = animal_house.loc[[animal]].T\n",
    "        for motif in motifs:\n",
    "            animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "        # Suming up columns and rows\n",
    "        animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "        # There is a quirk in Pandas where the Text column is added, will drop\n",
    "        animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "        # Casting to ints\n",
    "        animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "        # Add the sum into a count_column list\n",
    "        count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "        # Populate top 3 lists\n",
    "        top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "        # Look at only the populated data lists\n",
    "        df = animal_data[animal][top_per_animal[animal]]\n",
    "        df = df[(df.T != 0).any()]\n",
    "\n",
    "        # Make animal dict to convert back to text file\n",
    "        gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "    # Join all of the columns into one dataframe\n",
    "    count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "    #Total sum per row: \n",
    "    count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "    report_dict = {}\n",
    "\n",
    "    # Build a report dictionary \n",
    "    for animal in gene_dict:\n",
    "        report_dict[animal] = {}\n",
    "        for gene in gene_dict[animal]:\n",
    "            d = gene_dict[animal][gene]\n",
    "\n",
    "            #Flatten Dictionary\n",
    "            report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "    # Build the final report with proper formating\n",
    "    report = {}\n",
    "    for animal in animal_list:\n",
    "        text_list = []\n",
    "        for line in animal_list[animal]:\n",
    "            if line in report_dict[animal]:\n",
    "                text_list.append(line + \" \" + report_dict[animal][line])\n",
    "                text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "        report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "    # Send the report dictionary to .txt files, not sure where you want them\n",
    "    # Sending to different directory \"exomes_top\"\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_top\"):\n",
    "        os.mkdir(\"exomes_top\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_top\")\n",
    "        os.mkdir(\"exomes_top\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in report:\n",
    "        with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "            for line in report[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "    similar = animal_house.T.to_dict()\n",
    "    count_dict = {}\n",
    "    gene_list = []\n",
    "\n",
    "    # First make a list of all genes\n",
    "    for animal in similar:\n",
    "        for gene in similar[animal]:\n",
    "            gene_list.append(similar[animal][gene])\n",
    "\n",
    "    # Search for duplicates\n",
    "    for animal in similar:\n",
    "        count_dict[animal] = {}\n",
    "        for gene in similar[animal]:\n",
    "            count = gene_list.count(similar[animal][gene]) - 1\n",
    "            count_dict[animal][gene] = count\n",
    "            if count > 0: print(\"shared gene!\")\n",
    "\n",
    "    shared_genes = pd.DataFrame(count_dict)\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nShared Genes\\n\")\n",
    "        f.write(\"{}\".format(shared_genes[shared_genes != 0].dropna()))\n",
    "        f.write(\"\\n\\nTop Ten Motifs\\n\")\n",
    "        f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "        f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "        for animal in top_per_animal:\n",
    "            f.write(\"\\n{}\\n\".format(animal))\n",
    "            f.write(\"{}\".format(top_per_animal[animal]))\n",
    "        f.write(\"\\n\")\n",
    "        f.close()\n",
    "        \n",
    "def locate():\n",
    "    \n",
    "    top_house = {}\n",
    "    soggy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_top\"):\n",
    "        with open('exomes_top/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        top_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in top_house:\n",
    "        soggy_list = []\n",
    "        have_gg = 0\n",
    "        no_gg = 0\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"\\nSearching {} for NGG\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in top_house[animal]:\n",
    "            candidate = top_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                soggy_list.append(gene)\n",
    "                soggy_list.append(candidate)\n",
    "                have_gg += 1\n",
    "\n",
    "                # Reporter\n",
    "                with open(\"report.txt\", \"a+\") as f:\n",
    "                    f.write(\"Found NGG at {}\\n\".format(gene.split(\" \")[0]))\n",
    "                    f.close()\n",
    "\n",
    "            else:\n",
    "                no_gg += 1\n",
    "\n",
    "            # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"On {}, found {} NGG genes out of {} candidates\\n\".format(animal,\n",
    "                                                                              have_gg,\n",
    "                                                                              have_gg + no_gg))\n",
    "            f.write(\"-----------------\\n\")\n",
    "            f.close()\n",
    "        soggy_dict[animal] = soggy_list\n",
    "\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_precrispr\"):\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_precrispr\")\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in soggy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_precrispr.fasta\"\n",
    "        with open(\"exomes_precrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in soggy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "                \n",
    "def edit_genome():\n",
    "    big_house = {}\n",
    "    crispy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_precrispr\"):\n",
    "        with open('exomes_precrispr/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        big_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in big_house:\n",
    "        crispy_list = []\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"Inserting A at NGG sites on file {}\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in big_house[animal]:\n",
    "            candidate = big_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "            for ind in indices:\n",
    "                candidate = candidate[:ind] + \"A\" + candidate[ind:]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                crispy_list.append(gene)\n",
    "                crispy_list.append(candidate)\n",
    "\n",
    "        crispy_dict[animal] = crispy_list\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_postcrispr\"):\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_postcrispr\")\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in crispy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_postcrispr.fasta\"\n",
    "        with open(\"exomes_postcrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in crispy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above is not the best way to measure distance. I will look at Levenshtein distance measure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Levenshtein Distance Measure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "import Levenshtein as lev\n",
    "\n",
    "def copy_exomes():\n",
    " \n",
    "    now = datetime.datetime.now()\n",
    "    \n",
    "    # Making Report as we go\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nAnalysis Run Started:\\n\" + now.strftime(\"%c\") + \"\\n----------\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    \n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"The following entries are of interest:\\n\")\n",
    "        f.write(\"{}\".format(mover) + \"\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"copying {}.fasta file\\n\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))\n",
    "        \n",
    "def create_crispy():\n",
    "    animal_apartment = []\n",
    "    animal_list = {}\n",
    "    gene_dict = {}\n",
    "\n",
    "    #First build dataframe animals in exome dir\n",
    "    for fasta in os.listdir(\"exomes\"):\n",
    "        with open('exomes/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        animal_list[fasta[:-6]] = lines\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "        entry[\"name\"] = fasta[:-6]\n",
    "        animal_apartment.append(entry)\n",
    "\n",
    "    animal_house = pd.DataFrame(animal_apartment)\n",
    "    animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "    #Iterate through motif_list\n",
    "    with open(\"motif_list.txt\") as f:\n",
    "        motifs = [line.rstrip() for line in f]\n",
    "\n",
    "    # Build counting dataframe for each animal\n",
    "    animal_data = {}\n",
    "    count_list = []\n",
    "    top_per_animal = {}\n",
    "\n",
    "    for animal in list(animal_house.index):\n",
    "        animal_data[animal] = animal_house.loc[[animal]].T\n",
    "        for motif in motifs:\n",
    "            animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "        # Suming up columns and rows\n",
    "        animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "        # There is a quirk in Pandas where the Text column is added, will drop\n",
    "        animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "        # Casting to ints\n",
    "        animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "        # Add the sum into a count_column list\n",
    "        count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "        # Populate top 3 lists\n",
    "        top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "        # Look at only the populated data lists\n",
    "        df = animal_data[animal][top_per_animal[animal]]\n",
    "        df = df[(df.T != 0).any()]\n",
    "\n",
    "        # Make animal dict to convert back to text file\n",
    "        gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "    # Join all of the columns into one dataframe\n",
    "    count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "    #Total sum per row: \n",
    "    count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "    report_dict = {}\n",
    "\n",
    "    # Build a report dictionary \n",
    "    for animal in gene_dict:\n",
    "        report_dict[animal] = {}\n",
    "        for gene in gene_dict[animal]:\n",
    "            d = gene_dict[animal][gene]\n",
    "\n",
    "            #Flatten Dictionary\n",
    "            report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "    # Build the final report with proper formating\n",
    "    report = {}\n",
    "    for animal in animal_list:\n",
    "        text_list = []\n",
    "        for line in animal_list[animal]:\n",
    "            if line in report_dict[animal]:\n",
    "                text_list.append(line + \" \" + report_dict[animal][line])\n",
    "                text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "        report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "    # Send the report dictionary to .txt files, not sure where you want them\n",
    "    # Sending to different directory \"exomes_top\"\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_top\"):\n",
    "        os.mkdir(\"exomes_top\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_top\")\n",
    "        os.mkdir(\"exomes_top\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in report:\n",
    "        with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "            for line in report[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "    similar = animal_house.T.to_dict()\n",
    "    count_dict = {}\n",
    "    gene_list = []\n",
    "\n",
    "    # First make a list of all genes\n",
    "    for animal in similar:\n",
    "        for gene in similar[animal]:\n",
    "            gene_list.append(similar[animal][gene])\n",
    "\n",
    "    # Search for duplicates\n",
    "    for animal in similar:\n",
    "        count_dict[animal] = {}\n",
    "        for gene in similar[animal]:\n",
    "            count = gene_list.count(similar[animal][gene]) - 1\n",
    "            count_dict[animal][gene] = count\n",
    "            if count > 0: print(\"shared gene!\")\n",
    "\n",
    "    shared_genes = pd.DataFrame(count_dict)\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nShared Genes\\n\")\n",
    "        f.write(\"{}\".format(shared_genes[shared_genes != 0].dropna()))\n",
    "        f.write(\"\\n\\nTop Ten Motifs\\n\")\n",
    "        f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "        f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "        for animal in top_per_animal:\n",
    "            f.write(\"\\n{}\\n\".format(animal))\n",
    "            f.write(\"{}\".format(top_per_animal[animal]))\n",
    "        f.write(\"\\n\")\n",
    "        f.close()\n",
    "        \n",
    "def locate():\n",
    "    \n",
    "    top_house = {}\n",
    "    soggy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_top\"):\n",
    "        with open('exomes_top/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        top_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in top_house:\n",
    "        soggy_list = []\n",
    "        have_gg = 0\n",
    "        no_gg = 0\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"\\nSearching {} for NGG\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in top_house[animal]:\n",
    "            candidate = top_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                soggy_list.append(gene)\n",
    "                soggy_list.append(candidate)\n",
    "                have_gg += 1\n",
    "\n",
    "                # Reporter\n",
    "                with open(\"report.txt\", \"a+\") as f:\n",
    "                    f.write(\"Found NGG at {}\\n\".format(gene.split(\" \")[0]))\n",
    "                    f.close()\n",
    "\n",
    "            else:\n",
    "                no_gg += 1\n",
    "\n",
    "            # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"On {}, found {} NGG genes out of {} candidates\\n\".format(animal,\n",
    "                                                                              have_gg,\n",
    "                                                                              have_gg + no_gg))\n",
    "            f.write(\"-----------------\\n\")\n",
    "            f.close()\n",
    "        soggy_dict[animal] = soggy_list\n",
    "\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_precrispr\"):\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_precrispr\")\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in soggy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_precrispr.fasta\"\n",
    "        with open(\"exomes_precrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in soggy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "                \n",
    "def edit_genome():\n",
    "    big_house = {}\n",
    "    crispy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_precrispr\"):\n",
    "        with open('exomes_precrispr/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        big_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in big_house:\n",
    "        crispy_list = []\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"Inserting A at NGG sites on file {}\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in big_house[animal]:\n",
    "            candidate = big_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "            for ind in indices:\n",
    "                candidate = candidate[:ind] + \"A\" + candidate[ind:]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                crispy_list.append(gene)\n",
    "                crispy_list.append(candidate)\n",
    "\n",
    "        crispy_dict[animal] = crispy_list\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_postcrispr\"):\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_postcrispr\")\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in crispy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_postcrispr.fasta\"\n",
    "        with open(\"exomes_postcrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in crispy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "animal_apartment = []\n",
    "animal_list = {}\n",
    "gene_dict = {}\n",
    "\n",
    "#First build dataframe animals in exome dir\n",
    "for fasta in os.listdir(\"exomes\"):\n",
    "    with open('exomes/{}'.format(fasta)) as f:\n",
    "        lines = [line.rstrip() for line in f]\n",
    "\n",
    "    animal_list[fasta[:-6]] = lines\n",
    "    entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "    entry[\"name\"] = fasta[:-6]\n",
    "    animal_apartment.append(entry)\n",
    "\n",
    "animal_house = pd.DataFrame(animal_apartment)\n",
    "animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "#Iterate through motif_list\n",
    "with open(\"motif_list.txt\") as f:\n",
    "    motifs = [line.rstrip() for line in f]\n",
    "\n",
    "# Build counting dataframe for each animal\n",
    "animal_data = {}\n",
    "count_list = []\n",
    "top_per_animal = {}\n",
    "\n",
    "for animal in list(animal_house.index):\n",
    "    animal_data[animal] = animal_house.loc[[animal]].T\n",
    "    for motif in motifs:\n",
    "        animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "    # Suming up columns and rows\n",
    "    animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "    # There is a quirk in Pandas where the Text column is added, will drop\n",
    "    animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "    # Casting to ints\n",
    "    animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "    # Add the sum into a count_column list\n",
    "    count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "    # Populate top 3 lists\n",
    "    top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "    # Look at only the populated data lists\n",
    "    df = animal_data[animal][top_per_animal[animal]]\n",
    "    df = df[(df.T != 0).any()]\n",
    "\n",
    "    # Make animal dict to convert back to text file\n",
    "    gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "# Join all of the columns into one dataframe\n",
    "count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "#Total sum per row: \n",
    "count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "report_dict = {}\n",
    "\n",
    "# Build a report dictionary \n",
    "for animal in gene_dict:\n",
    "    report_dict[animal] = {}\n",
    "    for gene in gene_dict[animal]:\n",
    "        d = gene_dict[animal][gene]\n",
    "\n",
    "        #Flatten Dictionary\n",
    "        report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "# Build the final report with proper formating\n",
    "report = {}\n",
    "for animal in animal_list:\n",
    "    text_list = []\n",
    "    for line in animal_list[animal]:\n",
    "        if line in report_dict[animal]:\n",
    "            text_list.append(line + \" \" + report_dict[animal][line])\n",
    "            text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "    report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "# Send the report dictionary to .txt files, not sure where you want them\n",
    "# Sending to different directory \"exomes_top\"\n",
    "\n",
    "# Make Directory Structure\n",
    "if not os.path.exists(\"exomes_top\"):\n",
    "    os.mkdir(\"exomes_top\")\n",
    "else:\n",
    "    shutil.rmtree(\"exomes_top\")\n",
    "    os.mkdir(\"exomes_top\")\n",
    "\n",
    "# Create Files\n",
    "for file in report:\n",
    "    with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "        for line in report[file]:\n",
    "            crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "similar = animal_house.T.to_dict()\n",
    "count_dict = {}\n",
    "gene_list = []\n",
    "\n",
    "# First make a list of all genes\n",
    "for animal in similar:\n",
    "    for gene in similar[animal]:\n",
    "        gene_list.append(similar[animal][gene])\n",
    "\n",
    "# Search for duplicates\n",
    "for animal in similar:\n",
    "    count_dict[animal] = {}\n",
    "    for gene in similar[animal]:\n",
    "        count = gene_list.count(similar[animal][gene]) - 1\n",
    "        count_dict[animal][gene] = count\n",
    "        if count > 0: print(\"shared gene!\")\n",
    "\n",
    "shared_genes = pd.DataFrame(count_dict)\n",
    "\n",
    "with open(\"report.txt\", \"a+\") as f:\n",
    "    f.write(\"\\nShared Genes\\n\")\n",
    "    f.write(\"{}\".format(shared_genes[shared_genes != 0].dropna()))\n",
    "    f.write(\"\\n\\nTop Ten Motifs\\n\")\n",
    "    f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "    f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "    for animal in top_per_animal:\n",
    "        f.write(\"\\n{}\\n\".format(animal))\n",
    "        f.write(\"{}\".format(top_per_animal[animal]))\n",
    "    f.write(\"\\n\")\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "animal_book = animal_house.T.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Levenshtein as lev\n",
    "import pickle\n",
    "\n",
    "distance_dict = {}\n",
    "for animal in animal_book:\n",
    "    for gene in animal_book[animal]:\n",
    "        distance_dict[animal + \"_\" + gene] = animal_book[animal][gene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'lamb'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(distance_dict.keys())[0].split(\"_\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lamb_>gene1 complete\n",
      "lamb_>gene2 complete\n",
      "lamb_>gene3 complete\n",
      "lamb_>gene4 complete\n",
      "lamb_>gene5 complete\n",
      "lamb_>gene6 complete\n",
      "lamb_>gene7 complete\n",
      "lamb_>gene8 complete\n",
      "lamb_>gene9 complete\n",
      "lamb_>gene10 complete\n",
      "lamb_>gene11 complete\n",
      "lamb_>gene12 complete\n",
      "lamb_>gene13 complete\n",
      "lamb_>gene14 complete\n",
      "lamb_>gene15 complete\n",
      "lamb_>gene16 complete\n",
      "lamb_>gene17 complete\n",
      "lamb_>gene18 complete\n",
      "lamb_>gene19 complete\n",
      "lamb_>gene20 complete\n",
      "lamb_>gene21 complete\n",
      "lamb_>gene22 complete\n",
      "lamb_>gene23 complete\n",
      "lamb_>gene24 complete\n",
      "lamb_>gene25 complete\n",
      "lamb_>gene26 complete\n",
      "lamb_>gene27 complete\n",
      "lamb_>gene28 complete\n",
      "lamb_>gene29 complete\n",
      "lamb_>gene30 complete\n",
      "lamb_>gene31 complete\n",
      "lamb_>gene32 complete\n",
      "lamb_>gene33 complete\n",
      "lamb_>gene34 complete\n",
      "lamb_>gene35 complete\n",
      "lamb_>gene36 complete\n",
      "lamb_>gene37 complete\n",
      "lamb_>gene38 complete\n",
      "lamb_>gene39 complete\n",
      "lamb_>gene40 complete\n",
      "lamb_>gene41 complete\n",
      "lamb_>gene42 complete\n",
      "lamb_>gene43 complete\n",
      "lamb_>gene44 complete\n",
      "lamb_>gene45 complete\n",
      "lamb_>gene46 complete\n",
      "lamb_>gene47 complete\n",
      "lamb_>gene48 complete\n",
      "lamb_>gene49 complete\n",
      "lamb_>gene50 complete\n",
      "lamb_>gene51 complete\n",
      "lamb_>gene52 complete\n",
      "lamb_>gene53 complete\n",
      "lamb_>gene54 complete\n",
      "lamb_>gene55 complete\n",
      "lamb_>gene56 complete\n",
      "lamb_>gene57 complete\n",
      "lamb_>gene58 complete\n",
      "lamb_>gene59 complete\n",
      "lamb_>gene60 complete\n",
      "lamb_>gene61 complete\n",
      "lamb_>gene62 complete\n",
      "lamb_>gene63 complete\n",
      "lamb_>gene64 complete\n",
      "lamb_>gene65 complete\n",
      "lamb_>gene66 complete\n",
      "lamb_>gene67 complete\n",
      "lamb_>gene68 complete\n",
      "lamb_>gene69 complete\n",
      "lamb_>gene70 complete\n",
      "lamb_>gene71 complete\n",
      "lamb_>gene72 complete\n"
     ]
    }
   ],
   "source": [
    "# This is very inefficent, I sloppily ran this, \n",
    "# should have triangulaged first\n",
    "\n",
    "import json\n",
    "\n",
    "animal_class = \"\"\n",
    "square_dict = {}\n",
    "for animal_1 in distance_dict:\n",
    "    animal = animal_1.split(\"_\")[0]\n",
    "    if animal_class != animal:\n",
    "        animal_class = animal\n",
    "        with open('distance_{}.json'.format(animal), 'w') as fp:\n",
    "            json.dump(square_dict, fp)\n",
    "        square_dict = {}\n",
    "    square_dict[animal_1] = {}\n",
    "    for animal_2 in distance_dict:\n",
    "        square_dict[animal_1][animal_2] = lev.distance(distance_dict[animal_1], \n",
    "                                                       distance_dict[animal_2])\n",
    "    print(animal_1+\" complete\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "square_dict.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "square_dict = {}\n",
    "for animal_1 in distance_dict:\n",
    "    square_dict[animal_1] = {}\n",
    "    for animal_2 in distance_dict:\n",
    "        square_dict[animal_1][animal_2] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "square_dict.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d[9] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
