{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 4 Assignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Description\" data-toc-modified-id=\"Description-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Description</a></span><ul class=\"toc-item\"><li><span><a href=\"#readme.txt\" data-toc-modified-id=\"readme.txt-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>readme.txt</a></span></li><li><span><a href=\"#Story\" data-toc-modified-id=\"Story-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Story</a></span></li><li><span><a href=\"#File-Structure\" data-toc-modified-id=\"File-Structure-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>File Structure</a></span></li><li><span><a href=\"#copyExomes.sh\" data-toc-modified-id=\"copyExomes.sh-1.4\"><span class=\"toc-item-num\">1.4&nbsp;&nbsp;</span>copyExomes.sh</a></span></li><li><span><a href=\"#createCrisprReady.sh\" data-toc-modified-id=\"createCrisprReady.sh-1.5\"><span class=\"toc-item-num\">1.5&nbsp;&nbsp;</span>createCrisprReady.sh</a></span></li><li><span><a href=\"#identifyCrisprSite.sh\" data-toc-modified-id=\"identifyCrisprSite.sh-1.6\"><span class=\"toc-item-num\">1.6&nbsp;&nbsp;</span>identifyCrisprSite.sh</a></span></li><li><span><a href=\"#editGenome.sh\" data-toc-modified-id=\"editGenome.sh-1.7\"><span class=\"toc-item-num\">1.7&nbsp;&nbsp;</span>editGenome.sh</a></span></li><li><span><a href=\"#exomeReport.py\" data-toc-modified-id=\"exomeReport.py-1.8\"><span class=\"toc-item-num\">1.8&nbsp;&nbsp;</span>exomeReport.py</a></span></li></ul></li><li><span><a href=\"#Readme.txt\" data-toc-modified-id=\"Readme.txt-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Readme.txt</a></span></li><li><span><a href=\"#copyExomes.sh\" data-toc-modified-id=\"copyExomes.sh-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>copyExomes.sh</a></span><ul class=\"toc-item\"><li><span><a href=\"#Structure-Decision\" data-toc-modified-id=\"Structure-Decision-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>Structure Decision</a></span></li><li><span><a href=\"#Programing-Decision\" data-toc-modified-id=\"Programing-Decision-3.2\"><span class=\"toc-item-num\">3.2&nbsp;&nbsp;</span>Programing Decision</a></span></li><li><span><a href=\"#Read-in-clinical_data-with-pandas\" data-toc-modified-id=\"Read-in-clinical_data-with-pandas-3.3\"><span class=\"toc-item-num\">3.3&nbsp;&nbsp;</span>Read in clinical_data with pandas</a></span></li><li><span><a href=\"#Filter-dataframe-by-diameter-size\" data-toc-modified-id=\"Filter-dataframe-by-diameter-size-3.4\"><span class=\"toc-item-num\">3.4&nbsp;&nbsp;</span>Filter dataframe by diameter size</a></span></li><li><span><a href=\"#Make-exomes-directory-and-copy-over-filtered-files\" data-toc-modified-id=\"Make-exomes-directory-and-copy-over-filtered-files-3.5\"><span class=\"toc-item-num\">3.5&nbsp;&nbsp;</span>Make exomes directory and copy over filtered files</a></span></li><li><span><a href=\"#Create-motif_manager.py\" data-toc-modified-id=\"Create-motif_manager.py-3.6\"><span class=\"toc-item-num\">3.6&nbsp;&nbsp;</span>Create motif_manager.py</a></span></li><li><span><a href=\"#Task-1.4-Complete\" data-toc-modified-id=\"Task-1.4-Complete-3.7\"><span class=\"toc-item-num\">3.7&nbsp;&nbsp;</span>Task 1.4 Complete</a></span></li></ul></li><li><span><a href=\"#createCrisprReady.sh\" data-toc-modified-id=\"createCrisprReady.sh-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>createCrisprReady.sh</a></span><ul class=\"toc-item\"><li><span><a href=\"#Check-size-length-of-.fasta-files\" data-toc-modified-id=\"Check-size-length-of-.fasta-files-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>Check size length of .fasta files</a></span></li><li><span><a href=\"#Build-Animal-Dataframe\" data-toc-modified-id=\"Build-Animal-Dataframe-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Build Animal Dataframe</a></span></li><li><span><a href=\"#Build-dataframe-counter-for-each-animal\" data-toc-modified-id=\"Build-dataframe-counter-for-each-animal-4.3\"><span class=\"toc-item-num\">4.3&nbsp;&nbsp;</span>Build dataframe counter for each animal</a></span></li><li><span><a href=\"#Top-Ten-Motifs\" data-toc-modified-id=\"Top-Ten-Motifs-4.4\"><span class=\"toc-item-num\">4.4&nbsp;&nbsp;</span>Top Ten Motifs</a></span></li><li><span><a href=\"#Top-motifs-for-each-animal\" data-toc-modified-id=\"Top-motifs-for-each-animal-4.5\"><span class=\"toc-item-num\">4.5&nbsp;&nbsp;</span>Top motifs for each animal</a></span></li><li><span><a href=\"#Count-Report\" data-toc-modified-id=\"Count-Report-4.6\"><span class=\"toc-item-num\">4.6&nbsp;&nbsp;</span>Count Report</a></span></li><li><span><a href=\"#Create-motif_manager.py\" data-toc-modified-id=\"Create-motif_manager.py-4.7\"><span class=\"toc-item-num\">4.7&nbsp;&nbsp;</span>Create motif_manager.py</a></span></li></ul></li><li><span><a href=\"#identifyCrisprSite.sh\" data-toc-modified-id=\"identifyCrisprSite.sh-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>identifyCrisprSite.sh</a></span></li><li><span><a href=\"#editGenome.sh\" data-toc-modified-id=\"editGenome.sh-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>editGenome.sh</a></span></li><li><span><a href=\"#exomeReport.py\" data-toc-modified-id=\"exomeReport.py-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>exomeReport.py</a></span></li><li><span><a href=\"#Levenshtein-Distance-Measure\" data-toc-modified-id=\"Levenshtein-Distance-Measure-8\"><span class=\"toc-item-num\">8&nbsp;&nbsp;</span>Levenshtein Distance Measure</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Description"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<strong>Due Tuesday Week 4 </strong><br>\n",
    "Beginning Python scripting <br>\n",
    "Key concepts covered: Apply Bash and Python practices to solving problems"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### readme.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<strong>Important! Your assignment must contain a readme.txt file that indicates the exact command used to execute your script on the AWS server. It should indicate what inputs are used, and indicate what outputs are produced.\n",
    "Use relative paths so that if I copied your week4 folder to any other directory, it will still execute properly.</strong>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Story"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dr. Banner leads a team of researchers that have each discovered a unique eukaryotic organism. They all seem to be related to each other and the team has tried to sequence their specimens. Homologous genes between each eukaryote were labeled with the same gene name. Each of organisms have some degree of gamma ray resistance and Dr. Banner constructed a list of motifs that may be associated with resistance. He wants to use only the most abundant motif and knockout the genes that contain these motifs using CRISPR.  He first needs to identify a list of suitable CRISPR gene candidates and submit a report. \n",
    "\n",
    "Inputs:\n",
    "\n",
    "/home/rbif/week4/clinical_data.txt <br>\n",
    "/home/rbif/week4/motif_list.txt <br>\n",
    "/home/rbif/week4/exomes/*.fasta <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File Structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "notes: Task completed. Tim, I am not sure if you want me to make a bash script that does this with something like:\n",
    "```\n",
    "mkdir week4\n",
    "rsync . /home/ruggm/BioInformatics/001_RBIF-100-1DL/week4\n",
    "```\n",
    "or just make it here. I am going to assume that I have done this successfully, but let me know if you want me to make a little build_me script."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<strong>Task 1.3 completed</strong>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### copyExomes.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in the clinical data file and identify the samples that have a diameter between 20 and 30 mm long (inclusive) and have had their genomes sequenced. Copy the identified exomes using the sample code names to a new directory called exomes.<br><br><strong>Task 1.4 completed</strong>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### createCrisprReady.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the motif_list.txt file, identify the 3 highest occurring motifs in each exome inside the exomes folder. \n",
    "Output the headers and corresponding sequences to a new file called exomename_topmotifs.fasta.  **This means to only select for the genes that have at least one of the three top motifs and output to a new file\n",
    "Rename the headers of the exomename_topmotifs.fasta files to include the motif sequences and their count. (Example: >gene1  ===> >gene1 ATGC_1 AAAA_38)\n",
    "\n",
    "Note: The sequence does not need to contain all 3 motifs, it just needs to have at least 1 of them.\n",
    "“exomename_topmotifs.fasta” should contain the code name of the sample. For example, fox_topmotifs.fasta."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### identifyCrisprSite.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each gene inside the exomename_topmotifs.fasta files, identify a suitable CRISPR site.\n",
    "Find sequences that contain “NGG”, where “N” can be any base, that has at least 20 basepairs upstream.\n",
    "Example of upstream: ATGAACGTCTGTAAGAACTGCGGATCTGTCA (Everything left of CGG is upstream of the DNA) \n",
    "Output suitable candidates (headers and sequences) to a new file called exomenames_precrispr.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### editGenome.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using those files, write a script that will insert the letter A right before the NGG site. Output to a new file called exomename_postcrispr.fasta \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### exomeReport.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a single report that summarizes the findings. It should be a text file that lists the name of the discoverer of the organism, the diameter, the code name and the environment it came from. -Additionally, include how many genes the organism has in common with all of the other organisms, and which genes are unique to that organism.\n",
    "Note: You are only comparing that organism against the other organisms of the same cohort (20-30mm in diameter). ie You are summarizing the motif-containing exomes with just the CRISPR-ready genes, in otherwords using the exomenames_precrispr.fasta (or exomenames_postcrispr.fasta since the headers should be identical between those two files)\n",
    "There may not be any genes in common in that entire cohort.\n",
    "\n",
    "Example:\n",
    "Organism FOX, discovered by DISCOVERER, has a diameter of DIAMETER, and has the NUMBER of genes in common with the cohort, and the following genes unique to only itself: gene1, gene4, gene823"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Readme.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile Readme.txt\n",
    "\n",
    "Welcome to week4 homework! To run the script, step by step type: \n",
    "\n",
    "    ./copyExomes\n",
    "    ./createCrisprReady.sh\n",
    "    ./identifyCrisprSite.sh\n",
    "    ./editGenome.sh\n",
    "\n",
    "into terminal or type ./runAll.sh. If you use the runAll.sh pipe, it will\n",
    "delete old reports. \n",
    "\n",
    "About the report...I am not sure if you mean shared genes or motifs. It \n",
    "seems all genes are unique, but I will look at it tomorrow morning."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## copyExomes.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Structure Decision"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a little confusing to me Tim, we already have a file \n",
    "exomes. Do you want me to over write this folder and delete the \n",
    "Older .fasta files? \n",
    "\n",
    "Seems weird to overwrite everything like that and destroy\n",
    "The old .fasta files. So I am going to move the exomes folder to a \n",
    "exomes_given and make a new directory as asked."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Programing Decision"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It seems like I have two choices to make all of this work. I can make a bunch of little scripts that play together in a modular fashion or I can maybe a larger python script that is called in different bash scripts. For such a little project, it seems strained to make a bunch of little scripts to fry all together. Therefore, I am going to build one larger python script that will do most of the heavy lifting and call all of the parts of that script from different parts of the bash shell. <br><br>\n",
    "It looks like you are giving me a lot of latitude here, so I am going to use pandas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in clinical_data with pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def copy_exomes():\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    return df\n",
    "\n",
    "df = copy_exomes()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This looks rather clean, moving to filer by the requested diameters."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter dataframe by diameter size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "mover.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make exomes directory and copy over filtered files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(\"exomes\"):\n",
    "    os.mkdir(\"exomes\")\n",
    "else:\n",
    "    shutil.rmtree(\"exomes\")\n",
    "    os.mkdir(\"exomes\")\n",
    "    \n",
    "for i in mover[\"code_name\"].tolist():\n",
    "    print(\"copying {}.fasta file\".format(i))\n",
    "    shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                 \"exomes/{}.fasta\".format(i))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create motif_manager.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "\n",
    "def copy_exomes():\n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    print(\"The following entries are of interest:\")\n",
    "    print(mover)\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        print(\"copying {}.fasta file\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile copyExomes.sh\n",
    "#!/bin/bash\n",
    "\n",
    "python -c 'import motif_manager; motif_manager.copy_exomes()'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 1.4 Complete"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## createCrisprReady.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check size length of .fasta files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for fasta in os.listdir(\"exomes\"):\n",
    "    with open('exomes/{}'.format(fasta)) as f:\n",
    "        lines = [line.rstrip() for line in f]\n",
    "        print(len(lines))\n",
    "        \n",
    "with open(\"motif_list.txt\") as f:\n",
    "    motifs = [line.rstrip() for line in f]\n",
    "    \n",
    "print(motifs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is very helpful that all of these lists have the same size. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build Animal Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "animal_house = []\n",
    "animal_list = {}\n",
    "for fasta in os.listdir(\"exomes\"):\n",
    "    with open('exomes/{}'.format(fasta)) as f:\n",
    "        lines = [line.rstrip() for line in f]\n",
    "    animal_list[fasta] = lines\n",
    "    entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "    entry[\"name\"] = fasta[:-6]\n",
    "    animal_house.append(entry)\n",
    "\n",
    "        \n",
    "with open(\"motif_list.txt\") as f:\n",
    "    motifs = [line.rstrip() for line in f]\n",
    "        \n",
    "animal_house = pd.DataFrame(animal_house)\n",
    "animal_house.set_index('name', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "animal_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It looks like we have two different lines of attack here. First, I could make a tensor of all of the different motifs and then collapse the tensor by count. Also, I could crush down genes inside each animal and then explore the values. The tensor option seems the most fun, but I think we will stick to a plain Jane adding a column count inside of the dataframe."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build dataframe counter for each animal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(animal_house.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "animal_data = {}\n",
    "count_list = []\n",
    "top_per_animal = {}\n",
    "gene_dict = {}\n",
    "for animal in list(animal_house.index):\n",
    "    animal_data[animal] = animal_house.loc[[animal]].T\n",
    "    for motif in motifs:\n",
    "        animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "                \n",
    "    # Suming up columns and rows\n",
    "    animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "    \n",
    "    # There is a quirk in Pandas where the Text column is added, will drop\n",
    "    animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "    \n",
    "    # Casting to ints\n",
    "    animal_data[animal] = animal_data[animal].astype(int)\n",
    "    \n",
    "    \n",
    "    count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T)\n",
    "    top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "    \n",
    "    df = animal_data[animal][top_per_animal[animal]]\n",
    "    df = df[(df.T != 0).any()]\n",
    "    \n",
    "    gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "count_values = pd.concat(count_list, axis=1, sort = True)\n",
    "\n",
    "#Total sum per row: \n",
    "count_values.loc[:,'sum'] = count_values.sum(axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Top Ten Motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_values[\"sum\"].nlargest(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_values.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "count_values[\"sum\"].nlargest(10).plot.bar(x=\"name\");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Top motifs for each animal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_per_animal[\"lamb\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for animal in top_per_animal:\n",
    "    print(animal)\n",
    "    print(top_per_animal[animal])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we can see that the same three motifs do in fact dominate all\n",
    "of the various animals."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count Report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = animal_data[animal][top_per_animal[animal]]\n",
    "df = df[(df.T != 0).any()]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I do not really understand the reason why anyone would want this report type, versus something like a json or dataframe, but here you go. <br><br>\n",
    "I am taking the statement <b>The sequence does not need to contain all 3 motifs</b> to mean that I <b>can</b> include them. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I am running out of time, so I am going to build this all at once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "report.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Send the report dictionary to .txt files, not sure where you want them\n",
    "# Sending to different directory \"exomes_top\"\n",
    "\n",
    "# Make Directory Structure\n",
    "if not os.path.exists(\"exomes_top\"):\n",
    "    os.mkdir(\"exomes_top\")\n",
    "else:\n",
    "    shutil.rmtree(\"exomes_top\")\n",
    "    os.mkdir(\"exomes_top\")\n",
    "\n",
    "for file in report:\n",
    "    with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "        for line in report[file]:\n",
    "            crispy.write(\"{}\\n\".format(line))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create motif_manager.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "def copy_exomes():\n",
    " \n",
    "    now = datetime.datetime.now()\n",
    "    \n",
    "    # Making Report as we go\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nAnalysis Run Started:\\n\" + now.strftime(\"%c\") + \"\\n----------\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    \n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"The following entries are of interest:\\n\")\n",
    "        f.write(\"{}\".format(mover) + \"\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"copying {}.fasta file\\n\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))\n",
    "        \n",
    "def create_crispy():\n",
    "    animal_apartment = []\n",
    "    animal_list = {}\n",
    "    gene_dict = {}\n",
    "\n",
    "    #First build dataframe animals in exome dir\n",
    "    for fasta in os.listdir(\"exomes\"):\n",
    "        with open('exomes/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        animal_list[fasta[:-6]] = lines\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "        entry[\"name\"] = fasta[:-6]\n",
    "        animal_apartment.append(entry)\n",
    "\n",
    "    animal_house = pd.DataFrame(animal_apartment)\n",
    "    animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "    #Iterate through motif_list\n",
    "    with open(\"motif_list.txt\") as f:\n",
    "        motifs = [line.rstrip() for line in f]\n",
    "\n",
    "    # Build counting dataframe for each animal\n",
    "    animal_data = {}\n",
    "    count_list = []\n",
    "    top_per_animal = {}\n",
    "\n",
    "    for animal in list(animal_house.index):\n",
    "        animal_data[animal] = animal_house.loc[[animal]].T\n",
    "        for motif in motifs:\n",
    "            animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "        # Suming up columns and rows\n",
    "        animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "        # There is a quirk in Pandas where the Text column is added, will drop\n",
    "        animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "        # Casting to ints\n",
    "        animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "        # Add the sum into a count_column list\n",
    "        count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "        # Populate top 3 lists\n",
    "        top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "        # Look at only the populated data lists\n",
    "        df = animal_data[animal][top_per_animal[animal]]\n",
    "        df = df[(df.T != 0).any()]\n",
    "\n",
    "        # Make animal dict to convert back to text file\n",
    "        gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "    # Join all of the columns into one dataframe\n",
    "    count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "    #Total sum per row: \n",
    "    count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "    report_dict = {}\n",
    "\n",
    "    # Build a report dictionary \n",
    "    for animal in gene_dict:\n",
    "        report_dict[animal] = {}\n",
    "        for gene in gene_dict[animal]:\n",
    "            d = gene_dict[animal][gene]\n",
    "\n",
    "            #Flatten Dictionary\n",
    "            report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "    # Build the final report with proper formating\n",
    "    report = {}\n",
    "    for animal in animal_list:\n",
    "        text_list = []\n",
    "        for line in animal_list[animal]:\n",
    "            if line in report_dict[animal]:\n",
    "                text_list.append(line + \" \" + report_dict[animal][line])\n",
    "                text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "        report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "    # Send the report dictionary to .txt files, not sure where you want them\n",
    "    # Sending to different directory \"exomes_top\"\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_top\"):\n",
    "        os.mkdir(\"exomes_top\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_top\")\n",
    "        os.mkdir(\"exomes_top\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in report:\n",
    "        with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "            for line in report[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nTop Ten Motifs\\n\")\n",
    "        f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "        f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "        for animal in top_per_animal:\n",
    "            f.write(\"\\n{}\\n\".format(animal))\n",
    "            f.write(\"{}\".format(top_per_animal[animal]))\n",
    "        f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile createCrisprReady.sh\n",
    "#!/bin/bash\n",
    "\n",
    "python -c 'import motif_manager; motif_manager.create_crispy()'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## identifyCrisprSite.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I fell way behind on the time, so to catch up I am just going to script this right out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "def copy_exomes():\n",
    " \n",
    "    now = datetime.datetime.now()\n",
    "    \n",
    "    # Making Report as we go\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nAnalysis Run Started:\\n\" + now.strftime(\"%c\") + \"\\n----------\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    \n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"The following entries are of interest:\\n\")\n",
    "        f.write(\"{}\".format(mover) + \"\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"copying {}.fasta file\\n\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))\n",
    "        \n",
    "def create_crispy():\n",
    "    animal_apartment = []\n",
    "    animal_list = {}\n",
    "    gene_dict = {}\n",
    "\n",
    "    #First build dataframe animals in exome dir\n",
    "    for fasta in os.listdir(\"exomes\"):\n",
    "        with open('exomes/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        animal_list[fasta[:-6]] = lines\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "        entry[\"name\"] = fasta[:-6]\n",
    "        animal_apartment.append(entry)\n",
    "\n",
    "    animal_house = pd.DataFrame(animal_apartment)\n",
    "    animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "    #Iterate through motif_list\n",
    "    with open(\"motif_list.txt\") as f:\n",
    "        motifs = [line.rstrip() for line in f]\n",
    "\n",
    "    # Build counting dataframe for each animal\n",
    "    animal_data = {}\n",
    "    count_list = []\n",
    "    top_per_animal = {}\n",
    "\n",
    "    for animal in list(animal_house.index):\n",
    "        animal_data[animal] = animal_house.loc[[animal]].T\n",
    "        for motif in motifs:\n",
    "            animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "        # Suming up columns and rows\n",
    "        animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "        # There is a quirk in Pandas where the Text column is added, will drop\n",
    "        animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "        # Casting to ints\n",
    "        animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "        # Add the sum into a count_column list\n",
    "        count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "        # Populate top 3 lists\n",
    "        top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "        # Look at only the populated data lists\n",
    "        df = animal_data[animal][top_per_animal[animal]]\n",
    "        df = df[(df.T != 0).any()]\n",
    "\n",
    "        # Make animal dict to convert back to text file\n",
    "        gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "    # Join all of the columns into one dataframe\n",
    "    count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "    #Total sum per row: \n",
    "    count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "    report_dict = {}\n",
    "\n",
    "    # Build a report dictionary \n",
    "    for animal in gene_dict:\n",
    "        report_dict[animal] = {}\n",
    "        for gene in gene_dict[animal]:\n",
    "            d = gene_dict[animal][gene]\n",
    "\n",
    "            #Flatten Dictionary\n",
    "            report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "    # Build the final report with proper formating\n",
    "    report = {}\n",
    "    for animal in animal_list:\n",
    "        text_list = []\n",
    "        for line in animal_list[animal]:\n",
    "            if line in report_dict[animal]:\n",
    "                text_list.append(line + \" \" + report_dict[animal][line])\n",
    "                text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "        report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "    # Send the report dictionary to .txt files, not sure where you want them\n",
    "    # Sending to different directory \"exomes_top\"\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_top\"):\n",
    "        os.mkdir(\"exomes_top\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_top\")\n",
    "        os.mkdir(\"exomes_top\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in report:\n",
    "        with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "            for line in report[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nTop Ten Motifs\\n\")\n",
    "        f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "        f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "        for animal in top_per_animal:\n",
    "            f.write(\"\\n{}\\n\".format(animal))\n",
    "            f.write(\"{}\".format(top_per_animal[animal]))\n",
    "        f.write(\"\\n\")\n",
    "        f.close()\n",
    "        \n",
    "def locate():\n",
    "    \n",
    "    top_house = {}\n",
    "    soggy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_top\"):\n",
    "        with open('exomes_top/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        top_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in top_house:\n",
    "        soggy_list = []\n",
    "        have_gg = 0\n",
    "        no_gg = 0\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"\\nSearching {} for NGG\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in top_house[animal]:\n",
    "            candidate = top_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                soggy_list.append(gene)\n",
    "                soggy_list.append(candidate)\n",
    "                have_gg += 1\n",
    "\n",
    "                # Reporter\n",
    "                with open(\"report.txt\", \"a+\") as f:\n",
    "                    f.write(\"Found NGG at {}\\n\".format(gene.split(\" \")[0]))\n",
    "                    f.close()\n",
    "\n",
    "            else:\n",
    "                no_gg += 1\n",
    "\n",
    "            # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"On {}, found {} NGG genes out of {} candidates\\n\".format(animal,\n",
    "                                                                              have_gg,\n",
    "                                                                              have_gg + no_gg))\n",
    "            f.write(\"-----------------\\n\")\n",
    "            f.close()\n",
    "        soggy_dict[animal] = soggy_list\n",
    "\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_precrispr\"):\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_precrispr\")\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in soggy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_precrispr.fasta\"\n",
    "        with open(\"exomes_precrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in soggy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile identifyCrisprSite.sh\n",
    "#!/bin/bash\n",
    "\n",
    "python -c 'import motif_manager; motif_manager.locate()'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## editGenome.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Two quick assumptions, are we adding an A to multiple sites? I am going to assume yes. Next, are we inserting before the GG (NAGG) or before NGG (ANGG). I am assuming NAGG. <br><br>\n",
    "\n",
    "Probably could have done this all at once, but no big deal to do it in two steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "import Levenshtein as lev\n",
    "\n",
    "def copy_exomes():\n",
    " \n",
    "    now = datetime.datetime.now()\n",
    "    \n",
    "    # Making Report as we go\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nAnalysis Run Started:\\n\" + now.strftime(\"%c\") + \"\\n----------\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    \n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"The following entries are of interest:\\n\")\n",
    "        f.write(\"{}\".format(mover) + \"\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"copying {}.fasta file\\n\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))\n",
    "        \n",
    "def create_crispy():\n",
    "    animal_apartment = []\n",
    "    animal_list = {}\n",
    "    gene_dict = {}\n",
    "\n",
    "    #First build dataframe animals in exome dir\n",
    "    for fasta in os.listdir(\"exomes\"):\n",
    "        with open('exomes/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        animal_list[fasta[:-6]] = lines\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "        entry[\"name\"] = fasta[:-6]\n",
    "        animal_apartment.append(entry)\n",
    "\n",
    "    animal_house = pd.DataFrame(animal_apartment)\n",
    "    animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "    #Iterate through motif_list\n",
    "    with open(\"motif_list.txt\") as f:\n",
    "        motifs = [line.rstrip() for line in f]\n",
    "\n",
    "    # Build counting dataframe for each animal\n",
    "    animal_data = {}\n",
    "    count_list = []\n",
    "    top_per_animal = {}\n",
    "\n",
    "    for animal in list(animal_house.index):\n",
    "        animal_data[animal] = animal_house.loc[[animal]].T\n",
    "        for motif in motifs:\n",
    "            animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "        # Suming up columns and rows\n",
    "        animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "        # There is a quirk in Pandas where the Text column is added, will drop\n",
    "        animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "        # Casting to ints\n",
    "        animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "        # Add the sum into a count_column list\n",
    "        count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "        # Populate top 3 lists\n",
    "        top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "        # Look at only the populated data lists\n",
    "        df = animal_data[animal][top_per_animal[animal]]\n",
    "        df = df[(df.T != 0).any()]\n",
    "\n",
    "        # Make animal dict to convert back to text file\n",
    "        gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "    # Join all of the columns into one dataframe\n",
    "    count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "    #Total sum per row: \n",
    "    count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "    report_dict = {}\n",
    "\n",
    "    # Build a report dictionary \n",
    "    for animal in gene_dict:\n",
    "        report_dict[animal] = {}\n",
    "        for gene in gene_dict[animal]:\n",
    "            d = gene_dict[animal][gene]\n",
    "\n",
    "            #Flatten Dictionary\n",
    "            report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "    # Build the final report with proper formating\n",
    "    report = {}\n",
    "    for animal in animal_list:\n",
    "        text_list = []\n",
    "        for line in animal_list[animal]:\n",
    "            if line in report_dict[animal]:\n",
    "                text_list.append(line + \" \" + report_dict[animal][line])\n",
    "                text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "        report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "    # Send the report dictionary to .txt files, not sure where you want them\n",
    "    # Sending to different directory \"exomes_top\"\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_top\"):\n",
    "        os.mkdir(\"exomes_top\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_top\")\n",
    "        os.mkdir(\"exomes_top\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in report:\n",
    "        with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "            for line in report[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "    similar = animal_house.T.to_dict()\n",
    "    count_dict = {}\n",
    "    gene_list = []\n",
    "\n",
    "    # First make a list of all genes\n",
    "    for animal in similar:\n",
    "        for gene in similar[animal]:\n",
    "            gene_list.append(similar[animal][gene])\n",
    "\n",
    "    # Search for duplicates\n",
    "    for animal in similar:\n",
    "        count_dict[animal] = {}\n",
    "        for gene in similar[animal]:\n",
    "            count = gene_list.count(similar[animal][gene]) - 1\n",
    "            count_dict[animal][gene] = count\n",
    "            if count > 0: print(\"shared gene!\")\n",
    "\n",
    "    shared_genes = pd.DataFrame(count_dict)\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nShared Genes\\n\")\n",
    "        f.write(\"{}\".format(shared_genes[shared_genes != 0].dropna()))\n",
    "        f.write(\"\\n\\nTop Ten Motifs\\n\")\n",
    "        f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "        f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "        for animal in top_per_animal:\n",
    "            f.write(\"\\n{}\\n\".format(animal))\n",
    "            f.write(\"{}\".format(top_per_animal[animal]))\n",
    "        f.write(\"\\n\")\n",
    "        f.close()\n",
    "        \n",
    "def locate():\n",
    "    \n",
    "    top_house = {}\n",
    "    soggy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_top\"):\n",
    "        with open('exomes_top/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        top_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in top_house:\n",
    "        soggy_list = []\n",
    "        have_gg = 0\n",
    "        no_gg = 0\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"\\nSearching {} for NGG\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in top_house[animal]:\n",
    "            candidate = top_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                soggy_list.append(gene)\n",
    "                soggy_list.append(candidate)\n",
    "                have_gg += 1\n",
    "\n",
    "                # Reporter\n",
    "                with open(\"report.txt\", \"a+\") as f:\n",
    "                    f.write(\"Found NGG at {}\\n\".format(gene.split(\" \")[0]))\n",
    "                    f.close()\n",
    "\n",
    "            else:\n",
    "                no_gg += 1\n",
    "\n",
    "            # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"On {}, found {} NGG genes out of {} candidates\\n\".format(animal,\n",
    "                                                                              have_gg,\n",
    "                                                                              have_gg + no_gg))\n",
    "            f.write(\"-----------------\\n\")\n",
    "            f.close()\n",
    "        soggy_dict[animal] = soggy_list\n",
    "\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_precrispr\"):\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_precrispr\")\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in soggy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_precrispr.fasta\"\n",
    "        with open(\"exomes_precrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in soggy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "                \n",
    "def edit_genome():\n",
    "    big_house = {}\n",
    "    crispy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_precrispr\"):\n",
    "        with open('exomes_precrispr/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        big_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in big_house:\n",
    "        crispy_list = []\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"Inserting A at NGG sites on file {}\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in big_house[animal]:\n",
    "            candidate = big_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "            for ind in indices:\n",
    "                candidate = candidate[:ind] + \"A\" + candidate[ind:]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                crispy_list.append(gene)\n",
    "                crispy_list.append(candidate)\n",
    "\n",
    "        crispy_dict[animal] = crispy_list\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_postcrispr\"):\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_postcrispr\")\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in crispy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_postcrispr.fasta\"\n",
    "        with open(\"exomes_postcrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in crispy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile editGenome.sh\n",
    "#!/bin/bash\n",
    "\n",
    "python -c 'import motif_manager; motif_manager.edit_genome()'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile runAll.sh\n",
    "#!/bin/bash\n",
    "\n",
    "echo \"removing old reports\"\n",
    "rm report.txt\n",
    "./copyExomes.sh\n",
    "./createCrisprReady.sh\n",
    "./identifyCrisprSite.sh\n",
    "./editGenome.sh\n",
    "echo \"completed\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## exomeReport.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "import Levenshtein as lev\n",
    "\n",
    "def copy_exomes():\n",
    " \n",
    "    now = datetime.datetime.now()\n",
    "    \n",
    "    # Making Report as we go\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nAnalysis Run Started:\\n\" + now.strftime(\"%c\") + \"\\n----------\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    \n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"The following entries are of interest:\\n\")\n",
    "        f.write(\"{}\".format(mover) + \"\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"copying {}.fasta file\\n\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))\n",
    "        \n",
    "def create_crispy():\n",
    "    animal_apartment = []\n",
    "    animal_list = {}\n",
    "    gene_dict = {}\n",
    "\n",
    "    #First build dataframe animals in exome dir\n",
    "    for fasta in os.listdir(\"exomes\"):\n",
    "        with open('exomes/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        animal_list[fasta[:-6]] = lines\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "        entry[\"name\"] = fasta[:-6]\n",
    "        animal_apartment.append(entry)\n",
    "\n",
    "    animal_house = pd.DataFrame(animal_apartment)\n",
    "    animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "    #Iterate through motif_list\n",
    "    with open(\"motif_list.txt\") as f:\n",
    "        motifs = [line.rstrip() for line in f]\n",
    "\n",
    "    # Build counting dataframe for each animal\n",
    "    animal_data = {}\n",
    "    count_list = []\n",
    "    top_per_animal = {}\n",
    "\n",
    "    for animal in list(animal_house.index):\n",
    "        animal_data[animal] = animal_house.loc[[animal]].T\n",
    "        for motif in motifs:\n",
    "            animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "        # Suming up columns and rows\n",
    "        animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "        # There is a quirk in Pandas where the Text column is added, will drop\n",
    "        animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "        # Casting to ints\n",
    "        animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "        # Add the sum into a count_column list\n",
    "        count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "        # Populate top 3 lists\n",
    "        top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "        # Look at only the populated data lists\n",
    "        df = animal_data[animal][top_per_animal[animal]]\n",
    "        df = df[(df.T != 0).any()]\n",
    "\n",
    "        # Make animal dict to convert back to text file\n",
    "        gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "    # Join all of the columns into one dataframe\n",
    "    count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "    #Total sum per row: \n",
    "    count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "    report_dict = {}\n",
    "\n",
    "    # Build a report dictionary \n",
    "    for animal in gene_dict:\n",
    "        report_dict[animal] = {}\n",
    "        for gene in gene_dict[animal]:\n",
    "            d = gene_dict[animal][gene]\n",
    "\n",
    "            #Flatten Dictionary\n",
    "            report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "    # Build the final report with proper formating\n",
    "    report = {}\n",
    "    for animal in animal_list:\n",
    "        text_list = []\n",
    "        for line in animal_list[animal]:\n",
    "            if line in report_dict[animal]:\n",
    "                text_list.append(line + \" \" + report_dict[animal][line])\n",
    "                text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "        report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "    # Send the report dictionary to .txt files, not sure where you want them\n",
    "    # Sending to different directory \"exomes_top\"\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_top\"):\n",
    "        os.mkdir(\"exomes_top\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_top\")\n",
    "        os.mkdir(\"exomes_top\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in report:\n",
    "        with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "            for line in report[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "    similar = animal_house.T.to_dict()\n",
    "    count_dict = {}\n",
    "    gene_list = []\n",
    "\n",
    "    # First make a list of all genes\n",
    "    for animal in similar:\n",
    "        for gene in similar[animal]:\n",
    "            gene_list.append(similar[animal][gene])\n",
    "\n",
    "    # Search for duplicates\n",
    "    for animal in similar:\n",
    "        count_dict[animal] = {}\n",
    "        for gene in similar[animal]:\n",
    "            count = gene_list.count(similar[animal][gene]) - 1\n",
    "            count_dict[animal][gene] = count\n",
    "            if count > 0: print(\"shared gene!\")\n",
    "\n",
    "    shared_genes = pd.DataFrame(count_dict)\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nShared Genes\\n\")\n",
    "        f.write(\"{}\".format(shared_genes[shared_genes != 0].dropna()))\n",
    "        f.write(\"\\n\\nTop Ten Motifs\\n\")\n",
    "        f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "        f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "        for animal in top_per_animal:\n",
    "            f.write(\"\\n{}\\n\".format(animal))\n",
    "            f.write(\"{}\".format(top_per_animal[animal]))\n",
    "        f.write(\"\\n\")\n",
    "        f.close()\n",
    "        \n",
    "def locate():\n",
    "    \n",
    "    top_house = {}\n",
    "    soggy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_top\"):\n",
    "        with open('exomes_top/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        top_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in top_house:\n",
    "        soggy_list = []\n",
    "        have_gg = 0\n",
    "        no_gg = 0\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"\\nSearching {} for NGG\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in top_house[animal]:\n",
    "            candidate = top_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                soggy_list.append(gene)\n",
    "                soggy_list.append(candidate)\n",
    "                have_gg += 1\n",
    "\n",
    "                # Reporter\n",
    "                with open(\"report.txt\", \"a+\") as f:\n",
    "                    f.write(\"Found NGG at {}\\n\".format(gene.split(\" \")[0]))\n",
    "                    f.close()\n",
    "\n",
    "            else:\n",
    "                no_gg += 1\n",
    "\n",
    "            # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"On {}, found {} NGG genes out of {} candidates\\n\".format(animal,\n",
    "                                                                              have_gg,\n",
    "                                                                              have_gg + no_gg))\n",
    "            f.write(\"-----------------\\n\")\n",
    "            f.close()\n",
    "        soggy_dict[animal] = soggy_list\n",
    "\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_precrispr\"):\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_precrispr\")\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in soggy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_precrispr.fasta\"\n",
    "        with open(\"exomes_precrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in soggy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "                \n",
    "def edit_genome():\n",
    "    big_house = {}\n",
    "    crispy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_precrispr\"):\n",
    "        with open('exomes_precrispr/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        big_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in big_house:\n",
    "        crispy_list = []\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"Inserting A at NGG sites on file {}\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in big_house[animal]:\n",
    "            candidate = big_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "            for ind in indices:\n",
    "                candidate = candidate[:ind] + \"A\" + candidate[ind:]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                crispy_list.append(gene)\n",
    "                crispy_list.append(candidate)\n",
    "\n",
    "        crispy_dict[animal] = crispy_list\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_postcrispr\"):\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_postcrispr\")\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in crispy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_postcrispr.fasta\"\n",
    "        with open(\"exomes_postcrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in crispy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above is not the best way to measure distance. I will look at Levenshtein distance measure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Levenshtein Distance Measure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile motif_manager.py\n",
    "#!/usr/bin/env python\n",
    "\n",
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "import Levenshtein as lev\n",
    "\n",
    "def copy_exomes():\n",
    " \n",
    "    now = datetime.datetime.now()\n",
    "    \n",
    "    # Making Report as we go\n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nAnalysis Run Started:\\n\" + now.strftime(\"%c\") + \"\\n----------\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    \n",
    "    # Read in tab deliniated file\n",
    "    df = pd.read_csv(\"clinical_data.txt\", sep='\\t', engine='python')\n",
    "    \n",
    "    # Filter out by needed terms\n",
    "    mover = df[(20 <= df['Diamater (mm)']) & (df['Diamater (mm)'] <= 30)]\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"The following entries are of interest:\\n\")\n",
    "        f.write(\"{}\".format(mover) + \"\\n\")\n",
    "        f.close()\n",
    "    \n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "    \n",
    "    # Check to see if path exists, it it does, remove and create\n",
    "    if not os.path.exists(\"exomes\"):\n",
    "        os.mkdir(\"exomes\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes\")\n",
    "        os.mkdir(\"exomes\")\n",
    "\n",
    "    # Copy files\n",
    "    for i in mover[\"code_name\"].tolist():\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"copying {}.fasta file\\n\".format(i))\n",
    "        shutil.copy2(\"exomes_given/{}.fasta\".format(i), \n",
    "                     \"exomes/{}.fasta\".format(i))\n",
    "        \n",
    "def create_crispy():\n",
    "    animal_apartment = []\n",
    "    animal_list = {}\n",
    "    gene_dict = {}\n",
    "\n",
    "    #First build dataframe animals in exome dir\n",
    "    for fasta in os.listdir(\"exomes\"):\n",
    "        with open('exomes/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        animal_list[fasta[:-6]] = lines\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "        entry[\"name\"] = fasta[:-6]\n",
    "        animal_apartment.append(entry)\n",
    "\n",
    "    animal_house = pd.DataFrame(animal_apartment)\n",
    "    animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "    #Iterate through motif_list\n",
    "    with open(\"motif_list.txt\") as f:\n",
    "        motifs = [line.rstrip() for line in f]\n",
    "\n",
    "    # Build counting dataframe for each animal\n",
    "    animal_data = {}\n",
    "    count_list = []\n",
    "    top_per_animal = {}\n",
    "\n",
    "    for animal in list(animal_house.index):\n",
    "        animal_data[animal] = animal_house.loc[[animal]].T\n",
    "        for motif in motifs:\n",
    "            animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "        # Suming up columns and rows\n",
    "        animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "        # There is a quirk in Pandas where the Text column is added, will drop\n",
    "        animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "        # Casting to ints\n",
    "        animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "        # Add the sum into a count_column list\n",
    "        count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "        # Populate top 3 lists\n",
    "        top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "        # Look at only the populated data lists\n",
    "        df = animal_data[animal][top_per_animal[animal]]\n",
    "        df = df[(df.T != 0).any()]\n",
    "\n",
    "        # Make animal dict to convert back to text file\n",
    "        gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "    # Join all of the columns into one dataframe\n",
    "    count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "    #Total sum per row: \n",
    "    count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "    report_dict = {}\n",
    "\n",
    "    # Build a report dictionary \n",
    "    for animal in gene_dict:\n",
    "        report_dict[animal] = {}\n",
    "        for gene in gene_dict[animal]:\n",
    "            d = gene_dict[animal][gene]\n",
    "\n",
    "            #Flatten Dictionary\n",
    "            report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "    # Build the final report with proper formating\n",
    "    report = {}\n",
    "    for animal in animal_list:\n",
    "        text_list = []\n",
    "        for line in animal_list[animal]:\n",
    "            if line in report_dict[animal]:\n",
    "                text_list.append(line + \" \" + report_dict[animal][line])\n",
    "                text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "        report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "    # Send the report dictionary to .txt files, not sure where you want them\n",
    "    # Sending to different directory \"exomes_top\"\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_top\"):\n",
    "        os.mkdir(\"exomes_top\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_top\")\n",
    "        os.mkdir(\"exomes_top\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in report:\n",
    "        with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "            for line in report[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "    similar = animal_house.T.to_dict()\n",
    "    count_dict = {}\n",
    "    gene_list = []\n",
    "\n",
    "    # First make a list of all genes\n",
    "    for animal in similar:\n",
    "        for gene in similar[animal]:\n",
    "            gene_list.append(similar[animal][gene])\n",
    "\n",
    "    # Search for duplicates\n",
    "    for animal in similar:\n",
    "        count_dict[animal] = {}\n",
    "        for gene in similar[animal]:\n",
    "            count = gene_list.count(similar[animal][gene]) - 1\n",
    "            count_dict[animal][gene] = count\n",
    "            if count > 0: print(\"shared gene!\")\n",
    "\n",
    "    shared_genes = pd.DataFrame(count_dict)\n",
    "    \n",
    "    with open(\"report.txt\", \"a+\") as f:\n",
    "        f.write(\"\\nShared Genes\\n\")\n",
    "        f.write(\"{}\".format(shared_genes[shared_genes != 0].dropna()))\n",
    "        f.write(\"\\n\\nTop Ten Motifs\\n\")\n",
    "        f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "        f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "        for animal in top_per_animal:\n",
    "            f.write(\"\\n{}\\n\".format(animal))\n",
    "            f.write(\"{}\".format(top_per_animal[animal]))\n",
    "        f.write(\"\\n\")\n",
    "        f.close()\n",
    "        \n",
    "def locate():\n",
    "    \n",
    "    top_house = {}\n",
    "    soggy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_top\"):\n",
    "        with open('exomes_top/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        top_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in top_house:\n",
    "        soggy_list = []\n",
    "        have_gg = 0\n",
    "        no_gg = 0\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"\\nSearching {} for NGG\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in top_house[animal]:\n",
    "            candidate = top_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                soggy_list.append(gene)\n",
    "                soggy_list.append(candidate)\n",
    "                have_gg += 1\n",
    "\n",
    "                # Reporter\n",
    "                with open(\"report.txt\", \"a+\") as f:\n",
    "                    f.write(\"Found NGG at {}\\n\".format(gene.split(\" \")[0]))\n",
    "                    f.close()\n",
    "\n",
    "            else:\n",
    "                no_gg += 1\n",
    "\n",
    "            # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"On {}, found {} NGG genes out of {} candidates\\n\".format(animal,\n",
    "                                                                              have_gg,\n",
    "                                                                              have_gg + no_gg))\n",
    "            f.write(\"-----------------\\n\")\n",
    "            f.close()\n",
    "        soggy_dict[animal] = soggy_list\n",
    "\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_precrispr\"):\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_precrispr\")\n",
    "        os.mkdir(\"exomes_precrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in soggy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_precrispr.fasta\"\n",
    "        with open(\"exomes_precrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in soggy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n",
    "                \n",
    "def edit_genome():\n",
    "    big_house = {}\n",
    "    crispy_dict = {}\n",
    "\n",
    "    # First build dictionary for candidates\n",
    "    for fasta in os.listdir(\"exomes_precrispr\"):\n",
    "        with open('exomes_precrispr/{}'.format(fasta)) as f:\n",
    "            lines = [line.rstrip() for line in f]\n",
    "\n",
    "        entry = {lines[2*i]:lines[2*i + 1] for i in range(0,len(lines)//2)}\n",
    "        big_house[fasta] = entry\n",
    "\n",
    "    # Search through string for upsteam GG\n",
    "    for animal in big_house:\n",
    "        crispy_list = []\n",
    "\n",
    "        # Reporter\n",
    "        with open(\"report.txt\", \"a+\") as f:\n",
    "            f.write(\"Inserting A at NGG sites on file {}\\n-------------\\n\".format(animal))\n",
    "            f.close()\n",
    "\n",
    "        for gene in big_house[animal]:\n",
    "            candidate = big_house[animal][gene]\n",
    "            search = re.finditer(\"GG\", candidate)\n",
    "            indices = [pos for pos in [m.start(0) for m in search] if pos > 19]\n",
    "            for ind in indices:\n",
    "                candidate = candidate[:ind] + \"A\" + candidate[ind:]\n",
    "\n",
    "            if len(indices) > 0:\n",
    "                crispy_list.append(gene)\n",
    "                crispy_list.append(candidate)\n",
    "\n",
    "        crispy_dict[animal] = crispy_list\n",
    "\n",
    "    # Make Directory Structure\n",
    "    if not os.path.exists(\"exomes_postcrispr\"):\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "    else:\n",
    "        shutil.rmtree(\"exomes_postcrispr\")\n",
    "        os.mkdir(\"exomes_postcrispr\")\n",
    "\n",
    "    # Create Files\n",
    "    for file in crispy_dict:\n",
    "        listing = file.split(\"_\")[0] + \"_postcrispr.fasta\"\n",
    "        with open(\"exomes_postcrispr/{}\".format(listing), \"w\") as crispy:\n",
    "            for line in crispy_dict[file]:\n",
    "                crispy.write(\"{}\\n\".format(line))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import combinations, permutations\n",
    "import datetime\n",
    "import re\n",
    "\n",
    "animal_apartment = []\n",
    "animal_list = {}\n",
    "gene_dict = {}\n",
    "\n",
    "#First build dataframe animals in exome dir\n",
    "for fasta in os.listdir(\"exomes\"):\n",
    "    with open('exomes/{}'.format(fasta)) as f:\n",
    "        lines = [line.rstrip() for line in f]\n",
    "\n",
    "    animal_list[fasta[:-6]] = lines\n",
    "    entry = {lines[2*i]:lines[2*i + 1] for i in range(0,500)}\n",
    "    entry[\"name\"] = fasta[:-6]\n",
    "    animal_apartment.append(entry)\n",
    "\n",
    "animal_house = pd.DataFrame(animal_apartment)\n",
    "animal_house.set_index('name', inplace=True)  \n",
    "\n",
    "#Iterate through motif_list\n",
    "with open(\"motif_list.txt\") as f:\n",
    "    motifs = [line.rstrip() for line in f]\n",
    "\n",
    "# Build counting dataframe for each animal\n",
    "animal_data = {}\n",
    "count_list = []\n",
    "top_per_animal = {}\n",
    "\n",
    "for animal in list(animal_house.index):\n",
    "    animal_data[animal] = animal_house.loc[[animal]].T\n",
    "    for motif in motifs:\n",
    "        animal_data[animal][motif] = animal_data[animal][animal].apply(lambda a: a.count(motif))\n",
    "\n",
    "    # Suming up columns and rows\n",
    "    animal_data[animal].loc['{}'.format(animal),:]= animal_data[animal].sum(axis=0)\n",
    "\n",
    "    # There is a quirk in Pandas where the Text column is added, will drop\n",
    "    animal_data[animal].drop([animal], axis=1, inplace = True)\n",
    "\n",
    "    # Casting to ints\n",
    "    animal_data[animal] = animal_data[animal].astype(int)\n",
    "\n",
    "    # Add the sum into a count_column list\n",
    "    count_list.append(animal_data[animal].loc[['{}'.format(animal)]].T.dropna())\n",
    "\n",
    "    # Populate top 3 lists\n",
    "    top_per_animal[animal] = list(animal_data[animal].loc[animal].nlargest(3).index)\n",
    "\n",
    "    # Look at only the populated data lists\n",
    "    df = animal_data[animal][top_per_animal[animal]]\n",
    "    df = df[(df.T != 0).any()]\n",
    "\n",
    "    # Make animal dict to convert back to text file\n",
    "    gene_dict[animal] = df.T.to_dict()\n",
    "\n",
    "# Join all of the columns into one dataframe\n",
    "count_values = pd.concat(count_list, axis=1)\n",
    "\n",
    "#Total sum per row: \n",
    "count_values.loc[:,'sum'] = count_values.sum(axis=1)\n",
    "\n",
    "report_dict = {}\n",
    "\n",
    "# Build a report dictionary \n",
    "for animal in gene_dict:\n",
    "    report_dict[animal] = {}\n",
    "    for gene in gene_dict[animal]:\n",
    "        d = gene_dict[animal][gene]\n",
    "\n",
    "        #Flatten Dictionary\n",
    "        report_dict[animal][gene] = ' '.join(\"{!s}_{!r}\".format(key,val) for (key,val) in d.items())\n",
    "\n",
    "# Build the final report with proper formating\n",
    "report = {}\n",
    "for animal in animal_list:\n",
    "    text_list = []\n",
    "    for line in animal_list[animal]:\n",
    "        if line in report_dict[animal]:\n",
    "            text_list.append(line + \" \" + report_dict[animal][line])\n",
    "            text_list.append(animal_house.loc[animal][line])\n",
    "\n",
    "    report[\"{}_topmotifs.fasta\".format(animal)] = text_list\n",
    "\n",
    "# Send the report dictionary to .txt files, not sure where you want them\n",
    "# Sending to different directory \"exomes_top\"\n",
    "\n",
    "# Make Directory Structure\n",
    "if not os.path.exists(\"exomes_top\"):\n",
    "    os.mkdir(\"exomes_top\")\n",
    "else:\n",
    "    shutil.rmtree(\"exomes_top\")\n",
    "    os.mkdir(\"exomes_top\")\n",
    "\n",
    "# Create Files\n",
    "for file in report:\n",
    "    with open(\"exomes_top/{}\".format(file), \"w\") as crispy:\n",
    "        for line in report[file]:\n",
    "            crispy.write(\"{}\\n\".format(line))\n",
    "\n",
    "similar = animal_house.T.to_dict()\n",
    "count_dict = {}\n",
    "gene_list = []\n",
    "\n",
    "# First make a list of all genes\n",
    "for animal in similar:\n",
    "    for gene in similar[animal]:\n",
    "        gene_list.append(similar[animal][gene])\n",
    "\n",
    "# Search for duplicates\n",
    "for animal in similar:\n",
    "    count_dict[animal] = {}\n",
    "    for gene in similar[animal]:\n",
    "        count = gene_list.count(similar[animal][gene]) - 1\n",
    "        count_dict[animal][gene] = count\n",
    "        if count > 0: print(\"shared gene!\")\n",
    "\n",
    "shared_genes = pd.DataFrame(count_dict)\n",
    "\n",
    "with open(\"report.txt\", \"a+\") as f:\n",
    "    f.write(\"\\nShared Genes\\n\")\n",
    "    f.write(\"{}\".format(shared_genes[shared_genes != 0].dropna()))\n",
    "    f.write(\"\\n\\nTop Ten Motifs\\n\")\n",
    "    f.write(\"{}\".format(count_values[\"sum\"].nlargest(10)))\n",
    "    f.write('\\n\\nTop Motifs Per Animal\\n')\n",
    "    for animal in top_per_animal:\n",
    "        f.write(\"\\n{}\\n\".format(animal))\n",
    "        f.write(\"{}\".format(top_per_animal[animal]))\n",
    "    f.write(\"\\n\")\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "animal_book = animal_house.T.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Levenshtein as lev\n",
    "import pickle\n",
    "\n",
    "distance_dict = {}\n",
    "for animal in animal_book:\n",
    "    for gene in animal_book[animal]:\n",
    "        distance_dict[animal + \"_\" + gene] = animal_book[animal][gene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'lamb'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(distance_dict.keys())[0].split(\"_\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lamb_>gene1 complete\n",
      "lamb_>gene2 complete\n",
      "lamb_>gene3 complete\n",
      "lamb_>gene4 complete\n",
      "lamb_>gene5 complete\n",
      "lamb_>gene6 complete\n",
      "lamb_>gene7 complete\n",
      "lamb_>gene8 complete\n",
      "lamb_>gene9 complete\n",
      "lamb_>gene10 complete\n",
      "lamb_>gene11 complete\n",
      "lamb_>gene12 complete\n",
      "lamb_>gene13 complete\n",
      "lamb_>gene14 complete\n",
      "lamb_>gene15 complete\n",
      "lamb_>gene16 complete\n",
      "lamb_>gene17 complete\n",
      "lamb_>gene18 complete\n",
      "lamb_>gene19 complete\n",
      "lamb_>gene20 complete\n",
      "lamb_>gene21 complete\n",
      "lamb_>gene22 complete\n",
      "lamb_>gene23 complete\n",
      "lamb_>gene24 complete\n",
      "lamb_>gene25 complete\n",
      "lamb_>gene26 complete\n",
      "lamb_>gene27 complete\n",
      "lamb_>gene28 complete\n",
      "lamb_>gene29 complete\n",
      "lamb_>gene30 complete\n",
      "lamb_>gene31 complete\n",
      "lamb_>gene32 complete\n",
      "lamb_>gene33 complete\n",
      "lamb_>gene34 complete\n",
      "lamb_>gene35 complete\n",
      "lamb_>gene36 complete\n",
      "lamb_>gene37 complete\n",
      "lamb_>gene38 complete\n",
      "lamb_>gene39 complete\n",
      "lamb_>gene40 complete\n",
      "lamb_>gene41 complete\n",
      "lamb_>gene42 complete\n",
      "lamb_>gene43 complete\n",
      "lamb_>gene44 complete\n",
      "lamb_>gene45 complete\n",
      "lamb_>gene46 complete\n",
      "lamb_>gene47 complete\n",
      "lamb_>gene48 complete\n",
      "lamb_>gene49 complete\n",
      "lamb_>gene50 complete\n",
      "lamb_>gene51 complete\n",
      "lamb_>gene52 complete\n",
      "lamb_>gene53 complete\n",
      "lamb_>gene54 complete\n",
      "lamb_>gene55 complete\n",
      "lamb_>gene56 complete\n",
      "lamb_>gene57 complete\n",
      "lamb_>gene58 complete\n",
      "lamb_>gene59 complete\n",
      "lamb_>gene60 complete\n",
      "lamb_>gene61 complete\n",
      "lamb_>gene62 complete\n",
      "lamb_>gene63 complete\n",
      "lamb_>gene64 complete\n",
      "lamb_>gene65 complete\n",
      "lamb_>gene66 complete\n",
      "lamb_>gene67 complete\n",
      "lamb_>gene68 complete\n",
      "lamb_>gene69 complete\n",
      "lamb_>gene70 complete\n",
      "lamb_>gene71 complete\n",
      "lamb_>gene72 complete\n",
      "lamb_>gene73 complete\n",
      "lamb_>gene74 complete\n",
      "lamb_>gene75 complete\n",
      "lamb_>gene76 complete\n",
      "lamb_>gene77 complete\n",
      "lamb_>gene78 complete\n",
      "lamb_>gene79 complete\n",
      "lamb_>gene80 complete\n",
      "lamb_>gene81 complete\n",
      "lamb_>gene82 complete\n",
      "lamb_>gene83 complete\n",
      "lamb_>gene84 complete\n",
      "lamb_>gene85 complete\n",
      "lamb_>gene86 complete\n",
      "lamb_>gene87 complete\n",
      "lamb_>gene88 complete\n",
      "lamb_>gene89 complete\n",
      "lamb_>gene90 complete\n",
      "lamb_>gene91 complete\n",
      "lamb_>gene92 complete\n",
      "lamb_>gene93 complete\n",
      "lamb_>gene94 complete\n",
      "lamb_>gene95 complete\n",
      "lamb_>gene96 complete\n",
      "lamb_>gene97 complete\n",
      "lamb_>gene98 complete\n",
      "lamb_>gene99 complete\n",
      "lamb_>gene100 complete\n",
      "lamb_>gene101 complete\n",
      "lamb_>gene102 complete\n",
      "lamb_>gene103 complete\n",
      "lamb_>gene104 complete\n",
      "lamb_>gene105 complete\n",
      "lamb_>gene106 complete\n",
      "lamb_>gene107 complete\n",
      "lamb_>gene108 complete\n",
      "lamb_>gene109 complete\n",
      "lamb_>gene110 complete\n",
      "lamb_>gene111 complete\n",
      "lamb_>gene112 complete\n",
      "lamb_>gene113 complete\n",
      "lamb_>gene114 complete\n",
      "lamb_>gene115 complete\n",
      "lamb_>gene116 complete\n",
      "lamb_>gene117 complete\n",
      "lamb_>gene118 complete\n",
      "lamb_>gene119 complete\n",
      "lamb_>gene120 complete\n",
      "lamb_>gene121 complete\n",
      "lamb_>gene122 complete\n",
      "lamb_>gene123 complete\n",
      "lamb_>gene124 complete\n",
      "lamb_>gene125 complete\n",
      "lamb_>gene126 complete\n",
      "lamb_>gene127 complete\n",
      "lamb_>gene128 complete\n",
      "lamb_>gene129 complete\n",
      "lamb_>gene130 complete\n",
      "lamb_>gene131 complete\n",
      "lamb_>gene132 complete\n",
      "lamb_>gene133 complete\n",
      "lamb_>gene134 complete\n",
      "lamb_>gene135 complete\n",
      "lamb_>gene136 complete\n",
      "lamb_>gene137 complete\n",
      "lamb_>gene138 complete\n",
      "lamb_>gene139 complete\n",
      "lamb_>gene140 complete\n",
      "lamb_>gene141 complete\n",
      "lamb_>gene142 complete\n",
      "lamb_>gene143 complete\n",
      "lamb_>gene144 complete\n",
      "lamb_>gene145 complete\n",
      "lamb_>gene146 complete\n",
      "lamb_>gene147 complete\n",
      "lamb_>gene148 complete\n",
      "lamb_>gene149 complete\n",
      "lamb_>gene150 complete\n",
      "lamb_>gene151 complete\n",
      "lamb_>gene152 complete\n",
      "lamb_>gene153 complete\n",
      "lamb_>gene154 complete\n",
      "lamb_>gene155 complete\n",
      "lamb_>gene156 complete\n",
      "lamb_>gene157 complete\n",
      "lamb_>gene158 complete\n",
      "lamb_>gene159 complete\n",
      "lamb_>gene160 complete\n",
      "lamb_>gene161 complete\n",
      "lamb_>gene162 complete\n",
      "lamb_>gene163 complete\n",
      "lamb_>gene164 complete\n",
      "lamb_>gene165 complete\n",
      "lamb_>gene166 complete\n",
      "lamb_>gene167 complete\n",
      "lamb_>gene168 complete\n",
      "lamb_>gene169 complete\n",
      "lamb_>gene170 complete\n",
      "lamb_>gene171 complete\n",
      "lamb_>gene172 complete\n",
      "lamb_>gene173 complete\n",
      "lamb_>gene174 complete\n",
      "lamb_>gene175 complete\n",
      "lamb_>gene176 complete\n",
      "lamb_>gene177 complete\n",
      "lamb_>gene178 complete\n",
      "lamb_>gene179 complete\n",
      "lamb_>gene180 complete\n",
      "lamb_>gene181 complete\n",
      "lamb_>gene182 complete\n",
      "lamb_>gene183 complete\n",
      "lamb_>gene184 complete\n",
      "lamb_>gene185 complete\n",
      "lamb_>gene186 complete\n",
      "lamb_>gene187 complete\n",
      "lamb_>gene188 complete\n",
      "lamb_>gene189 complete\n",
      "lamb_>gene190 complete\n",
      "lamb_>gene191 complete\n",
      "lamb_>gene192 complete\n",
      "lamb_>gene193 complete\n",
      "lamb_>gene194 complete\n",
      "lamb_>gene195 complete\n",
      "lamb_>gene196 complete\n",
      "lamb_>gene197 complete\n",
      "lamb_>gene198 complete\n",
      "lamb_>gene199 complete\n",
      "lamb_>gene200 complete\n",
      "lamb_>gene201 complete\n",
      "lamb_>gene202 complete\n",
      "lamb_>gene203 complete\n",
      "lamb_>gene204 complete\n",
      "lamb_>gene205 complete\n",
      "lamb_>gene206 complete\n",
      "lamb_>gene207 complete\n",
      "lamb_>gene208 complete\n",
      "lamb_>gene209 complete\n",
      "lamb_>gene210 complete\n",
      "lamb_>gene211 complete\n",
      "lamb_>gene212 complete\n",
      "lamb_>gene213 complete\n",
      "lamb_>gene214 complete\n",
      "lamb_>gene215 complete\n",
      "lamb_>gene216 complete\n",
      "lamb_>gene217 complete\n",
      "lamb_>gene218 complete\n",
      "lamb_>gene219 complete\n",
      "lamb_>gene220 complete\n",
      "lamb_>gene221 complete\n",
      "lamb_>gene222 complete\n",
      "lamb_>gene223 complete\n",
      "lamb_>gene224 complete\n",
      "lamb_>gene225 complete\n",
      "lamb_>gene226 complete\n",
      "lamb_>gene227 complete\n",
      "lamb_>gene228 complete\n",
      "lamb_>gene229 complete\n",
      "lamb_>gene230 complete\n",
      "lamb_>gene231 complete\n",
      "lamb_>gene232 complete\n",
      "lamb_>gene233 complete\n",
      "lamb_>gene234 complete\n",
      "lamb_>gene235 complete\n",
      "lamb_>gene236 complete\n",
      "lamb_>gene237 complete\n",
      "lamb_>gene238 complete\n",
      "lamb_>gene239 complete\n",
      "lamb_>gene240 complete\n",
      "lamb_>gene241 complete\n",
      "lamb_>gene242 complete\n",
      "lamb_>gene243 complete\n",
      "lamb_>gene244 complete\n",
      "lamb_>gene245 complete\n",
      "lamb_>gene246 complete\n",
      "lamb_>gene247 complete\n",
      "lamb_>gene248 complete\n",
      "lamb_>gene249 complete\n",
      "lamb_>gene250 complete\n",
      "lamb_>gene251 complete\n",
      "lamb_>gene252 complete\n",
      "lamb_>gene253 complete\n",
      "lamb_>gene254 complete\n",
      "lamb_>gene255 complete\n",
      "lamb_>gene256 complete\n",
      "lamb_>gene257 complete\n",
      "lamb_>gene258 complete\n",
      "lamb_>gene259 complete\n",
      "lamb_>gene260 complete\n",
      "lamb_>gene261 complete\n",
      "lamb_>gene262 complete\n",
      "lamb_>gene263 complete\n",
      "lamb_>gene264 complete\n",
      "lamb_>gene265 complete\n",
      "lamb_>gene266 complete\n",
      "lamb_>gene267 complete\n",
      "lamb_>gene268 complete\n",
      "lamb_>gene269 complete\n",
      "lamb_>gene270 complete\n",
      "lamb_>gene271 complete\n",
      "lamb_>gene272 complete\n",
      "lamb_>gene273 complete\n",
      "lamb_>gene274 complete\n",
      "lamb_>gene275 complete\n",
      "lamb_>gene276 complete\n",
      "lamb_>gene277 complete\n",
      "lamb_>gene278 complete\n",
      "lamb_>gene279 complete\n",
      "lamb_>gene280 complete\n",
      "lamb_>gene281 complete\n",
      "lamb_>gene282 complete\n",
      "lamb_>gene283 complete\n",
      "lamb_>gene284 complete\n",
      "lamb_>gene285 complete\n",
      "lamb_>gene286 complete\n",
      "lamb_>gene287 complete\n",
      "lamb_>gene288 complete\n",
      "lamb_>gene289 complete\n",
      "lamb_>gene290 complete\n",
      "lamb_>gene291 complete\n",
      "lamb_>gene292 complete\n",
      "lamb_>gene293 complete\n",
      "lamb_>gene294 complete\n",
      "lamb_>gene295 complete\n",
      "lamb_>gene296 complete\n",
      "lamb_>gene297 complete\n",
      "lamb_>gene298 complete\n",
      "lamb_>gene299 complete\n",
      "lamb_>gene300 complete\n",
      "lamb_>gene301 complete\n",
      "lamb_>gene302 complete\n",
      "lamb_>gene303 complete\n",
      "lamb_>gene304 complete\n",
      "lamb_>gene305 complete\n",
      "lamb_>gene306 complete\n",
      "lamb_>gene307 complete\n",
      "lamb_>gene308 complete\n",
      "lamb_>gene309 complete\n",
      "lamb_>gene310 complete\n",
      "lamb_>gene311 complete\n",
      "lamb_>gene312 complete\n",
      "lamb_>gene313 complete\n",
      "lamb_>gene314 complete\n",
      "lamb_>gene315 complete\n",
      "lamb_>gene316 complete\n",
      "lamb_>gene317 complete\n",
      "lamb_>gene318 complete\n",
      "lamb_>gene319 complete\n",
      "lamb_>gene320 complete\n",
      "lamb_>gene321 complete\n",
      "lamb_>gene322 complete\n",
      "lamb_>gene323 complete\n",
      "lamb_>gene324 complete\n",
      "lamb_>gene325 complete\n",
      "lamb_>gene326 complete\n",
      "lamb_>gene327 complete\n",
      "lamb_>gene328 complete\n",
      "lamb_>gene329 complete\n",
      "lamb_>gene330 complete\n",
      "lamb_>gene331 complete\n",
      "lamb_>gene332 complete\n",
      "lamb_>gene333 complete\n",
      "lamb_>gene334 complete\n",
      "lamb_>gene335 complete\n",
      "lamb_>gene336 complete\n",
      "lamb_>gene337 complete\n",
      "lamb_>gene338 complete\n",
      "lamb_>gene339 complete\n",
      "lamb_>gene340 complete\n",
      "lamb_>gene341 complete\n",
      "lamb_>gene342 complete\n",
      "lamb_>gene343 complete\n",
      "lamb_>gene344 complete\n",
      "lamb_>gene345 complete\n",
      "lamb_>gene346 complete\n",
      "lamb_>gene347 complete\n",
      "lamb_>gene348 complete\n",
      "lamb_>gene349 complete\n",
      "lamb_>gene350 complete\n",
      "lamb_>gene351 complete\n",
      "lamb_>gene352 complete\n",
      "lamb_>gene353 complete\n",
      "lamb_>gene354 complete\n",
      "lamb_>gene355 complete\n",
      "lamb_>gene356 complete\n",
      "lamb_>gene357 complete\n",
      "lamb_>gene358 complete\n",
      "lamb_>gene359 complete\n",
      "lamb_>gene360 complete\n",
      "lamb_>gene361 complete\n",
      "lamb_>gene362 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lamb_>gene363 complete\n",
      "lamb_>gene364 complete\n",
      "lamb_>gene365 complete\n",
      "lamb_>gene366 complete\n",
      "lamb_>gene367 complete\n",
      "lamb_>gene368 complete\n",
      "lamb_>gene369 complete\n",
      "lamb_>gene370 complete\n",
      "lamb_>gene371 complete\n",
      "lamb_>gene372 complete\n",
      "lamb_>gene373 complete\n",
      "lamb_>gene374 complete\n",
      "lamb_>gene375 complete\n",
      "lamb_>gene376 complete\n",
      "lamb_>gene377 complete\n",
      "lamb_>gene378 complete\n",
      "lamb_>gene379 complete\n",
      "lamb_>gene380 complete\n",
      "lamb_>gene381 complete\n",
      "lamb_>gene382 complete\n",
      "lamb_>gene383 complete\n",
      "lamb_>gene384 complete\n",
      "lamb_>gene385 complete\n",
      "lamb_>gene386 complete\n",
      "lamb_>gene387 complete\n",
      "lamb_>gene388 complete\n",
      "lamb_>gene389 complete\n",
      "lamb_>gene390 complete\n",
      "lamb_>gene391 complete\n",
      "lamb_>gene392 complete\n",
      "lamb_>gene393 complete\n",
      "lamb_>gene394 complete\n",
      "lamb_>gene395 complete\n",
      "lamb_>gene396 complete\n",
      "lamb_>gene397 complete\n",
      "lamb_>gene398 complete\n",
      "lamb_>gene399 complete\n",
      "lamb_>gene400 complete\n",
      "lamb_>gene401 complete\n",
      "lamb_>gene402 complete\n",
      "lamb_>gene403 complete\n",
      "lamb_>gene404 complete\n",
      "lamb_>gene405 complete\n",
      "lamb_>gene406 complete\n",
      "lamb_>gene407 complete\n",
      "lamb_>gene408 complete\n",
      "lamb_>gene409 complete\n",
      "lamb_>gene410 complete\n",
      "lamb_>gene411 complete\n",
      "lamb_>gene412 complete\n",
      "lamb_>gene413 complete\n",
      "lamb_>gene414 complete\n",
      "lamb_>gene415 complete\n",
      "lamb_>gene416 complete\n",
      "lamb_>gene417 complete\n",
      "lamb_>gene418 complete\n",
      "lamb_>gene419 complete\n",
      "lamb_>gene420 complete\n",
      "lamb_>gene421 complete\n",
      "lamb_>gene422 complete\n",
      "lamb_>gene423 complete\n",
      "lamb_>gene424 complete\n",
      "lamb_>gene425 complete\n",
      "lamb_>gene426 complete\n",
      "lamb_>gene427 complete\n",
      "lamb_>gene428 complete\n",
      "lamb_>gene429 complete\n",
      "lamb_>gene430 complete\n",
      "lamb_>gene431 complete\n",
      "lamb_>gene432 complete\n",
      "lamb_>gene433 complete\n",
      "lamb_>gene434 complete\n",
      "lamb_>gene435 complete\n",
      "lamb_>gene436 complete\n",
      "lamb_>gene437 complete\n",
      "lamb_>gene438 complete\n",
      "lamb_>gene439 complete\n",
      "lamb_>gene440 complete\n",
      "lamb_>gene441 complete\n",
      "lamb_>gene442 complete\n",
      "lamb_>gene443 complete\n",
      "lamb_>gene444 complete\n",
      "lamb_>gene445 complete\n",
      "lamb_>gene446 complete\n",
      "lamb_>gene447 complete\n",
      "lamb_>gene448 complete\n",
      "lamb_>gene449 complete\n",
      "lamb_>gene450 complete\n",
      "lamb_>gene451 complete\n",
      "lamb_>gene452 complete\n",
      "lamb_>gene453 complete\n",
      "lamb_>gene454 complete\n",
      "lamb_>gene455 complete\n",
      "lamb_>gene456 complete\n",
      "lamb_>gene457 complete\n",
      "lamb_>gene458 complete\n",
      "lamb_>gene459 complete\n",
      "lamb_>gene460 complete\n",
      "lamb_>gene461 complete\n",
      "lamb_>gene462 complete\n",
      "lamb_>gene463 complete\n",
      "lamb_>gene464 complete\n",
      "lamb_>gene465 complete\n",
      "lamb_>gene466 complete\n",
      "lamb_>gene467 complete\n",
      "lamb_>gene468 complete\n",
      "lamb_>gene469 complete\n",
      "lamb_>gene470 complete\n",
      "lamb_>gene471 complete\n",
      "lamb_>gene472 complete\n",
      "lamb_>gene473 complete\n",
      "lamb_>gene474 complete\n",
      "lamb_>gene475 complete\n",
      "lamb_>gene476 complete\n",
      "lamb_>gene477 complete\n",
      "lamb_>gene478 complete\n",
      "lamb_>gene479 complete\n",
      "lamb_>gene480 complete\n",
      "lamb_>gene481 complete\n",
      "lamb_>gene482 complete\n",
      "lamb_>gene483 complete\n",
      "lamb_>gene484 complete\n",
      "lamb_>gene485 complete\n",
      "lamb_>gene486 complete\n",
      "lamb_>gene487 complete\n",
      "lamb_>gene488 complete\n",
      "lamb_>gene489 complete\n",
      "lamb_>gene490 complete\n",
      "lamb_>gene491 complete\n",
      "lamb_>gene492 complete\n",
      "lamb_>gene493 complete\n",
      "lamb_>gene494 complete\n",
      "lamb_>gene495 complete\n",
      "lamb_>gene496 complete\n",
      "lamb_>gene497 complete\n",
      "lamb_>gene498 complete\n",
      "lamb_>gene499 complete\n",
      "lamb_>gene500 complete\n",
      "dromedary_>gene1 complete\n",
      "dromedary_>gene2 complete\n",
      "dromedary_>gene3 complete\n",
      "dromedary_>gene4 complete\n",
      "dromedary_>gene5 complete\n",
      "dromedary_>gene6 complete\n",
      "dromedary_>gene7 complete\n",
      "dromedary_>gene8 complete\n",
      "dromedary_>gene9 complete\n",
      "dromedary_>gene10 complete\n",
      "dromedary_>gene11 complete\n",
      "dromedary_>gene12 complete\n",
      "dromedary_>gene13 complete\n",
      "dromedary_>gene14 complete\n",
      "dromedary_>gene15 complete\n",
      "dromedary_>gene16 complete\n",
      "dromedary_>gene17 complete\n",
      "dromedary_>gene18 complete\n",
      "dromedary_>gene19 complete\n",
      "dromedary_>gene20 complete\n",
      "dromedary_>gene21 complete\n",
      "dromedary_>gene22 complete\n",
      "dromedary_>gene23 complete\n",
      "dromedary_>gene24 complete\n",
      "dromedary_>gene25 complete\n",
      "dromedary_>gene26 complete\n",
      "dromedary_>gene27 complete\n",
      "dromedary_>gene28 complete\n",
      "dromedary_>gene29 complete\n",
      "dromedary_>gene30 complete\n",
      "dromedary_>gene31 complete\n",
      "dromedary_>gene32 complete\n",
      "dromedary_>gene33 complete\n",
      "dromedary_>gene34 complete\n",
      "dromedary_>gene35 complete\n",
      "dromedary_>gene36 complete\n",
      "dromedary_>gene37 complete\n",
      "dromedary_>gene38 complete\n",
      "dromedary_>gene39 complete\n",
      "dromedary_>gene40 complete\n",
      "dromedary_>gene41 complete\n",
      "dromedary_>gene42 complete\n",
      "dromedary_>gene43 complete\n",
      "dromedary_>gene44 complete\n",
      "dromedary_>gene45 complete\n",
      "dromedary_>gene46 complete\n",
      "dromedary_>gene47 complete\n",
      "dromedary_>gene48 complete\n",
      "dromedary_>gene49 complete\n",
      "dromedary_>gene50 complete\n",
      "dromedary_>gene51 complete\n",
      "dromedary_>gene52 complete\n",
      "dromedary_>gene53 complete\n",
      "dromedary_>gene54 complete\n",
      "dromedary_>gene55 complete\n",
      "dromedary_>gene56 complete\n",
      "dromedary_>gene57 complete\n",
      "dromedary_>gene58 complete\n",
      "dromedary_>gene59 complete\n",
      "dromedary_>gene60 complete\n",
      "dromedary_>gene61 complete\n",
      "dromedary_>gene62 complete\n",
      "dromedary_>gene63 complete\n",
      "dromedary_>gene64 complete\n",
      "dromedary_>gene65 complete\n",
      "dromedary_>gene66 complete\n",
      "dromedary_>gene67 complete\n",
      "dromedary_>gene68 complete\n",
      "dromedary_>gene69 complete\n",
      "dromedary_>gene70 complete\n",
      "dromedary_>gene71 complete\n",
      "dromedary_>gene72 complete\n",
      "dromedary_>gene73 complete\n",
      "dromedary_>gene74 complete\n",
      "dromedary_>gene75 complete\n",
      "dromedary_>gene76 complete\n",
      "dromedary_>gene77 complete\n",
      "dromedary_>gene78 complete\n",
      "dromedary_>gene79 complete\n",
      "dromedary_>gene80 complete\n",
      "dromedary_>gene81 complete\n",
      "dromedary_>gene82 complete\n",
      "dromedary_>gene83 complete\n",
      "dromedary_>gene84 complete\n",
      "dromedary_>gene85 complete\n",
      "dromedary_>gene86 complete\n",
      "dromedary_>gene87 complete\n",
      "dromedary_>gene88 complete\n",
      "dromedary_>gene89 complete\n",
      "dromedary_>gene90 complete\n",
      "dromedary_>gene91 complete\n",
      "dromedary_>gene92 complete\n",
      "dromedary_>gene93 complete\n",
      "dromedary_>gene94 complete\n",
      "dromedary_>gene95 complete\n",
      "dromedary_>gene96 complete\n",
      "dromedary_>gene97 complete\n",
      "dromedary_>gene98 complete\n",
      "dromedary_>gene99 complete\n",
      "dromedary_>gene100 complete\n",
      "dromedary_>gene101 complete\n",
      "dromedary_>gene102 complete\n",
      "dromedary_>gene103 complete\n",
      "dromedary_>gene104 complete\n",
      "dromedary_>gene105 complete\n",
      "dromedary_>gene106 complete\n",
      "dromedary_>gene107 complete\n",
      "dromedary_>gene108 complete\n",
      "dromedary_>gene109 complete\n",
      "dromedary_>gene110 complete\n",
      "dromedary_>gene111 complete\n",
      "dromedary_>gene112 complete\n",
      "dromedary_>gene113 complete\n",
      "dromedary_>gene114 complete\n",
      "dromedary_>gene115 complete\n",
      "dromedary_>gene116 complete\n",
      "dromedary_>gene117 complete\n",
      "dromedary_>gene118 complete\n",
      "dromedary_>gene119 complete\n",
      "dromedary_>gene120 complete\n",
      "dromedary_>gene121 complete\n",
      "dromedary_>gene122 complete\n",
      "dromedary_>gene123 complete\n",
      "dromedary_>gene124 complete\n",
      "dromedary_>gene125 complete\n",
      "dromedary_>gene126 complete\n",
      "dromedary_>gene127 complete\n",
      "dromedary_>gene128 complete\n",
      "dromedary_>gene129 complete\n",
      "dromedary_>gene130 complete\n",
      "dromedary_>gene131 complete\n",
      "dromedary_>gene132 complete\n",
      "dromedary_>gene133 complete\n",
      "dromedary_>gene134 complete\n",
      "dromedary_>gene135 complete\n",
      "dromedary_>gene136 complete\n",
      "dromedary_>gene137 complete\n",
      "dromedary_>gene138 complete\n",
      "dromedary_>gene139 complete\n",
      "dromedary_>gene140 complete\n",
      "dromedary_>gene141 complete\n",
      "dromedary_>gene142 complete\n",
      "dromedary_>gene143 complete\n",
      "dromedary_>gene144 complete\n",
      "dromedary_>gene145 complete\n",
      "dromedary_>gene146 complete\n",
      "dromedary_>gene147 complete\n",
      "dromedary_>gene148 complete\n",
      "dromedary_>gene149 complete\n",
      "dromedary_>gene150 complete\n",
      "dromedary_>gene151 complete\n",
      "dromedary_>gene152 complete\n",
      "dromedary_>gene153 complete\n",
      "dromedary_>gene154 complete\n",
      "dromedary_>gene155 complete\n",
      "dromedary_>gene156 complete\n",
      "dromedary_>gene157 complete\n",
      "dromedary_>gene158 complete\n",
      "dromedary_>gene159 complete\n",
      "dromedary_>gene160 complete\n",
      "dromedary_>gene161 complete\n",
      "dromedary_>gene162 complete\n",
      "dromedary_>gene163 complete\n",
      "dromedary_>gene164 complete\n",
      "dromedary_>gene165 complete\n",
      "dromedary_>gene166 complete\n",
      "dromedary_>gene167 complete\n",
      "dromedary_>gene168 complete\n",
      "dromedary_>gene169 complete\n",
      "dromedary_>gene170 complete\n",
      "dromedary_>gene171 complete\n",
      "dromedary_>gene172 complete\n",
      "dromedary_>gene173 complete\n",
      "dromedary_>gene174 complete\n",
      "dromedary_>gene175 complete\n",
      "dromedary_>gene176 complete\n",
      "dromedary_>gene177 complete\n",
      "dromedary_>gene178 complete\n",
      "dromedary_>gene179 complete\n",
      "dromedary_>gene180 complete\n",
      "dromedary_>gene181 complete\n",
      "dromedary_>gene182 complete\n",
      "dromedary_>gene183 complete\n",
      "dromedary_>gene184 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dromedary_>gene185 complete\n",
      "dromedary_>gene186 complete\n",
      "dromedary_>gene187 complete\n",
      "dromedary_>gene188 complete\n",
      "dromedary_>gene189 complete\n",
      "dromedary_>gene190 complete\n",
      "dromedary_>gene191 complete\n",
      "dromedary_>gene192 complete\n",
      "dromedary_>gene193 complete\n",
      "dromedary_>gene194 complete\n",
      "dromedary_>gene195 complete\n",
      "dromedary_>gene196 complete\n",
      "dromedary_>gene197 complete\n",
      "dromedary_>gene198 complete\n",
      "dromedary_>gene199 complete\n",
      "dromedary_>gene200 complete\n",
      "dromedary_>gene201 complete\n",
      "dromedary_>gene202 complete\n",
      "dromedary_>gene203 complete\n",
      "dromedary_>gene204 complete\n",
      "dromedary_>gene205 complete\n",
      "dromedary_>gene206 complete\n",
      "dromedary_>gene207 complete\n",
      "dromedary_>gene208 complete\n",
      "dromedary_>gene209 complete\n",
      "dromedary_>gene210 complete\n",
      "dromedary_>gene211 complete\n",
      "dromedary_>gene212 complete\n",
      "dromedary_>gene213 complete\n",
      "dromedary_>gene214 complete\n",
      "dromedary_>gene215 complete\n",
      "dromedary_>gene216 complete\n",
      "dromedary_>gene217 complete\n",
      "dromedary_>gene218 complete\n",
      "dromedary_>gene219 complete\n",
      "dromedary_>gene220 complete\n",
      "dromedary_>gene221 complete\n",
      "dromedary_>gene222 complete\n",
      "dromedary_>gene223 complete\n",
      "dromedary_>gene224 complete\n",
      "dromedary_>gene225 complete\n",
      "dromedary_>gene226 complete\n",
      "dromedary_>gene227 complete\n",
      "dromedary_>gene228 complete\n",
      "dromedary_>gene229 complete\n",
      "dromedary_>gene230 complete\n",
      "dromedary_>gene231 complete\n",
      "dromedary_>gene232 complete\n",
      "dromedary_>gene233 complete\n",
      "dromedary_>gene234 complete\n",
      "dromedary_>gene235 complete\n",
      "dromedary_>gene236 complete\n",
      "dromedary_>gene237 complete\n",
      "dromedary_>gene238 complete\n",
      "dromedary_>gene239 complete\n",
      "dromedary_>gene240 complete\n",
      "dromedary_>gene241 complete\n",
      "dromedary_>gene242 complete\n",
      "dromedary_>gene243 complete\n",
      "dromedary_>gene244 complete\n",
      "dromedary_>gene245 complete\n",
      "dromedary_>gene246 complete\n",
      "dromedary_>gene247 complete\n",
      "dromedary_>gene248 complete\n",
      "dromedary_>gene249 complete\n",
      "dromedary_>gene250 complete\n",
      "dromedary_>gene251 complete\n",
      "dromedary_>gene252 complete\n",
      "dromedary_>gene253 complete\n",
      "dromedary_>gene254 complete\n",
      "dromedary_>gene255 complete\n",
      "dromedary_>gene256 complete\n",
      "dromedary_>gene257 complete\n",
      "dromedary_>gene258 complete\n",
      "dromedary_>gene259 complete\n",
      "dromedary_>gene260 complete\n",
      "dromedary_>gene261 complete\n",
      "dromedary_>gene262 complete\n",
      "dromedary_>gene263 complete\n",
      "dromedary_>gene264 complete\n",
      "dromedary_>gene265 complete\n",
      "dromedary_>gene266 complete\n",
      "dromedary_>gene267 complete\n",
      "dromedary_>gene268 complete\n",
      "dromedary_>gene269 complete\n",
      "dromedary_>gene270 complete\n",
      "dromedary_>gene271 complete\n",
      "dromedary_>gene272 complete\n",
      "dromedary_>gene273 complete\n",
      "dromedary_>gene274 complete\n",
      "dromedary_>gene275 complete\n",
      "dromedary_>gene276 complete\n",
      "dromedary_>gene277 complete\n",
      "dromedary_>gene278 complete\n",
      "dromedary_>gene279 complete\n",
      "dromedary_>gene280 complete\n",
      "dromedary_>gene281 complete\n",
      "dromedary_>gene282 complete\n",
      "dromedary_>gene283 complete\n",
      "dromedary_>gene284 complete\n",
      "dromedary_>gene285 complete\n",
      "dromedary_>gene286 complete\n",
      "dromedary_>gene287 complete\n",
      "dromedary_>gene288 complete\n",
      "dromedary_>gene289 complete\n",
      "dromedary_>gene290 complete\n",
      "dromedary_>gene291 complete\n",
      "dromedary_>gene292 complete\n",
      "dromedary_>gene293 complete\n",
      "dromedary_>gene294 complete\n",
      "dromedary_>gene295 complete\n",
      "dromedary_>gene296 complete\n",
      "dromedary_>gene297 complete\n",
      "dromedary_>gene298 complete\n",
      "dromedary_>gene299 complete\n",
      "dromedary_>gene300 complete\n",
      "dromedary_>gene301 complete\n",
      "dromedary_>gene302 complete\n",
      "dromedary_>gene303 complete\n",
      "dromedary_>gene304 complete\n",
      "dromedary_>gene305 complete\n",
      "dromedary_>gene306 complete\n",
      "dromedary_>gene307 complete\n",
      "dromedary_>gene308 complete\n",
      "dromedary_>gene309 complete\n",
      "dromedary_>gene310 complete\n",
      "dromedary_>gene311 complete\n",
      "dromedary_>gene312 complete\n",
      "dromedary_>gene313 complete\n",
      "dromedary_>gene314 complete\n",
      "dromedary_>gene315 complete\n",
      "dromedary_>gene316 complete\n",
      "dromedary_>gene317 complete\n",
      "dromedary_>gene318 complete\n",
      "dromedary_>gene319 complete\n",
      "dromedary_>gene320 complete\n",
      "dromedary_>gene321 complete\n",
      "dromedary_>gene322 complete\n",
      "dromedary_>gene323 complete\n",
      "dromedary_>gene324 complete\n",
      "dromedary_>gene325 complete\n",
      "dromedary_>gene326 complete\n",
      "dromedary_>gene327 complete\n",
      "dromedary_>gene328 complete\n",
      "dromedary_>gene329 complete\n",
      "dromedary_>gene330 complete\n",
      "dromedary_>gene331 complete\n",
      "dromedary_>gene332 complete\n",
      "dromedary_>gene333 complete\n",
      "dromedary_>gene334 complete\n",
      "dromedary_>gene335 complete\n",
      "dromedary_>gene336 complete\n",
      "dromedary_>gene337 complete\n",
      "dromedary_>gene338 complete\n",
      "dromedary_>gene339 complete\n",
      "dromedary_>gene340 complete\n",
      "dromedary_>gene341 complete\n",
      "dromedary_>gene342 complete\n",
      "dromedary_>gene343 complete\n",
      "dromedary_>gene344 complete\n",
      "dromedary_>gene345 complete\n",
      "dromedary_>gene346 complete\n",
      "dromedary_>gene347 complete\n",
      "dromedary_>gene348 complete\n",
      "dromedary_>gene349 complete\n",
      "dromedary_>gene350 complete\n",
      "dromedary_>gene351 complete\n",
      "dromedary_>gene352 complete\n",
      "dromedary_>gene353 complete\n",
      "dromedary_>gene354 complete\n",
      "dromedary_>gene355 complete\n",
      "dromedary_>gene356 complete\n",
      "dromedary_>gene357 complete\n",
      "dromedary_>gene358 complete\n",
      "dromedary_>gene359 complete\n",
      "dromedary_>gene360 complete\n",
      "dromedary_>gene361 complete\n",
      "dromedary_>gene362 complete\n",
      "dromedary_>gene363 complete\n",
      "dromedary_>gene364 complete\n",
      "dromedary_>gene365 complete\n",
      "dromedary_>gene366 complete\n",
      "dromedary_>gene367 complete\n",
      "dromedary_>gene368 complete\n",
      "dromedary_>gene369 complete\n",
      "dromedary_>gene370 complete\n",
      "dromedary_>gene371 complete\n",
      "dromedary_>gene372 complete\n",
      "dromedary_>gene373 complete\n",
      "dromedary_>gene374 complete\n",
      "dromedary_>gene375 complete\n",
      "dromedary_>gene376 complete\n",
      "dromedary_>gene377 complete\n",
      "dromedary_>gene378 complete\n",
      "dromedary_>gene379 complete\n",
      "dromedary_>gene380 complete\n",
      "dromedary_>gene381 complete\n",
      "dromedary_>gene382 complete\n",
      "dromedary_>gene383 complete\n",
      "dromedary_>gene384 complete\n",
      "dromedary_>gene385 complete\n",
      "dromedary_>gene386 complete\n",
      "dromedary_>gene387 complete\n",
      "dromedary_>gene388 complete\n",
      "dromedary_>gene389 complete\n",
      "dromedary_>gene390 complete\n",
      "dromedary_>gene391 complete\n",
      "dromedary_>gene392 complete\n",
      "dromedary_>gene393 complete\n",
      "dromedary_>gene394 complete\n",
      "dromedary_>gene395 complete\n",
      "dromedary_>gene396 complete\n",
      "dromedary_>gene397 complete\n",
      "dromedary_>gene398 complete\n",
      "dromedary_>gene399 complete\n",
      "dromedary_>gene400 complete\n",
      "dromedary_>gene401 complete\n",
      "dromedary_>gene402 complete\n",
      "dromedary_>gene403 complete\n",
      "dromedary_>gene404 complete\n",
      "dromedary_>gene405 complete\n",
      "dromedary_>gene406 complete\n",
      "dromedary_>gene407 complete\n",
      "dromedary_>gene408 complete\n",
      "dromedary_>gene409 complete\n",
      "dromedary_>gene410 complete\n",
      "dromedary_>gene411 complete\n",
      "dromedary_>gene412 complete\n",
      "dromedary_>gene413 complete\n",
      "dromedary_>gene414 complete\n",
      "dromedary_>gene415 complete\n",
      "dromedary_>gene416 complete\n",
      "dromedary_>gene417 complete\n",
      "dromedary_>gene418 complete\n",
      "dromedary_>gene419 complete\n",
      "dromedary_>gene420 complete\n",
      "dromedary_>gene421 complete\n",
      "dromedary_>gene422 complete\n",
      "dromedary_>gene423 complete\n",
      "dromedary_>gene424 complete\n",
      "dromedary_>gene425 complete\n",
      "dromedary_>gene426 complete\n",
      "dromedary_>gene427 complete\n",
      "dromedary_>gene428 complete\n",
      "dromedary_>gene429 complete\n",
      "dromedary_>gene430 complete\n",
      "dromedary_>gene431 complete\n",
      "dromedary_>gene432 complete\n",
      "dromedary_>gene433 complete\n",
      "dromedary_>gene434 complete\n",
      "dromedary_>gene435 complete\n",
      "dromedary_>gene436 complete\n",
      "dromedary_>gene437 complete\n",
      "dromedary_>gene438 complete\n",
      "dromedary_>gene439 complete\n",
      "dromedary_>gene440 complete\n",
      "dromedary_>gene441 complete\n",
      "dromedary_>gene442 complete\n",
      "dromedary_>gene443 complete\n",
      "dromedary_>gene444 complete\n",
      "dromedary_>gene445 complete\n",
      "dromedary_>gene446 complete\n",
      "dromedary_>gene447 complete\n",
      "dromedary_>gene448 complete\n",
      "dromedary_>gene449 complete\n",
      "dromedary_>gene450 complete\n",
      "dromedary_>gene451 complete\n",
      "dromedary_>gene452 complete\n",
      "dromedary_>gene453 complete\n",
      "dromedary_>gene454 complete\n",
      "dromedary_>gene455 complete\n",
      "dromedary_>gene456 complete\n",
      "dromedary_>gene457 complete\n",
      "dromedary_>gene458 complete\n",
      "dromedary_>gene459 complete\n",
      "dromedary_>gene460 complete\n",
      "dromedary_>gene461 complete\n",
      "dromedary_>gene462 complete\n",
      "dromedary_>gene463 complete\n",
      "dromedary_>gene464 complete\n",
      "dromedary_>gene465 complete\n",
      "dromedary_>gene466 complete\n",
      "dromedary_>gene467 complete\n",
      "dromedary_>gene468 complete\n",
      "dromedary_>gene469 complete\n",
      "dromedary_>gene470 complete\n",
      "dromedary_>gene471 complete\n",
      "dromedary_>gene472 complete\n",
      "dromedary_>gene473 complete\n",
      "dromedary_>gene474 complete\n",
      "dromedary_>gene475 complete\n",
      "dromedary_>gene476 complete\n",
      "dromedary_>gene477 complete\n",
      "dromedary_>gene478 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dromedary_>gene479 complete\n",
      "dromedary_>gene480 complete\n",
      "dromedary_>gene481 complete\n",
      "dromedary_>gene482 complete\n",
      "dromedary_>gene483 complete\n",
      "dromedary_>gene484 complete\n",
      "dromedary_>gene485 complete\n",
      "dromedary_>gene486 complete\n",
      "dromedary_>gene487 complete\n",
      "dromedary_>gene488 complete\n",
      "dromedary_>gene489 complete\n",
      "dromedary_>gene490 complete\n",
      "dromedary_>gene491 complete\n",
      "dromedary_>gene492 complete\n",
      "dromedary_>gene493 complete\n",
      "dromedary_>gene494 complete\n",
      "dromedary_>gene495 complete\n",
      "dromedary_>gene496 complete\n",
      "dromedary_>gene497 complete\n",
      "dromedary_>gene498 complete\n",
      "dromedary_>gene499 complete\n",
      "dromedary_>gene500 complete\n",
      "turtle_>gene1 complete\n",
      "turtle_>gene2 complete\n",
      "turtle_>gene3 complete\n",
      "turtle_>gene4 complete\n",
      "turtle_>gene5 complete\n",
      "turtle_>gene6 complete\n",
      "turtle_>gene7 complete\n",
      "turtle_>gene8 complete\n",
      "turtle_>gene9 complete\n",
      "turtle_>gene10 complete\n",
      "turtle_>gene11 complete\n",
      "turtle_>gene12 complete\n",
      "turtle_>gene13 complete\n",
      "turtle_>gene14 complete\n",
      "turtle_>gene15 complete\n",
      "turtle_>gene16 complete\n",
      "turtle_>gene17 complete\n",
      "turtle_>gene18 complete\n",
      "turtle_>gene19 complete\n",
      "turtle_>gene20 complete\n",
      "turtle_>gene21 complete\n",
      "turtle_>gene22 complete\n",
      "turtle_>gene23 complete\n",
      "turtle_>gene24 complete\n",
      "turtle_>gene25 complete\n",
      "turtle_>gene26 complete\n",
      "turtle_>gene27 complete\n",
      "turtle_>gene28 complete\n",
      "turtle_>gene29 complete\n",
      "turtle_>gene30 complete\n",
      "turtle_>gene31 complete\n",
      "turtle_>gene32 complete\n",
      "turtle_>gene33 complete\n",
      "turtle_>gene34 complete\n",
      "turtle_>gene35 complete\n",
      "turtle_>gene36 complete\n",
      "turtle_>gene37 complete\n",
      "turtle_>gene38 complete\n",
      "turtle_>gene39 complete\n",
      "turtle_>gene40 complete\n",
      "turtle_>gene41 complete\n",
      "turtle_>gene42 complete\n",
      "turtle_>gene43 complete\n",
      "turtle_>gene44 complete\n",
      "turtle_>gene45 complete\n",
      "turtle_>gene46 complete\n",
      "turtle_>gene47 complete\n",
      "turtle_>gene48 complete\n",
      "turtle_>gene49 complete\n",
      "turtle_>gene50 complete\n",
      "turtle_>gene51 complete\n",
      "turtle_>gene52 complete\n",
      "turtle_>gene53 complete\n",
      "turtle_>gene54 complete\n",
      "turtle_>gene55 complete\n",
      "turtle_>gene56 complete\n",
      "turtle_>gene57 complete\n",
      "turtle_>gene58 complete\n",
      "turtle_>gene59 complete\n",
      "turtle_>gene60 complete\n",
      "turtle_>gene61 complete\n",
      "turtle_>gene62 complete\n",
      "turtle_>gene63 complete\n",
      "turtle_>gene64 complete\n",
      "turtle_>gene65 complete\n",
      "turtle_>gene66 complete\n",
      "turtle_>gene67 complete\n",
      "turtle_>gene68 complete\n",
      "turtle_>gene69 complete\n",
      "turtle_>gene70 complete\n",
      "turtle_>gene71 complete\n",
      "turtle_>gene72 complete\n",
      "turtle_>gene73 complete\n",
      "turtle_>gene74 complete\n",
      "turtle_>gene75 complete\n",
      "turtle_>gene76 complete\n",
      "turtle_>gene77 complete\n",
      "turtle_>gene78 complete\n",
      "turtle_>gene79 complete\n",
      "turtle_>gene80 complete\n",
      "turtle_>gene81 complete\n",
      "turtle_>gene82 complete\n",
      "turtle_>gene83 complete\n",
      "turtle_>gene84 complete\n",
      "turtle_>gene85 complete\n",
      "turtle_>gene86 complete\n",
      "turtle_>gene87 complete\n",
      "turtle_>gene88 complete\n",
      "turtle_>gene89 complete\n",
      "turtle_>gene90 complete\n",
      "turtle_>gene91 complete\n",
      "turtle_>gene92 complete\n",
      "turtle_>gene93 complete\n",
      "turtle_>gene94 complete\n",
      "turtle_>gene95 complete\n",
      "turtle_>gene96 complete\n",
      "turtle_>gene97 complete\n",
      "turtle_>gene98 complete\n",
      "turtle_>gene99 complete\n",
      "turtle_>gene100 complete\n",
      "turtle_>gene101 complete\n",
      "turtle_>gene102 complete\n",
      "turtle_>gene103 complete\n",
      "turtle_>gene104 complete\n",
      "turtle_>gene105 complete\n",
      "turtle_>gene106 complete\n",
      "turtle_>gene107 complete\n",
      "turtle_>gene108 complete\n",
      "turtle_>gene109 complete\n",
      "turtle_>gene110 complete\n",
      "turtle_>gene111 complete\n",
      "turtle_>gene112 complete\n",
      "turtle_>gene113 complete\n",
      "turtle_>gene114 complete\n",
      "turtle_>gene115 complete\n",
      "turtle_>gene116 complete\n",
      "turtle_>gene117 complete\n",
      "turtle_>gene118 complete\n",
      "turtle_>gene119 complete\n",
      "turtle_>gene120 complete\n",
      "turtle_>gene121 complete\n",
      "turtle_>gene122 complete\n",
      "turtle_>gene123 complete\n",
      "turtle_>gene124 complete\n",
      "turtle_>gene125 complete\n",
      "turtle_>gene126 complete\n",
      "turtle_>gene127 complete\n",
      "turtle_>gene128 complete\n",
      "turtle_>gene129 complete\n",
      "turtle_>gene130 complete\n",
      "turtle_>gene131 complete\n",
      "turtle_>gene132 complete\n",
      "turtle_>gene133 complete\n",
      "turtle_>gene134 complete\n",
      "turtle_>gene135 complete\n",
      "turtle_>gene136 complete\n",
      "turtle_>gene137 complete\n",
      "turtle_>gene138 complete\n",
      "turtle_>gene139 complete\n",
      "turtle_>gene140 complete\n",
      "turtle_>gene141 complete\n",
      "turtle_>gene142 complete\n",
      "turtle_>gene143 complete\n",
      "turtle_>gene144 complete\n",
      "turtle_>gene145 complete\n",
      "turtle_>gene146 complete\n",
      "turtle_>gene147 complete\n",
      "turtle_>gene148 complete\n",
      "turtle_>gene149 complete\n",
      "turtle_>gene150 complete\n",
      "turtle_>gene151 complete\n",
      "turtle_>gene152 complete\n",
      "turtle_>gene153 complete\n",
      "turtle_>gene154 complete\n",
      "turtle_>gene155 complete\n",
      "turtle_>gene156 complete\n",
      "turtle_>gene157 complete\n",
      "turtle_>gene158 complete\n",
      "turtle_>gene159 complete\n",
      "turtle_>gene160 complete\n",
      "turtle_>gene161 complete\n",
      "turtle_>gene162 complete\n",
      "turtle_>gene163 complete\n",
      "turtle_>gene164 complete\n",
      "turtle_>gene165 complete\n",
      "turtle_>gene166 complete\n",
      "turtle_>gene167 complete\n",
      "turtle_>gene168 complete\n",
      "turtle_>gene169 complete\n",
      "turtle_>gene170 complete\n",
      "turtle_>gene171 complete\n",
      "turtle_>gene172 complete\n",
      "turtle_>gene173 complete\n",
      "turtle_>gene174 complete\n",
      "turtle_>gene175 complete\n",
      "turtle_>gene176 complete\n",
      "turtle_>gene177 complete\n",
      "turtle_>gene178 complete\n",
      "turtle_>gene179 complete\n",
      "turtle_>gene180 complete\n",
      "turtle_>gene181 complete\n",
      "turtle_>gene182 complete\n",
      "turtle_>gene183 complete\n",
      "turtle_>gene184 complete\n",
      "turtle_>gene185 complete\n",
      "turtle_>gene186 complete\n",
      "turtle_>gene187 complete\n",
      "turtle_>gene188 complete\n",
      "turtle_>gene189 complete\n",
      "turtle_>gene190 complete\n",
      "turtle_>gene191 complete\n",
      "turtle_>gene192 complete\n",
      "turtle_>gene193 complete\n",
      "turtle_>gene194 complete\n",
      "turtle_>gene195 complete\n",
      "turtle_>gene196 complete\n",
      "turtle_>gene197 complete\n",
      "turtle_>gene198 complete\n",
      "turtle_>gene199 complete\n",
      "turtle_>gene200 complete\n",
      "turtle_>gene201 complete\n",
      "turtle_>gene202 complete\n",
      "turtle_>gene203 complete\n",
      "turtle_>gene204 complete\n",
      "turtle_>gene205 complete\n",
      "turtle_>gene206 complete\n",
      "turtle_>gene207 complete\n",
      "turtle_>gene208 complete\n",
      "turtle_>gene209 complete\n",
      "turtle_>gene210 complete\n",
      "turtle_>gene211 complete\n",
      "turtle_>gene212 complete\n",
      "turtle_>gene213 complete\n",
      "turtle_>gene214 complete\n",
      "turtle_>gene215 complete\n",
      "turtle_>gene216 complete\n",
      "turtle_>gene217 complete\n",
      "turtle_>gene218 complete\n",
      "turtle_>gene219 complete\n",
      "turtle_>gene220 complete\n",
      "turtle_>gene221 complete\n",
      "turtle_>gene222 complete\n",
      "turtle_>gene223 complete\n",
      "turtle_>gene224 complete\n",
      "turtle_>gene225 complete\n",
      "turtle_>gene226 complete\n",
      "turtle_>gene227 complete\n",
      "turtle_>gene228 complete\n",
      "turtle_>gene229 complete\n",
      "turtle_>gene230 complete\n",
      "turtle_>gene231 complete\n",
      "turtle_>gene232 complete\n",
      "turtle_>gene233 complete\n",
      "turtle_>gene234 complete\n",
      "turtle_>gene235 complete\n",
      "turtle_>gene236 complete\n",
      "turtle_>gene237 complete\n",
      "turtle_>gene238 complete\n",
      "turtle_>gene239 complete\n",
      "turtle_>gene240 complete\n",
      "turtle_>gene241 complete\n",
      "turtle_>gene242 complete\n",
      "turtle_>gene243 complete\n",
      "turtle_>gene244 complete\n",
      "turtle_>gene245 complete\n",
      "turtle_>gene246 complete\n",
      "turtle_>gene247 complete\n",
      "turtle_>gene248 complete\n",
      "turtle_>gene249 complete\n",
      "turtle_>gene250 complete\n",
      "turtle_>gene251 complete\n",
      "turtle_>gene252 complete\n",
      "turtle_>gene253 complete\n",
      "turtle_>gene254 complete\n",
      "turtle_>gene255 complete\n",
      "turtle_>gene256 complete\n",
      "turtle_>gene257 complete\n",
      "turtle_>gene258 complete\n",
      "turtle_>gene259 complete\n",
      "turtle_>gene260 complete\n",
      "turtle_>gene261 complete\n",
      "turtle_>gene262 complete\n",
      "turtle_>gene263 complete\n",
      "turtle_>gene264 complete\n",
      "turtle_>gene265 complete\n",
      "turtle_>gene266 complete\n",
      "turtle_>gene267 complete\n",
      "turtle_>gene268 complete\n",
      "turtle_>gene269 complete\n",
      "turtle_>gene270 complete\n",
      "turtle_>gene271 complete\n",
      "turtle_>gene272 complete\n",
      "turtle_>gene273 complete\n",
      "turtle_>gene274 complete\n",
      "turtle_>gene275 complete\n",
      "turtle_>gene276 complete\n",
      "turtle_>gene277 complete\n",
      "turtle_>gene278 complete\n",
      "turtle_>gene279 complete\n",
      "turtle_>gene280 complete\n",
      "turtle_>gene281 complete\n",
      "turtle_>gene282 complete\n",
      "turtle_>gene283 complete\n",
      "turtle_>gene284 complete\n",
      "turtle_>gene285 complete\n",
      "turtle_>gene286 complete\n",
      "turtle_>gene287 complete\n",
      "turtle_>gene288 complete\n",
      "turtle_>gene289 complete\n",
      "turtle_>gene290 complete\n",
      "turtle_>gene291 complete\n",
      "turtle_>gene292 complete\n",
      "turtle_>gene293 complete\n",
      "turtle_>gene294 complete\n",
      "turtle_>gene295 complete\n",
      "turtle_>gene296 complete\n",
      "turtle_>gene297 complete\n",
      "turtle_>gene298 complete\n",
      "turtle_>gene299 complete\n",
      "turtle_>gene300 complete\n",
      "turtle_>gene301 complete\n",
      "turtle_>gene302 complete\n",
      "turtle_>gene303 complete\n",
      "turtle_>gene304 complete\n",
      "turtle_>gene305 complete\n",
      "turtle_>gene306 complete\n",
      "turtle_>gene307 complete\n",
      "turtle_>gene308 complete\n",
      "turtle_>gene309 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "turtle_>gene310 complete\n",
      "turtle_>gene311 complete\n",
      "turtle_>gene312 complete\n",
      "turtle_>gene313 complete\n",
      "turtle_>gene314 complete\n",
      "turtle_>gene315 complete\n",
      "turtle_>gene316 complete\n",
      "turtle_>gene317 complete\n",
      "turtle_>gene318 complete\n",
      "turtle_>gene319 complete\n",
      "turtle_>gene320 complete\n",
      "turtle_>gene321 complete\n",
      "turtle_>gene322 complete\n",
      "turtle_>gene323 complete\n",
      "turtle_>gene324 complete\n",
      "turtle_>gene325 complete\n",
      "turtle_>gene326 complete\n",
      "turtle_>gene327 complete\n",
      "turtle_>gene328 complete\n",
      "turtle_>gene329 complete\n",
      "turtle_>gene330 complete\n",
      "turtle_>gene331 complete\n",
      "turtle_>gene332 complete\n",
      "turtle_>gene333 complete\n",
      "turtle_>gene334 complete\n",
      "turtle_>gene335 complete\n",
      "turtle_>gene336 complete\n",
      "turtle_>gene337 complete\n",
      "turtle_>gene338 complete\n",
      "turtle_>gene339 complete\n",
      "turtle_>gene340 complete\n",
      "turtle_>gene341 complete\n",
      "turtle_>gene342 complete\n",
      "turtle_>gene343 complete\n",
      "turtle_>gene344 complete\n",
      "turtle_>gene345 complete\n",
      "turtle_>gene346 complete\n",
      "turtle_>gene347 complete\n",
      "turtle_>gene348 complete\n",
      "turtle_>gene349 complete\n",
      "turtle_>gene350 complete\n",
      "turtle_>gene351 complete\n",
      "turtle_>gene352 complete\n",
      "turtle_>gene353 complete\n",
      "turtle_>gene354 complete\n",
      "turtle_>gene355 complete\n",
      "turtle_>gene356 complete\n",
      "turtle_>gene357 complete\n",
      "turtle_>gene358 complete\n",
      "turtle_>gene359 complete\n",
      "turtle_>gene360 complete\n",
      "turtle_>gene361 complete\n",
      "turtle_>gene362 complete\n",
      "turtle_>gene363 complete\n",
      "turtle_>gene364 complete\n",
      "turtle_>gene365 complete\n",
      "turtle_>gene366 complete\n",
      "turtle_>gene367 complete\n",
      "turtle_>gene368 complete\n",
      "turtle_>gene369 complete\n",
      "turtle_>gene370 complete\n",
      "turtle_>gene371 complete\n",
      "turtle_>gene372 complete\n",
      "turtle_>gene373 complete\n",
      "turtle_>gene374 complete\n",
      "turtle_>gene375 complete\n",
      "turtle_>gene376 complete\n",
      "turtle_>gene377 complete\n",
      "turtle_>gene378 complete\n",
      "turtle_>gene379 complete\n",
      "turtle_>gene380 complete\n",
      "turtle_>gene381 complete\n",
      "turtle_>gene382 complete\n",
      "turtle_>gene383 complete\n",
      "turtle_>gene384 complete\n",
      "turtle_>gene385 complete\n",
      "turtle_>gene386 complete\n",
      "turtle_>gene387 complete\n",
      "turtle_>gene388 complete\n",
      "turtle_>gene389 complete\n",
      "turtle_>gene390 complete\n",
      "turtle_>gene391 complete\n",
      "turtle_>gene392 complete\n",
      "turtle_>gene393 complete\n",
      "turtle_>gene394 complete\n",
      "turtle_>gene395 complete\n",
      "turtle_>gene396 complete\n",
      "turtle_>gene397 complete\n",
      "turtle_>gene398 complete\n",
      "turtle_>gene399 complete\n",
      "turtle_>gene400 complete\n",
      "turtle_>gene401 complete\n",
      "turtle_>gene402 complete\n",
      "turtle_>gene403 complete\n",
      "turtle_>gene404 complete\n",
      "turtle_>gene405 complete\n",
      "turtle_>gene406 complete\n",
      "turtle_>gene407 complete\n",
      "turtle_>gene408 complete\n",
      "turtle_>gene409 complete\n",
      "turtle_>gene410 complete\n",
      "turtle_>gene411 complete\n",
      "turtle_>gene412 complete\n",
      "turtle_>gene413 complete\n",
      "turtle_>gene414 complete\n",
      "turtle_>gene415 complete\n",
      "turtle_>gene416 complete\n",
      "turtle_>gene417 complete\n",
      "turtle_>gene418 complete\n",
      "turtle_>gene419 complete\n",
      "turtle_>gene420 complete\n",
      "turtle_>gene421 complete\n",
      "turtle_>gene422 complete\n",
      "turtle_>gene423 complete\n",
      "turtle_>gene424 complete\n",
      "turtle_>gene425 complete\n",
      "turtle_>gene426 complete\n",
      "turtle_>gene427 complete\n",
      "turtle_>gene428 complete\n",
      "turtle_>gene429 complete\n",
      "turtle_>gene430 complete\n",
      "turtle_>gene431 complete\n",
      "turtle_>gene432 complete\n",
      "turtle_>gene433 complete\n",
      "turtle_>gene434 complete\n",
      "turtle_>gene435 complete\n",
      "turtle_>gene436 complete\n",
      "turtle_>gene437 complete\n",
      "turtle_>gene438 complete\n",
      "turtle_>gene439 complete\n",
      "turtle_>gene440 complete\n",
      "turtle_>gene441 complete\n",
      "turtle_>gene442 complete\n",
      "turtle_>gene443 complete\n",
      "turtle_>gene444 complete\n",
      "turtle_>gene445 complete\n",
      "turtle_>gene446 complete\n",
      "turtle_>gene447 complete\n",
      "turtle_>gene448 complete\n",
      "turtle_>gene449 complete\n",
      "turtle_>gene450 complete\n",
      "turtle_>gene451 complete\n",
      "turtle_>gene452 complete\n",
      "turtle_>gene453 complete\n",
      "turtle_>gene454 complete\n",
      "turtle_>gene455 complete\n",
      "turtle_>gene456 complete\n",
      "turtle_>gene457 complete\n",
      "turtle_>gene458 complete\n",
      "turtle_>gene459 complete\n",
      "turtle_>gene460 complete\n",
      "turtle_>gene461 complete\n",
      "turtle_>gene462 complete\n",
      "turtle_>gene463 complete\n",
      "turtle_>gene464 complete\n",
      "turtle_>gene465 complete\n",
      "turtle_>gene466 complete\n",
      "turtle_>gene467 complete\n",
      "turtle_>gene468 complete\n",
      "turtle_>gene469 complete\n",
      "turtle_>gene470 complete\n",
      "turtle_>gene471 complete\n",
      "turtle_>gene472 complete\n",
      "turtle_>gene473 complete\n",
      "turtle_>gene474 complete\n",
      "turtle_>gene475 complete\n",
      "turtle_>gene476 complete\n",
      "turtle_>gene477 complete\n",
      "turtle_>gene478 complete\n",
      "turtle_>gene479 complete\n",
      "turtle_>gene480 complete\n",
      "turtle_>gene481 complete\n",
      "turtle_>gene482 complete\n",
      "turtle_>gene483 complete\n",
      "turtle_>gene484 complete\n",
      "turtle_>gene485 complete\n",
      "turtle_>gene486 complete\n",
      "turtle_>gene487 complete\n",
      "turtle_>gene488 complete\n",
      "turtle_>gene489 complete\n",
      "turtle_>gene490 complete\n",
      "turtle_>gene491 complete\n",
      "turtle_>gene492 complete\n",
      "turtle_>gene493 complete\n",
      "turtle_>gene494 complete\n",
      "turtle_>gene495 complete\n",
      "turtle_>gene496 complete\n",
      "turtle_>gene497 complete\n",
      "turtle_>gene498 complete\n",
      "turtle_>gene499 complete\n",
      "turtle_>gene500 complete\n",
      "bunny_>gene1 complete\n",
      "bunny_>gene2 complete\n",
      "bunny_>gene3 complete\n",
      "bunny_>gene4 complete\n",
      "bunny_>gene5 complete\n",
      "bunny_>gene6 complete\n",
      "bunny_>gene7 complete\n",
      "bunny_>gene8 complete\n",
      "bunny_>gene9 complete\n",
      "bunny_>gene10 complete\n",
      "bunny_>gene11 complete\n",
      "bunny_>gene12 complete\n",
      "bunny_>gene13 complete\n",
      "bunny_>gene14 complete\n",
      "bunny_>gene15 complete\n",
      "bunny_>gene16 complete\n",
      "bunny_>gene17 complete\n",
      "bunny_>gene18 complete\n",
      "bunny_>gene19 complete\n",
      "bunny_>gene20 complete\n",
      "bunny_>gene21 complete\n",
      "bunny_>gene22 complete\n",
      "bunny_>gene23 complete\n",
      "bunny_>gene24 complete\n",
      "bunny_>gene25 complete\n",
      "bunny_>gene26 complete\n",
      "bunny_>gene27 complete\n",
      "bunny_>gene28 complete\n",
      "bunny_>gene29 complete\n",
      "bunny_>gene30 complete\n",
      "bunny_>gene31 complete\n",
      "bunny_>gene32 complete\n",
      "bunny_>gene33 complete\n",
      "bunny_>gene34 complete\n",
      "bunny_>gene35 complete\n",
      "bunny_>gene36 complete\n",
      "bunny_>gene37 complete\n",
      "bunny_>gene38 complete\n",
      "bunny_>gene39 complete\n",
      "bunny_>gene40 complete\n",
      "bunny_>gene41 complete\n",
      "bunny_>gene42 complete\n",
      "bunny_>gene43 complete\n",
      "bunny_>gene44 complete\n",
      "bunny_>gene45 complete\n",
      "bunny_>gene46 complete\n",
      "bunny_>gene47 complete\n",
      "bunny_>gene48 complete\n",
      "bunny_>gene49 complete\n",
      "bunny_>gene50 complete\n",
      "bunny_>gene51 complete\n",
      "bunny_>gene52 complete\n",
      "bunny_>gene53 complete\n",
      "bunny_>gene54 complete\n",
      "bunny_>gene55 complete\n",
      "bunny_>gene56 complete\n",
      "bunny_>gene57 complete\n",
      "bunny_>gene58 complete\n",
      "bunny_>gene59 complete\n",
      "bunny_>gene60 complete\n",
      "bunny_>gene61 complete\n",
      "bunny_>gene62 complete\n",
      "bunny_>gene63 complete\n",
      "bunny_>gene64 complete\n",
      "bunny_>gene65 complete\n",
      "bunny_>gene66 complete\n",
      "bunny_>gene67 complete\n",
      "bunny_>gene68 complete\n",
      "bunny_>gene69 complete\n",
      "bunny_>gene70 complete\n",
      "bunny_>gene71 complete\n",
      "bunny_>gene72 complete\n",
      "bunny_>gene73 complete\n",
      "bunny_>gene74 complete\n",
      "bunny_>gene75 complete\n",
      "bunny_>gene76 complete\n",
      "bunny_>gene77 complete\n",
      "bunny_>gene78 complete\n",
      "bunny_>gene79 complete\n",
      "bunny_>gene80 complete\n",
      "bunny_>gene81 complete\n",
      "bunny_>gene82 complete\n",
      "bunny_>gene83 complete\n",
      "bunny_>gene84 complete\n",
      "bunny_>gene85 complete\n",
      "bunny_>gene86 complete\n",
      "bunny_>gene87 complete\n",
      "bunny_>gene88 complete\n",
      "bunny_>gene89 complete\n",
      "bunny_>gene90 complete\n",
      "bunny_>gene91 complete\n",
      "bunny_>gene92 complete\n",
      "bunny_>gene93 complete\n",
      "bunny_>gene94 complete\n",
      "bunny_>gene95 complete\n",
      "bunny_>gene96 complete\n",
      "bunny_>gene97 complete\n",
      "bunny_>gene98 complete\n",
      "bunny_>gene99 complete\n",
      "bunny_>gene100 complete\n",
      "bunny_>gene101 complete\n",
      "bunny_>gene102 complete\n",
      "bunny_>gene103 complete\n",
      "bunny_>gene104 complete\n",
      "bunny_>gene105 complete\n",
      "bunny_>gene106 complete\n",
      "bunny_>gene107 complete\n",
      "bunny_>gene108 complete\n",
      "bunny_>gene109 complete\n",
      "bunny_>gene110 complete\n",
      "bunny_>gene111 complete\n",
      "bunny_>gene112 complete\n",
      "bunny_>gene113 complete\n",
      "bunny_>gene114 complete\n",
      "bunny_>gene115 complete\n",
      "bunny_>gene116 complete\n",
      "bunny_>gene117 complete\n",
      "bunny_>gene118 complete\n",
      "bunny_>gene119 complete\n",
      "bunny_>gene120 complete\n",
      "bunny_>gene121 complete\n",
      "bunny_>gene122 complete\n",
      "bunny_>gene123 complete\n",
      "bunny_>gene124 complete\n",
      "bunny_>gene125 complete\n",
      "bunny_>gene126 complete\n",
      "bunny_>gene127 complete\n",
      "bunny_>gene128 complete\n",
      "bunny_>gene129 complete\n",
      "bunny_>gene130 complete\n",
      "bunny_>gene131 complete\n",
      "bunny_>gene132 complete\n",
      "bunny_>gene133 complete\n",
      "bunny_>gene134 complete\n",
      "bunny_>gene135 complete\n",
      "bunny_>gene136 complete\n",
      "bunny_>gene137 complete\n",
      "bunny_>gene138 complete\n",
      "bunny_>gene139 complete\n",
      "bunny_>gene140 complete\n",
      "bunny_>gene141 complete\n",
      "bunny_>gene142 complete\n",
      "bunny_>gene143 complete\n",
      "bunny_>gene144 complete\n",
      "bunny_>gene145 complete\n",
      "bunny_>gene146 complete\n",
      "bunny_>gene147 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bunny_>gene148 complete\n",
      "bunny_>gene149 complete\n",
      "bunny_>gene150 complete\n",
      "bunny_>gene151 complete\n",
      "bunny_>gene152 complete\n",
      "bunny_>gene153 complete\n",
      "bunny_>gene154 complete\n",
      "bunny_>gene155 complete\n",
      "bunny_>gene156 complete\n",
      "bunny_>gene157 complete\n",
      "bunny_>gene158 complete\n",
      "bunny_>gene159 complete\n",
      "bunny_>gene160 complete\n",
      "bunny_>gene161 complete\n",
      "bunny_>gene162 complete\n",
      "bunny_>gene163 complete\n",
      "bunny_>gene164 complete\n",
      "bunny_>gene165 complete\n",
      "bunny_>gene166 complete\n",
      "bunny_>gene167 complete\n",
      "bunny_>gene168 complete\n",
      "bunny_>gene169 complete\n",
      "bunny_>gene170 complete\n",
      "bunny_>gene171 complete\n",
      "bunny_>gene172 complete\n",
      "bunny_>gene173 complete\n",
      "bunny_>gene174 complete\n",
      "bunny_>gene175 complete\n",
      "bunny_>gene176 complete\n",
      "bunny_>gene177 complete\n",
      "bunny_>gene178 complete\n",
      "bunny_>gene179 complete\n",
      "bunny_>gene180 complete\n",
      "bunny_>gene181 complete\n",
      "bunny_>gene182 complete\n",
      "bunny_>gene183 complete\n",
      "bunny_>gene184 complete\n",
      "bunny_>gene185 complete\n",
      "bunny_>gene186 complete\n",
      "bunny_>gene187 complete\n",
      "bunny_>gene188 complete\n",
      "bunny_>gene189 complete\n",
      "bunny_>gene190 complete\n",
      "bunny_>gene191 complete\n",
      "bunny_>gene192 complete\n",
      "bunny_>gene193 complete\n",
      "bunny_>gene194 complete\n",
      "bunny_>gene195 complete\n",
      "bunny_>gene196 complete\n",
      "bunny_>gene197 complete\n",
      "bunny_>gene198 complete\n",
      "bunny_>gene199 complete\n",
      "bunny_>gene200 complete\n",
      "bunny_>gene201 complete\n",
      "bunny_>gene202 complete\n",
      "bunny_>gene203 complete\n",
      "bunny_>gene204 complete\n",
      "bunny_>gene205 complete\n",
      "bunny_>gene206 complete\n",
      "bunny_>gene207 complete\n",
      "bunny_>gene208 complete\n",
      "bunny_>gene209 complete\n",
      "bunny_>gene210 complete\n",
      "bunny_>gene211 complete\n",
      "bunny_>gene212 complete\n",
      "bunny_>gene213 complete\n",
      "bunny_>gene214 complete\n",
      "bunny_>gene215 complete\n",
      "bunny_>gene216 complete\n",
      "bunny_>gene217 complete\n",
      "bunny_>gene218 complete\n",
      "bunny_>gene219 complete\n",
      "bunny_>gene220 complete\n",
      "bunny_>gene221 complete\n",
      "bunny_>gene222 complete\n",
      "bunny_>gene223 complete\n",
      "bunny_>gene224 complete\n",
      "bunny_>gene225 complete\n",
      "bunny_>gene226 complete\n",
      "bunny_>gene227 complete\n",
      "bunny_>gene228 complete\n",
      "bunny_>gene229 complete\n",
      "bunny_>gene230 complete\n",
      "bunny_>gene231 complete\n",
      "bunny_>gene232 complete\n",
      "bunny_>gene233 complete\n",
      "bunny_>gene234 complete\n",
      "bunny_>gene235 complete\n",
      "bunny_>gene236 complete\n",
      "bunny_>gene237 complete\n",
      "bunny_>gene238 complete\n",
      "bunny_>gene239 complete\n",
      "bunny_>gene240 complete\n",
      "bunny_>gene241 complete\n",
      "bunny_>gene242 complete\n",
      "bunny_>gene243 complete\n",
      "bunny_>gene244 complete\n",
      "bunny_>gene245 complete\n",
      "bunny_>gene246 complete\n",
      "bunny_>gene247 complete\n",
      "bunny_>gene248 complete\n",
      "bunny_>gene249 complete\n",
      "bunny_>gene250 complete\n",
      "bunny_>gene251 complete\n",
      "bunny_>gene252 complete\n",
      "bunny_>gene253 complete\n",
      "bunny_>gene254 complete\n",
      "bunny_>gene255 complete\n",
      "bunny_>gene256 complete\n",
      "bunny_>gene257 complete\n",
      "bunny_>gene258 complete\n",
      "bunny_>gene259 complete\n",
      "bunny_>gene260 complete\n",
      "bunny_>gene261 complete\n",
      "bunny_>gene262 complete\n",
      "bunny_>gene263 complete\n",
      "bunny_>gene264 complete\n",
      "bunny_>gene265 complete\n",
      "bunny_>gene266 complete\n",
      "bunny_>gene267 complete\n",
      "bunny_>gene268 complete\n",
      "bunny_>gene269 complete\n",
      "bunny_>gene270 complete\n",
      "bunny_>gene271 complete\n",
      "bunny_>gene272 complete\n",
      "bunny_>gene273 complete\n",
      "bunny_>gene274 complete\n",
      "bunny_>gene275 complete\n",
      "bunny_>gene276 complete\n",
      "bunny_>gene277 complete\n",
      "bunny_>gene278 complete\n",
      "bunny_>gene279 complete\n",
      "bunny_>gene280 complete\n",
      "bunny_>gene281 complete\n",
      "bunny_>gene282 complete\n",
      "bunny_>gene283 complete\n",
      "bunny_>gene284 complete\n",
      "bunny_>gene285 complete\n",
      "bunny_>gene286 complete\n",
      "bunny_>gene287 complete\n",
      "bunny_>gene288 complete\n",
      "bunny_>gene289 complete\n",
      "bunny_>gene290 complete\n",
      "bunny_>gene291 complete\n",
      "bunny_>gene292 complete\n",
      "bunny_>gene293 complete\n",
      "bunny_>gene294 complete\n",
      "bunny_>gene295 complete\n",
      "bunny_>gene296 complete\n",
      "bunny_>gene297 complete\n",
      "bunny_>gene298 complete\n",
      "bunny_>gene299 complete\n",
      "bunny_>gene300 complete\n",
      "bunny_>gene301 complete\n",
      "bunny_>gene302 complete\n",
      "bunny_>gene303 complete\n",
      "bunny_>gene304 complete\n",
      "bunny_>gene305 complete\n",
      "bunny_>gene306 complete\n",
      "bunny_>gene307 complete\n",
      "bunny_>gene308 complete\n",
      "bunny_>gene309 complete\n",
      "bunny_>gene310 complete\n",
      "bunny_>gene311 complete\n",
      "bunny_>gene312 complete\n",
      "bunny_>gene313 complete\n",
      "bunny_>gene314 complete\n",
      "bunny_>gene315 complete\n",
      "bunny_>gene316 complete\n",
      "bunny_>gene317 complete\n",
      "bunny_>gene318 complete\n",
      "bunny_>gene319 complete\n",
      "bunny_>gene320 complete\n",
      "bunny_>gene321 complete\n",
      "bunny_>gene322 complete\n",
      "bunny_>gene323 complete\n",
      "bunny_>gene324 complete\n",
      "bunny_>gene325 complete\n",
      "bunny_>gene326 complete\n",
      "bunny_>gene327 complete\n",
      "bunny_>gene328 complete\n",
      "bunny_>gene329 complete\n",
      "bunny_>gene330 complete\n",
      "bunny_>gene331 complete\n",
      "bunny_>gene332 complete\n",
      "bunny_>gene333 complete\n",
      "bunny_>gene334 complete\n",
      "bunny_>gene335 complete\n",
      "bunny_>gene336 complete\n",
      "bunny_>gene337 complete\n",
      "bunny_>gene338 complete\n",
      "bunny_>gene339 complete\n",
      "bunny_>gene340 complete\n",
      "bunny_>gene341 complete\n",
      "bunny_>gene342 complete\n",
      "bunny_>gene343 complete\n",
      "bunny_>gene344 complete\n",
      "bunny_>gene345 complete\n",
      "bunny_>gene346 complete\n",
      "bunny_>gene347 complete\n",
      "bunny_>gene348 complete\n",
      "bunny_>gene349 complete\n",
      "bunny_>gene350 complete\n",
      "bunny_>gene351 complete\n",
      "bunny_>gene352 complete\n",
      "bunny_>gene353 complete\n",
      "bunny_>gene354 complete\n",
      "bunny_>gene355 complete\n",
      "bunny_>gene356 complete\n",
      "bunny_>gene357 complete\n",
      "bunny_>gene358 complete\n",
      "bunny_>gene359 complete\n",
      "bunny_>gene360 complete\n",
      "bunny_>gene361 complete\n",
      "bunny_>gene362 complete\n",
      "bunny_>gene363 complete\n",
      "bunny_>gene364 complete\n",
      "bunny_>gene365 complete\n",
      "bunny_>gene366 complete\n",
      "bunny_>gene367 complete\n",
      "bunny_>gene368 complete\n",
      "bunny_>gene369 complete\n",
      "bunny_>gene370 complete\n",
      "bunny_>gene371 complete\n",
      "bunny_>gene372 complete\n",
      "bunny_>gene373 complete\n",
      "bunny_>gene374 complete\n",
      "bunny_>gene375 complete\n",
      "bunny_>gene376 complete\n",
      "bunny_>gene377 complete\n",
      "bunny_>gene378 complete\n",
      "bunny_>gene379 complete\n",
      "bunny_>gene380 complete\n",
      "bunny_>gene381 complete\n",
      "bunny_>gene382 complete\n",
      "bunny_>gene383 complete\n",
      "bunny_>gene384 complete\n",
      "bunny_>gene385 complete\n",
      "bunny_>gene386 complete\n",
      "bunny_>gene387 complete\n",
      "bunny_>gene388 complete\n",
      "bunny_>gene389 complete\n",
      "bunny_>gene390 complete\n",
      "bunny_>gene391 complete\n",
      "bunny_>gene392 complete\n",
      "bunny_>gene393 complete\n",
      "bunny_>gene394 complete\n",
      "bunny_>gene395 complete\n",
      "bunny_>gene396 complete\n",
      "bunny_>gene397 complete\n",
      "bunny_>gene398 complete\n",
      "bunny_>gene399 complete\n",
      "bunny_>gene400 complete\n",
      "bunny_>gene401 complete\n",
      "bunny_>gene402 complete\n",
      "bunny_>gene403 complete\n",
      "bunny_>gene404 complete\n",
      "bunny_>gene405 complete\n",
      "bunny_>gene406 complete\n",
      "bunny_>gene407 complete\n",
      "bunny_>gene408 complete\n",
      "bunny_>gene409 complete\n",
      "bunny_>gene410 complete\n",
      "bunny_>gene411 complete\n",
      "bunny_>gene412 complete\n",
      "bunny_>gene413 complete\n",
      "bunny_>gene414 complete\n",
      "bunny_>gene415 complete\n",
      "bunny_>gene416 complete\n",
      "bunny_>gene417 complete\n",
      "bunny_>gene418 complete\n",
      "bunny_>gene419 complete\n",
      "bunny_>gene420 complete\n",
      "bunny_>gene421 complete\n",
      "bunny_>gene422 complete\n",
      "bunny_>gene423 complete\n",
      "bunny_>gene424 complete\n",
      "bunny_>gene425 complete\n",
      "bunny_>gene426 complete\n",
      "bunny_>gene427 complete\n",
      "bunny_>gene428 complete\n",
      "bunny_>gene429 complete\n",
      "bunny_>gene430 complete\n",
      "bunny_>gene431 complete\n",
      "bunny_>gene432 complete\n",
      "bunny_>gene433 complete\n",
      "bunny_>gene434 complete\n",
      "bunny_>gene435 complete\n",
      "bunny_>gene436 complete\n",
      "bunny_>gene437 complete\n",
      "bunny_>gene438 complete\n",
      "bunny_>gene439 complete\n",
      "bunny_>gene440 complete\n",
      "bunny_>gene441 complete\n",
      "bunny_>gene442 complete\n",
      "bunny_>gene443 complete\n",
      "bunny_>gene444 complete\n",
      "bunny_>gene445 complete\n",
      "bunny_>gene446 complete\n",
      "bunny_>gene447 complete\n",
      "bunny_>gene448 complete\n",
      "bunny_>gene449 complete\n",
      "bunny_>gene450 complete\n",
      "bunny_>gene451 complete\n",
      "bunny_>gene452 complete\n",
      "bunny_>gene453 complete\n",
      "bunny_>gene454 complete\n",
      "bunny_>gene455 complete\n",
      "bunny_>gene456 complete\n",
      "bunny_>gene457 complete\n",
      "bunny_>gene458 complete\n",
      "bunny_>gene459 complete\n",
      "bunny_>gene460 complete\n",
      "bunny_>gene461 complete\n",
      "bunny_>gene462 complete\n",
      "bunny_>gene463 complete\n",
      "bunny_>gene464 complete\n",
      "bunny_>gene465 complete\n",
      "bunny_>gene466 complete\n",
      "bunny_>gene467 complete\n",
      "bunny_>gene468 complete\n",
      "bunny_>gene469 complete\n",
      "bunny_>gene470 complete\n",
      "bunny_>gene471 complete\n",
      "bunny_>gene472 complete\n",
      "bunny_>gene473 complete\n",
      "bunny_>gene474 complete\n",
      "bunny_>gene475 complete\n",
      "bunny_>gene476 complete\n",
      "bunny_>gene477 complete\n",
      "bunny_>gene478 complete\n",
      "bunny_>gene479 complete\n",
      "bunny_>gene480 complete\n",
      "bunny_>gene481 complete\n",
      "bunny_>gene482 complete\n",
      "bunny_>gene483 complete\n",
      "bunny_>gene484 complete\n",
      "bunny_>gene485 complete\n",
      "bunny_>gene486 complete\n",
      "bunny_>gene487 complete\n",
      "bunny_>gene488 complete\n",
      "bunny_>gene489 complete\n",
      "bunny_>gene490 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bunny_>gene491 complete\n",
      "bunny_>gene492 complete\n",
      "bunny_>gene493 complete\n",
      "bunny_>gene494 complete\n",
      "bunny_>gene495 complete\n",
      "bunny_>gene496 complete\n",
      "bunny_>gene497 complete\n",
      "bunny_>gene498 complete\n",
      "bunny_>gene499 complete\n",
      "bunny_>gene500 complete\n",
      "lovebird_>gene1 complete\n",
      "lovebird_>gene2 complete\n",
      "lovebird_>gene3 complete\n",
      "lovebird_>gene4 complete\n",
      "lovebird_>gene5 complete\n",
      "lovebird_>gene6 complete\n",
      "lovebird_>gene7 complete\n",
      "lovebird_>gene8 complete\n",
      "lovebird_>gene9 complete\n",
      "lovebird_>gene10 complete\n",
      "lovebird_>gene11 complete\n",
      "lovebird_>gene12 complete\n",
      "lovebird_>gene13 complete\n",
      "lovebird_>gene14 complete\n",
      "lovebird_>gene15 complete\n",
      "lovebird_>gene16 complete\n",
      "lovebird_>gene17 complete\n",
      "lovebird_>gene18 complete\n",
      "lovebird_>gene19 complete\n",
      "lovebird_>gene20 complete\n",
      "lovebird_>gene21 complete\n",
      "lovebird_>gene22 complete\n",
      "lovebird_>gene23 complete\n",
      "lovebird_>gene24 complete\n",
      "lovebird_>gene25 complete\n",
      "lovebird_>gene26 complete\n",
      "lovebird_>gene27 complete\n",
      "lovebird_>gene28 complete\n",
      "lovebird_>gene29 complete\n",
      "lovebird_>gene30 complete\n",
      "lovebird_>gene31 complete\n",
      "lovebird_>gene32 complete\n",
      "lovebird_>gene33 complete\n",
      "lovebird_>gene34 complete\n",
      "lovebird_>gene35 complete\n",
      "lovebird_>gene36 complete\n",
      "lovebird_>gene37 complete\n",
      "lovebird_>gene38 complete\n",
      "lovebird_>gene39 complete\n",
      "lovebird_>gene40 complete\n",
      "lovebird_>gene41 complete\n",
      "lovebird_>gene42 complete\n",
      "lovebird_>gene43 complete\n",
      "lovebird_>gene44 complete\n",
      "lovebird_>gene45 complete\n",
      "lovebird_>gene46 complete\n",
      "lovebird_>gene47 complete\n",
      "lovebird_>gene48 complete\n",
      "lovebird_>gene49 complete\n",
      "lovebird_>gene50 complete\n",
      "lovebird_>gene51 complete\n",
      "lovebird_>gene52 complete\n",
      "lovebird_>gene53 complete\n",
      "lovebird_>gene54 complete\n",
      "lovebird_>gene55 complete\n",
      "lovebird_>gene56 complete\n",
      "lovebird_>gene57 complete\n",
      "lovebird_>gene58 complete\n",
      "lovebird_>gene59 complete\n",
      "lovebird_>gene60 complete\n",
      "lovebird_>gene61 complete\n",
      "lovebird_>gene62 complete\n",
      "lovebird_>gene63 complete\n",
      "lovebird_>gene64 complete\n",
      "lovebird_>gene65 complete\n",
      "lovebird_>gene66 complete\n",
      "lovebird_>gene67 complete\n",
      "lovebird_>gene68 complete\n",
      "lovebird_>gene69 complete\n",
      "lovebird_>gene70 complete\n",
      "lovebird_>gene71 complete\n",
      "lovebird_>gene72 complete\n",
      "lovebird_>gene73 complete\n",
      "lovebird_>gene74 complete\n",
      "lovebird_>gene75 complete\n",
      "lovebird_>gene76 complete\n",
      "lovebird_>gene77 complete\n",
      "lovebird_>gene78 complete\n",
      "lovebird_>gene79 complete\n",
      "lovebird_>gene80 complete\n",
      "lovebird_>gene81 complete\n",
      "lovebird_>gene82 complete\n",
      "lovebird_>gene83 complete\n",
      "lovebird_>gene84 complete\n",
      "lovebird_>gene85 complete\n",
      "lovebird_>gene86 complete\n",
      "lovebird_>gene87 complete\n",
      "lovebird_>gene88 complete\n",
      "lovebird_>gene89 complete\n",
      "lovebird_>gene90 complete\n",
      "lovebird_>gene91 complete\n",
      "lovebird_>gene92 complete\n",
      "lovebird_>gene93 complete\n",
      "lovebird_>gene94 complete\n",
      "lovebird_>gene95 complete\n",
      "lovebird_>gene96 complete\n",
      "lovebird_>gene97 complete\n",
      "lovebird_>gene98 complete\n",
      "lovebird_>gene99 complete\n",
      "lovebird_>gene100 complete\n",
      "lovebird_>gene101 complete\n",
      "lovebird_>gene102 complete\n",
      "lovebird_>gene103 complete\n",
      "lovebird_>gene104 complete\n",
      "lovebird_>gene105 complete\n",
      "lovebird_>gene106 complete\n",
      "lovebird_>gene107 complete\n",
      "lovebird_>gene108 complete\n",
      "lovebird_>gene109 complete\n",
      "lovebird_>gene110 complete\n",
      "lovebird_>gene111 complete\n",
      "lovebird_>gene112 complete\n",
      "lovebird_>gene113 complete\n",
      "lovebird_>gene114 complete\n",
      "lovebird_>gene115 complete\n",
      "lovebird_>gene116 complete\n",
      "lovebird_>gene117 complete\n",
      "lovebird_>gene118 complete\n",
      "lovebird_>gene119 complete\n",
      "lovebird_>gene120 complete\n",
      "lovebird_>gene121 complete\n",
      "lovebird_>gene122 complete\n",
      "lovebird_>gene123 complete\n",
      "lovebird_>gene124 complete\n",
      "lovebird_>gene125 complete\n",
      "lovebird_>gene126 complete\n",
      "lovebird_>gene127 complete\n",
      "lovebird_>gene128 complete\n",
      "lovebird_>gene129 complete\n",
      "lovebird_>gene130 complete\n",
      "lovebird_>gene131 complete\n",
      "lovebird_>gene132 complete\n",
      "lovebird_>gene133 complete\n",
      "lovebird_>gene134 complete\n",
      "lovebird_>gene135 complete\n",
      "lovebird_>gene136 complete\n",
      "lovebird_>gene137 complete\n",
      "lovebird_>gene138 complete\n",
      "lovebird_>gene139 complete\n",
      "lovebird_>gene140 complete\n",
      "lovebird_>gene141 complete\n",
      "lovebird_>gene142 complete\n",
      "lovebird_>gene143 complete\n",
      "lovebird_>gene144 complete\n",
      "lovebird_>gene145 complete\n",
      "lovebird_>gene146 complete\n",
      "lovebird_>gene147 complete\n",
      "lovebird_>gene148 complete\n",
      "lovebird_>gene149 complete\n",
      "lovebird_>gene150 complete\n",
      "lovebird_>gene151 complete\n",
      "lovebird_>gene152 complete\n",
      "lovebird_>gene153 complete\n",
      "lovebird_>gene154 complete\n",
      "lovebird_>gene155 complete\n",
      "lovebird_>gene156 complete\n",
      "lovebird_>gene157 complete\n",
      "lovebird_>gene158 complete\n",
      "lovebird_>gene159 complete\n",
      "lovebird_>gene160 complete\n",
      "lovebird_>gene161 complete\n",
      "lovebird_>gene162 complete\n",
      "lovebird_>gene163 complete\n",
      "lovebird_>gene164 complete\n",
      "lovebird_>gene165 complete\n",
      "lovebird_>gene166 complete\n",
      "lovebird_>gene167 complete\n",
      "lovebird_>gene168 complete\n",
      "lovebird_>gene169 complete\n",
      "lovebird_>gene170 complete\n",
      "lovebird_>gene171 complete\n",
      "lovebird_>gene172 complete\n",
      "lovebird_>gene173 complete\n",
      "lovebird_>gene174 complete\n",
      "lovebird_>gene175 complete\n",
      "lovebird_>gene176 complete\n",
      "lovebird_>gene177 complete\n",
      "lovebird_>gene178 complete\n",
      "lovebird_>gene179 complete\n",
      "lovebird_>gene180 complete\n",
      "lovebird_>gene181 complete\n",
      "lovebird_>gene182 complete\n",
      "lovebird_>gene183 complete\n",
      "lovebird_>gene184 complete\n",
      "lovebird_>gene185 complete\n",
      "lovebird_>gene186 complete\n",
      "lovebird_>gene187 complete\n",
      "lovebird_>gene188 complete\n",
      "lovebird_>gene189 complete\n",
      "lovebird_>gene190 complete\n",
      "lovebird_>gene191 complete\n",
      "lovebird_>gene192 complete\n",
      "lovebird_>gene193 complete\n",
      "lovebird_>gene194 complete\n",
      "lovebird_>gene195 complete\n",
      "lovebird_>gene196 complete\n",
      "lovebird_>gene197 complete\n",
      "lovebird_>gene198 complete\n",
      "lovebird_>gene199 complete\n",
      "lovebird_>gene200 complete\n",
      "lovebird_>gene201 complete\n",
      "lovebird_>gene202 complete\n",
      "lovebird_>gene203 complete\n",
      "lovebird_>gene204 complete\n",
      "lovebird_>gene205 complete\n",
      "lovebird_>gene206 complete\n",
      "lovebird_>gene207 complete\n",
      "lovebird_>gene208 complete\n",
      "lovebird_>gene209 complete\n",
      "lovebird_>gene210 complete\n",
      "lovebird_>gene211 complete\n",
      "lovebird_>gene212 complete\n",
      "lovebird_>gene213 complete\n",
      "lovebird_>gene214 complete\n",
      "lovebird_>gene215 complete\n",
      "lovebird_>gene216 complete\n",
      "lovebird_>gene217 complete\n",
      "lovebird_>gene218 complete\n",
      "lovebird_>gene219 complete\n",
      "lovebird_>gene220 complete\n",
      "lovebird_>gene221 complete\n",
      "lovebird_>gene222 complete\n",
      "lovebird_>gene223 complete\n",
      "lovebird_>gene224 complete\n",
      "lovebird_>gene225 complete\n",
      "lovebird_>gene226 complete\n",
      "lovebird_>gene227 complete\n",
      "lovebird_>gene228 complete\n",
      "lovebird_>gene229 complete\n",
      "lovebird_>gene230 complete\n",
      "lovebird_>gene231 complete\n",
      "lovebird_>gene232 complete\n",
      "lovebird_>gene233 complete\n",
      "lovebird_>gene234 complete\n",
      "lovebird_>gene235 complete\n",
      "lovebird_>gene236 complete\n",
      "lovebird_>gene237 complete\n",
      "lovebird_>gene238 complete\n",
      "lovebird_>gene239 complete\n",
      "lovebird_>gene240 complete\n",
      "lovebird_>gene241 complete\n",
      "lovebird_>gene242 complete\n",
      "lovebird_>gene243 complete\n",
      "lovebird_>gene244 complete\n",
      "lovebird_>gene245 complete\n",
      "lovebird_>gene246 complete\n",
      "lovebird_>gene247 complete\n",
      "lovebird_>gene248 complete\n",
      "lovebird_>gene249 complete\n",
      "lovebird_>gene250 complete\n",
      "lovebird_>gene251 complete\n",
      "lovebird_>gene252 complete\n",
      "lovebird_>gene253 complete\n",
      "lovebird_>gene254 complete\n",
      "lovebird_>gene255 complete\n",
      "lovebird_>gene256 complete\n",
      "lovebird_>gene257 complete\n",
      "lovebird_>gene258 complete\n",
      "lovebird_>gene259 complete\n",
      "lovebird_>gene260 complete\n",
      "lovebird_>gene261 complete\n",
      "lovebird_>gene262 complete\n",
      "lovebird_>gene263 complete\n",
      "lovebird_>gene264 complete\n",
      "lovebird_>gene265 complete\n",
      "lovebird_>gene266 complete\n",
      "lovebird_>gene267 complete\n",
      "lovebird_>gene268 complete\n",
      "lovebird_>gene269 complete\n",
      "lovebird_>gene270 complete\n",
      "lovebird_>gene271 complete\n",
      "lovebird_>gene272 complete\n",
      "lovebird_>gene273 complete\n",
      "lovebird_>gene274 complete\n",
      "lovebird_>gene275 complete\n",
      "lovebird_>gene276 complete\n",
      "lovebird_>gene277 complete\n",
      "lovebird_>gene278 complete\n",
      "lovebird_>gene279 complete\n",
      "lovebird_>gene280 complete\n",
      "lovebird_>gene281 complete\n",
      "lovebird_>gene282 complete\n",
      "lovebird_>gene283 complete\n",
      "lovebird_>gene284 complete\n",
      "lovebird_>gene285 complete\n",
      "lovebird_>gene286 complete\n",
      "lovebird_>gene287 complete\n",
      "lovebird_>gene288 complete\n",
      "lovebird_>gene289 complete\n",
      "lovebird_>gene290 complete\n",
      "lovebird_>gene291 complete\n",
      "lovebird_>gene292 complete\n",
      "lovebird_>gene293 complete\n",
      "lovebird_>gene294 complete\n",
      "lovebird_>gene295 complete\n",
      "lovebird_>gene296 complete\n",
      "lovebird_>gene297 complete\n",
      "lovebird_>gene298 complete\n",
      "lovebird_>gene299 complete\n",
      "lovebird_>gene300 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lovebird_>gene301 complete\n",
      "lovebird_>gene302 complete\n",
      "lovebird_>gene303 complete\n",
      "lovebird_>gene304 complete\n",
      "lovebird_>gene305 complete\n",
      "lovebird_>gene306 complete\n",
      "lovebird_>gene307 complete\n",
      "lovebird_>gene308 complete\n",
      "lovebird_>gene309 complete\n",
      "lovebird_>gene310 complete\n",
      "lovebird_>gene311 complete\n",
      "lovebird_>gene312 complete\n",
      "lovebird_>gene313 complete\n",
      "lovebird_>gene314 complete\n",
      "lovebird_>gene315 complete\n",
      "lovebird_>gene316 complete\n",
      "lovebird_>gene317 complete\n",
      "lovebird_>gene318 complete\n",
      "lovebird_>gene319 complete\n",
      "lovebird_>gene320 complete\n",
      "lovebird_>gene321 complete\n",
      "lovebird_>gene322 complete\n",
      "lovebird_>gene323 complete\n",
      "lovebird_>gene324 complete\n",
      "lovebird_>gene325 complete\n",
      "lovebird_>gene326 complete\n",
      "lovebird_>gene327 complete\n",
      "lovebird_>gene328 complete\n",
      "lovebird_>gene329 complete\n",
      "lovebird_>gene330 complete\n",
      "lovebird_>gene331 complete\n",
      "lovebird_>gene332 complete\n",
      "lovebird_>gene333 complete\n",
      "lovebird_>gene334 complete\n",
      "lovebird_>gene335 complete\n",
      "lovebird_>gene336 complete\n",
      "lovebird_>gene337 complete\n",
      "lovebird_>gene338 complete\n",
      "lovebird_>gene339 complete\n",
      "lovebird_>gene340 complete\n",
      "lovebird_>gene341 complete\n",
      "lovebird_>gene342 complete\n",
      "lovebird_>gene343 complete\n",
      "lovebird_>gene344 complete\n",
      "lovebird_>gene345 complete\n",
      "lovebird_>gene346 complete\n",
      "lovebird_>gene347 complete\n",
      "lovebird_>gene348 complete\n",
      "lovebird_>gene349 complete\n",
      "lovebird_>gene350 complete\n",
      "lovebird_>gene351 complete\n",
      "lovebird_>gene352 complete\n",
      "lovebird_>gene353 complete\n",
      "lovebird_>gene354 complete\n",
      "lovebird_>gene355 complete\n",
      "lovebird_>gene356 complete\n",
      "lovebird_>gene357 complete\n",
      "lovebird_>gene358 complete\n",
      "lovebird_>gene359 complete\n",
      "lovebird_>gene360 complete\n",
      "lovebird_>gene361 complete\n",
      "lovebird_>gene362 complete\n",
      "lovebird_>gene363 complete\n",
      "lovebird_>gene364 complete\n",
      "lovebird_>gene365 complete\n",
      "lovebird_>gene366 complete\n",
      "lovebird_>gene367 complete\n",
      "lovebird_>gene368 complete\n",
      "lovebird_>gene369 complete\n",
      "lovebird_>gene370 complete\n",
      "lovebird_>gene371 complete\n",
      "lovebird_>gene372 complete\n",
      "lovebird_>gene373 complete\n",
      "lovebird_>gene374 complete\n",
      "lovebird_>gene375 complete\n",
      "lovebird_>gene376 complete\n",
      "lovebird_>gene377 complete\n",
      "lovebird_>gene378 complete\n",
      "lovebird_>gene379 complete\n",
      "lovebird_>gene380 complete\n",
      "lovebird_>gene381 complete\n",
      "lovebird_>gene382 complete\n",
      "lovebird_>gene383 complete\n",
      "lovebird_>gene384 complete\n",
      "lovebird_>gene385 complete\n",
      "lovebird_>gene386 complete\n",
      "lovebird_>gene387 complete\n",
      "lovebird_>gene388 complete\n",
      "lovebird_>gene389 complete\n",
      "lovebird_>gene390 complete\n",
      "lovebird_>gene391 complete\n",
      "lovebird_>gene392 complete\n",
      "lovebird_>gene393 complete\n",
      "lovebird_>gene394 complete\n",
      "lovebird_>gene395 complete\n",
      "lovebird_>gene396 complete\n",
      "lovebird_>gene397 complete\n",
      "lovebird_>gene398 complete\n",
      "lovebird_>gene399 complete\n",
      "lovebird_>gene400 complete\n",
      "lovebird_>gene401 complete\n",
      "lovebird_>gene402 complete\n",
      "lovebird_>gene403 complete\n",
      "lovebird_>gene404 complete\n",
      "lovebird_>gene405 complete\n",
      "lovebird_>gene406 complete\n",
      "lovebird_>gene407 complete\n",
      "lovebird_>gene408 complete\n",
      "lovebird_>gene409 complete\n",
      "lovebird_>gene410 complete\n",
      "lovebird_>gene411 complete\n",
      "lovebird_>gene412 complete\n",
      "lovebird_>gene413 complete\n",
      "lovebird_>gene414 complete\n",
      "lovebird_>gene415 complete\n",
      "lovebird_>gene416 complete\n",
      "lovebird_>gene417 complete\n",
      "lovebird_>gene418 complete\n",
      "lovebird_>gene419 complete\n",
      "lovebird_>gene420 complete\n",
      "lovebird_>gene421 complete\n",
      "lovebird_>gene422 complete\n",
      "lovebird_>gene423 complete\n",
      "lovebird_>gene424 complete\n",
      "lovebird_>gene425 complete\n",
      "lovebird_>gene426 complete\n",
      "lovebird_>gene427 complete\n",
      "lovebird_>gene428 complete\n",
      "lovebird_>gene429 complete\n",
      "lovebird_>gene430 complete\n",
      "lovebird_>gene431 complete\n",
      "lovebird_>gene432 complete\n",
      "lovebird_>gene433 complete\n",
      "lovebird_>gene434 complete\n",
      "lovebird_>gene435 complete\n",
      "lovebird_>gene436 complete\n",
      "lovebird_>gene437 complete\n",
      "lovebird_>gene438 complete\n",
      "lovebird_>gene439 complete\n",
      "lovebird_>gene440 complete\n",
      "lovebird_>gene441 complete\n",
      "lovebird_>gene442 complete\n",
      "lovebird_>gene443 complete\n",
      "lovebird_>gene444 complete\n",
      "lovebird_>gene445 complete\n",
      "lovebird_>gene446 complete\n",
      "lovebird_>gene447 complete\n",
      "lovebird_>gene448 complete\n",
      "lovebird_>gene449 complete\n",
      "lovebird_>gene450 complete\n",
      "lovebird_>gene451 complete\n",
      "lovebird_>gene452 complete\n",
      "lovebird_>gene453 complete\n",
      "lovebird_>gene454 complete\n",
      "lovebird_>gene455 complete\n",
      "lovebird_>gene456 complete\n",
      "lovebird_>gene457 complete\n",
      "lovebird_>gene458 complete\n",
      "lovebird_>gene459 complete\n",
      "lovebird_>gene460 complete\n",
      "lovebird_>gene461 complete\n",
      "lovebird_>gene462 complete\n",
      "lovebird_>gene463 complete\n",
      "lovebird_>gene464 complete\n",
      "lovebird_>gene465 complete\n",
      "lovebird_>gene466 complete\n",
      "lovebird_>gene467 complete\n",
      "lovebird_>gene468 complete\n",
      "lovebird_>gene469 complete\n",
      "lovebird_>gene470 complete\n",
      "lovebird_>gene471 complete\n",
      "lovebird_>gene472 complete\n",
      "lovebird_>gene473 complete\n",
      "lovebird_>gene474 complete\n",
      "lovebird_>gene475 complete\n",
      "lovebird_>gene476 complete\n",
      "lovebird_>gene477 complete\n",
      "lovebird_>gene478 complete\n",
      "lovebird_>gene479 complete\n",
      "lovebird_>gene480 complete\n",
      "lovebird_>gene481 complete\n",
      "lovebird_>gene482 complete\n",
      "lovebird_>gene483 complete\n",
      "lovebird_>gene484 complete\n",
      "lovebird_>gene485 complete\n",
      "lovebird_>gene486 complete\n",
      "lovebird_>gene487 complete\n",
      "lovebird_>gene488 complete\n",
      "lovebird_>gene489 complete\n",
      "lovebird_>gene490 complete\n",
      "lovebird_>gene491 complete\n",
      "lovebird_>gene492 complete\n",
      "lovebird_>gene493 complete\n",
      "lovebird_>gene494 complete\n",
      "lovebird_>gene495 complete\n",
      "lovebird_>gene496 complete\n",
      "lovebird_>gene497 complete\n",
      "lovebird_>gene498 complete\n",
      "lovebird_>gene499 complete\n",
      "lovebird_>gene500 complete\n",
      "gorilla_>gene1 complete\n",
      "gorilla_>gene2 complete\n",
      "gorilla_>gene3 complete\n",
      "gorilla_>gene4 complete\n",
      "gorilla_>gene5 complete\n",
      "gorilla_>gene6 complete\n",
      "gorilla_>gene7 complete\n",
      "gorilla_>gene8 complete\n",
      "gorilla_>gene9 complete\n",
      "gorilla_>gene10 complete\n",
      "gorilla_>gene11 complete\n",
      "gorilla_>gene12 complete\n",
      "gorilla_>gene13 complete\n",
      "gorilla_>gene14 complete\n",
      "gorilla_>gene15 complete\n",
      "gorilla_>gene16 complete\n",
      "gorilla_>gene17 complete\n",
      "gorilla_>gene18 complete\n",
      "gorilla_>gene19 complete\n",
      "gorilla_>gene20 complete\n",
      "gorilla_>gene21 complete\n",
      "gorilla_>gene22 complete\n",
      "gorilla_>gene23 complete\n",
      "gorilla_>gene24 complete\n",
      "gorilla_>gene25 complete\n",
      "gorilla_>gene26 complete\n",
      "gorilla_>gene27 complete\n",
      "gorilla_>gene28 complete\n",
      "gorilla_>gene29 complete\n",
      "gorilla_>gene30 complete\n",
      "gorilla_>gene31 complete\n",
      "gorilla_>gene32 complete\n",
      "gorilla_>gene33 complete\n",
      "gorilla_>gene34 complete\n",
      "gorilla_>gene35 complete\n",
      "gorilla_>gene36 complete\n",
      "gorilla_>gene37 complete\n",
      "gorilla_>gene38 complete\n",
      "gorilla_>gene39 complete\n",
      "gorilla_>gene40 complete\n",
      "gorilla_>gene41 complete\n",
      "gorilla_>gene42 complete\n",
      "gorilla_>gene43 complete\n",
      "gorilla_>gene44 complete\n",
      "gorilla_>gene45 complete\n",
      "gorilla_>gene46 complete\n",
      "gorilla_>gene47 complete\n",
      "gorilla_>gene48 complete\n",
      "gorilla_>gene49 complete\n",
      "gorilla_>gene50 complete\n",
      "gorilla_>gene51 complete\n",
      "gorilla_>gene52 complete\n",
      "gorilla_>gene53 complete\n",
      "gorilla_>gene54 complete\n",
      "gorilla_>gene55 complete\n",
      "gorilla_>gene56 complete\n",
      "gorilla_>gene57 complete\n",
      "gorilla_>gene58 complete\n",
      "gorilla_>gene59 complete\n",
      "gorilla_>gene60 complete\n",
      "gorilla_>gene61 complete\n",
      "gorilla_>gene62 complete\n",
      "gorilla_>gene63 complete\n",
      "gorilla_>gene64 complete\n",
      "gorilla_>gene65 complete\n",
      "gorilla_>gene66 complete\n",
      "gorilla_>gene67 complete\n",
      "gorilla_>gene68 complete\n",
      "gorilla_>gene69 complete\n",
      "gorilla_>gene70 complete\n",
      "gorilla_>gene71 complete\n",
      "gorilla_>gene72 complete\n",
      "gorilla_>gene73 complete\n",
      "gorilla_>gene74 complete\n",
      "gorilla_>gene75 complete\n",
      "gorilla_>gene76 complete\n",
      "gorilla_>gene77 complete\n",
      "gorilla_>gene78 complete\n",
      "gorilla_>gene79 complete\n",
      "gorilla_>gene80 complete\n",
      "gorilla_>gene81 complete\n",
      "gorilla_>gene82 complete\n",
      "gorilla_>gene83 complete\n",
      "gorilla_>gene84 complete\n",
      "gorilla_>gene85 complete\n",
      "gorilla_>gene86 complete\n",
      "gorilla_>gene87 complete\n",
      "gorilla_>gene88 complete\n",
      "gorilla_>gene89 complete\n",
      "gorilla_>gene90 complete\n",
      "gorilla_>gene91 complete\n",
      "gorilla_>gene92 complete\n",
      "gorilla_>gene93 complete\n",
      "gorilla_>gene94 complete\n",
      "gorilla_>gene95 complete\n",
      "gorilla_>gene96 complete\n",
      "gorilla_>gene97 complete\n",
      "gorilla_>gene98 complete\n",
      "gorilla_>gene99 complete\n",
      "gorilla_>gene100 complete\n",
      "gorilla_>gene101 complete\n",
      "gorilla_>gene102 complete\n",
      "gorilla_>gene103 complete\n",
      "gorilla_>gene104 complete\n",
      "gorilla_>gene105 complete\n",
      "gorilla_>gene106 complete\n",
      "gorilla_>gene107 complete\n",
      "gorilla_>gene108 complete\n",
      "gorilla_>gene109 complete\n",
      "gorilla_>gene110 complete\n",
      "gorilla_>gene111 complete\n",
      "gorilla_>gene112 complete\n",
      "gorilla_>gene113 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gorilla_>gene114 complete\n",
      "gorilla_>gene115 complete\n",
      "gorilla_>gene116 complete\n",
      "gorilla_>gene117 complete\n",
      "gorilla_>gene118 complete\n",
      "gorilla_>gene119 complete\n",
      "gorilla_>gene120 complete\n",
      "gorilla_>gene121 complete\n",
      "gorilla_>gene122 complete\n",
      "gorilla_>gene123 complete\n",
      "gorilla_>gene124 complete\n",
      "gorilla_>gene125 complete\n",
      "gorilla_>gene126 complete\n",
      "gorilla_>gene127 complete\n",
      "gorilla_>gene128 complete\n",
      "gorilla_>gene129 complete\n",
      "gorilla_>gene130 complete\n",
      "gorilla_>gene131 complete\n",
      "gorilla_>gene132 complete\n",
      "gorilla_>gene133 complete\n",
      "gorilla_>gene134 complete\n",
      "gorilla_>gene135 complete\n",
      "gorilla_>gene136 complete\n",
      "gorilla_>gene137 complete\n",
      "gorilla_>gene138 complete\n",
      "gorilla_>gene139 complete\n",
      "gorilla_>gene140 complete\n",
      "gorilla_>gene141 complete\n",
      "gorilla_>gene142 complete\n",
      "gorilla_>gene143 complete\n",
      "gorilla_>gene144 complete\n",
      "gorilla_>gene145 complete\n",
      "gorilla_>gene146 complete\n",
      "gorilla_>gene147 complete\n",
      "gorilla_>gene148 complete\n",
      "gorilla_>gene149 complete\n",
      "gorilla_>gene150 complete\n",
      "gorilla_>gene151 complete\n",
      "gorilla_>gene152 complete\n",
      "gorilla_>gene153 complete\n",
      "gorilla_>gene154 complete\n",
      "gorilla_>gene155 complete\n",
      "gorilla_>gene156 complete\n",
      "gorilla_>gene157 complete\n",
      "gorilla_>gene158 complete\n",
      "gorilla_>gene159 complete\n",
      "gorilla_>gene160 complete\n",
      "gorilla_>gene161 complete\n",
      "gorilla_>gene162 complete\n",
      "gorilla_>gene163 complete\n",
      "gorilla_>gene164 complete\n",
      "gorilla_>gene165 complete\n",
      "gorilla_>gene166 complete\n",
      "gorilla_>gene167 complete\n",
      "gorilla_>gene168 complete\n",
      "gorilla_>gene169 complete\n",
      "gorilla_>gene170 complete\n",
      "gorilla_>gene171 complete\n",
      "gorilla_>gene172 complete\n",
      "gorilla_>gene173 complete\n",
      "gorilla_>gene174 complete\n",
      "gorilla_>gene175 complete\n",
      "gorilla_>gene176 complete\n",
      "gorilla_>gene177 complete\n",
      "gorilla_>gene178 complete\n",
      "gorilla_>gene179 complete\n",
      "gorilla_>gene180 complete\n",
      "gorilla_>gene181 complete\n",
      "gorilla_>gene182 complete\n",
      "gorilla_>gene183 complete\n",
      "gorilla_>gene184 complete\n",
      "gorilla_>gene185 complete\n",
      "gorilla_>gene186 complete\n",
      "gorilla_>gene187 complete\n",
      "gorilla_>gene188 complete\n",
      "gorilla_>gene189 complete\n",
      "gorilla_>gene190 complete\n",
      "gorilla_>gene191 complete\n",
      "gorilla_>gene192 complete\n",
      "gorilla_>gene193 complete\n",
      "gorilla_>gene194 complete\n",
      "gorilla_>gene195 complete\n",
      "gorilla_>gene196 complete\n",
      "gorilla_>gene197 complete\n",
      "gorilla_>gene198 complete\n",
      "gorilla_>gene199 complete\n",
      "gorilla_>gene200 complete\n",
      "gorilla_>gene201 complete\n",
      "gorilla_>gene202 complete\n",
      "gorilla_>gene203 complete\n",
      "gorilla_>gene204 complete\n",
      "gorilla_>gene205 complete\n",
      "gorilla_>gene206 complete\n",
      "gorilla_>gene207 complete\n",
      "gorilla_>gene208 complete\n",
      "gorilla_>gene209 complete\n",
      "gorilla_>gene210 complete\n",
      "gorilla_>gene211 complete\n",
      "gorilla_>gene212 complete\n",
      "gorilla_>gene213 complete\n",
      "gorilla_>gene214 complete\n",
      "gorilla_>gene215 complete\n",
      "gorilla_>gene216 complete\n",
      "gorilla_>gene217 complete\n",
      "gorilla_>gene218 complete\n",
      "gorilla_>gene219 complete\n",
      "gorilla_>gene220 complete\n",
      "gorilla_>gene221 complete\n",
      "gorilla_>gene222 complete\n",
      "gorilla_>gene223 complete\n",
      "gorilla_>gene224 complete\n",
      "gorilla_>gene225 complete\n",
      "gorilla_>gene226 complete\n",
      "gorilla_>gene227 complete\n",
      "gorilla_>gene228 complete\n",
      "gorilla_>gene229 complete\n",
      "gorilla_>gene230 complete\n",
      "gorilla_>gene231 complete\n",
      "gorilla_>gene232 complete\n",
      "gorilla_>gene233 complete\n",
      "gorilla_>gene234 complete\n",
      "gorilla_>gene235 complete\n",
      "gorilla_>gene236 complete\n",
      "gorilla_>gene237 complete\n",
      "gorilla_>gene238 complete\n",
      "gorilla_>gene239 complete\n",
      "gorilla_>gene240 complete\n",
      "gorilla_>gene241 complete\n",
      "gorilla_>gene242 complete\n",
      "gorilla_>gene243 complete\n",
      "gorilla_>gene244 complete\n",
      "gorilla_>gene245 complete\n",
      "gorilla_>gene246 complete\n",
      "gorilla_>gene247 complete\n",
      "gorilla_>gene248 complete\n",
      "gorilla_>gene249 complete\n",
      "gorilla_>gene250 complete\n",
      "gorilla_>gene251 complete\n",
      "gorilla_>gene252 complete\n",
      "gorilla_>gene253 complete\n",
      "gorilla_>gene254 complete\n",
      "gorilla_>gene255 complete\n",
      "gorilla_>gene256 complete\n",
      "gorilla_>gene257 complete\n",
      "gorilla_>gene258 complete\n",
      "gorilla_>gene259 complete\n",
      "gorilla_>gene260 complete\n",
      "gorilla_>gene261 complete\n",
      "gorilla_>gene262 complete\n",
      "gorilla_>gene263 complete\n",
      "gorilla_>gene264 complete\n",
      "gorilla_>gene265 complete\n",
      "gorilla_>gene266 complete\n",
      "gorilla_>gene267 complete\n",
      "gorilla_>gene268 complete\n",
      "gorilla_>gene269 complete\n",
      "gorilla_>gene270 complete\n",
      "gorilla_>gene271 complete\n",
      "gorilla_>gene272 complete\n",
      "gorilla_>gene273 complete\n",
      "gorilla_>gene274 complete\n",
      "gorilla_>gene275 complete\n",
      "gorilla_>gene276 complete\n",
      "gorilla_>gene277 complete\n",
      "gorilla_>gene278 complete\n",
      "gorilla_>gene279 complete\n",
      "gorilla_>gene280 complete\n",
      "gorilla_>gene281 complete\n",
      "gorilla_>gene282 complete\n",
      "gorilla_>gene283 complete\n",
      "gorilla_>gene284 complete\n",
      "gorilla_>gene285 complete\n",
      "gorilla_>gene286 complete\n",
      "gorilla_>gene287 complete\n",
      "gorilla_>gene288 complete\n",
      "gorilla_>gene289 complete\n",
      "gorilla_>gene290 complete\n",
      "gorilla_>gene291 complete\n",
      "gorilla_>gene292 complete\n",
      "gorilla_>gene293 complete\n",
      "gorilla_>gene294 complete\n",
      "gorilla_>gene295 complete\n",
      "gorilla_>gene296 complete\n",
      "gorilla_>gene297 complete\n",
      "gorilla_>gene298 complete\n",
      "gorilla_>gene299 complete\n",
      "gorilla_>gene300 complete\n",
      "gorilla_>gene301 complete\n",
      "gorilla_>gene302 complete\n",
      "gorilla_>gene303 complete\n",
      "gorilla_>gene304 complete\n",
      "gorilla_>gene305 complete\n",
      "gorilla_>gene306 complete\n",
      "gorilla_>gene307 complete\n",
      "gorilla_>gene308 complete\n",
      "gorilla_>gene309 complete\n",
      "gorilla_>gene310 complete\n",
      "gorilla_>gene311 complete\n",
      "gorilla_>gene312 complete\n",
      "gorilla_>gene313 complete\n",
      "gorilla_>gene314 complete\n",
      "gorilla_>gene315 complete\n",
      "gorilla_>gene316 complete\n",
      "gorilla_>gene317 complete\n",
      "gorilla_>gene318 complete\n",
      "gorilla_>gene319 complete\n",
      "gorilla_>gene320 complete\n",
      "gorilla_>gene321 complete\n",
      "gorilla_>gene322 complete\n",
      "gorilla_>gene323 complete\n",
      "gorilla_>gene324 complete\n",
      "gorilla_>gene325 complete\n",
      "gorilla_>gene326 complete\n",
      "gorilla_>gene327 complete\n",
      "gorilla_>gene328 complete\n",
      "gorilla_>gene329 complete\n",
      "gorilla_>gene330 complete\n",
      "gorilla_>gene331 complete\n",
      "gorilla_>gene332 complete\n",
      "gorilla_>gene333 complete\n",
      "gorilla_>gene334 complete\n",
      "gorilla_>gene335 complete\n",
      "gorilla_>gene336 complete\n",
      "gorilla_>gene337 complete\n",
      "gorilla_>gene338 complete\n",
      "gorilla_>gene339 complete\n",
      "gorilla_>gene340 complete\n",
      "gorilla_>gene341 complete\n",
      "gorilla_>gene342 complete\n",
      "gorilla_>gene343 complete\n",
      "gorilla_>gene344 complete\n",
      "gorilla_>gene345 complete\n",
      "gorilla_>gene346 complete\n",
      "gorilla_>gene347 complete\n",
      "gorilla_>gene348 complete\n",
      "gorilla_>gene349 complete\n",
      "gorilla_>gene350 complete\n",
      "gorilla_>gene351 complete\n",
      "gorilla_>gene352 complete\n",
      "gorilla_>gene353 complete\n",
      "gorilla_>gene354 complete\n",
      "gorilla_>gene355 complete\n",
      "gorilla_>gene356 complete\n",
      "gorilla_>gene357 complete\n",
      "gorilla_>gene358 complete\n",
      "gorilla_>gene359 complete\n",
      "gorilla_>gene360 complete\n",
      "gorilla_>gene361 complete\n",
      "gorilla_>gene362 complete\n",
      "gorilla_>gene363 complete\n",
      "gorilla_>gene364 complete\n",
      "gorilla_>gene365 complete\n",
      "gorilla_>gene366 complete\n",
      "gorilla_>gene367 complete\n",
      "gorilla_>gene368 complete\n",
      "gorilla_>gene369 complete\n",
      "gorilla_>gene370 complete\n",
      "gorilla_>gene371 complete\n",
      "gorilla_>gene372 complete\n",
      "gorilla_>gene373 complete\n",
      "gorilla_>gene374 complete\n",
      "gorilla_>gene375 complete\n",
      "gorilla_>gene376 complete\n",
      "gorilla_>gene377 complete\n",
      "gorilla_>gene378 complete\n",
      "gorilla_>gene379 complete\n",
      "gorilla_>gene380 complete\n",
      "gorilla_>gene381 complete\n",
      "gorilla_>gene382 complete\n",
      "gorilla_>gene383 complete\n",
      "gorilla_>gene384 complete\n",
      "gorilla_>gene385 complete\n",
      "gorilla_>gene386 complete\n",
      "gorilla_>gene387 complete\n",
      "gorilla_>gene388 complete\n",
      "gorilla_>gene389 complete\n",
      "gorilla_>gene390 complete\n",
      "gorilla_>gene391 complete\n",
      "gorilla_>gene392 complete\n",
      "gorilla_>gene393 complete\n",
      "gorilla_>gene394 complete\n",
      "gorilla_>gene395 complete\n",
      "gorilla_>gene396 complete\n",
      "gorilla_>gene397 complete\n",
      "gorilla_>gene398 complete\n",
      "gorilla_>gene399 complete\n",
      "gorilla_>gene400 complete\n",
      "gorilla_>gene401 complete\n",
      "gorilla_>gene402 complete\n",
      "gorilla_>gene403 complete\n",
      "gorilla_>gene404 complete\n",
      "gorilla_>gene405 complete\n",
      "gorilla_>gene406 complete\n",
      "gorilla_>gene407 complete\n",
      "gorilla_>gene408 complete\n",
      "gorilla_>gene409 complete\n",
      "gorilla_>gene410 complete\n",
      "gorilla_>gene411 complete\n",
      "gorilla_>gene412 complete\n",
      "gorilla_>gene413 complete\n",
      "gorilla_>gene414 complete\n",
      "gorilla_>gene415 complete\n",
      "gorilla_>gene416 complete\n",
      "gorilla_>gene417 complete\n",
      "gorilla_>gene418 complete\n",
      "gorilla_>gene419 complete\n",
      "gorilla_>gene420 complete\n",
      "gorilla_>gene421 complete\n",
      "gorilla_>gene422 complete\n",
      "gorilla_>gene423 complete\n",
      "gorilla_>gene424 complete\n",
      "gorilla_>gene425 complete\n",
      "gorilla_>gene426 complete\n",
      "gorilla_>gene427 complete\n",
      "gorilla_>gene428 complete\n",
      "gorilla_>gene429 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gorilla_>gene430 complete\n",
      "gorilla_>gene431 complete\n",
      "gorilla_>gene432 complete\n",
      "gorilla_>gene433 complete\n",
      "gorilla_>gene434 complete\n",
      "gorilla_>gene435 complete\n",
      "gorilla_>gene436 complete\n",
      "gorilla_>gene437 complete\n",
      "gorilla_>gene438 complete\n",
      "gorilla_>gene439 complete\n",
      "gorilla_>gene440 complete\n",
      "gorilla_>gene441 complete\n",
      "gorilla_>gene442 complete\n",
      "gorilla_>gene443 complete\n",
      "gorilla_>gene444 complete\n",
      "gorilla_>gene445 complete\n",
      "gorilla_>gene446 complete\n",
      "gorilla_>gene447 complete\n",
      "gorilla_>gene448 complete\n",
      "gorilla_>gene449 complete\n",
      "gorilla_>gene450 complete\n",
      "gorilla_>gene451 complete\n",
      "gorilla_>gene452 complete\n",
      "gorilla_>gene453 complete\n",
      "gorilla_>gene454 complete\n",
      "gorilla_>gene455 complete\n",
      "gorilla_>gene456 complete\n",
      "gorilla_>gene457 complete\n",
      "gorilla_>gene458 complete\n",
      "gorilla_>gene459 complete\n",
      "gorilla_>gene460 complete\n",
      "gorilla_>gene461 complete\n",
      "gorilla_>gene462 complete\n",
      "gorilla_>gene463 complete\n",
      "gorilla_>gene464 complete\n",
      "gorilla_>gene465 complete\n",
      "gorilla_>gene466 complete\n",
      "gorilla_>gene467 complete\n",
      "gorilla_>gene468 complete\n",
      "gorilla_>gene469 complete\n",
      "gorilla_>gene470 complete\n",
      "gorilla_>gene471 complete\n",
      "gorilla_>gene472 complete\n",
      "gorilla_>gene473 complete\n",
      "gorilla_>gene474 complete\n",
      "gorilla_>gene475 complete\n",
      "gorilla_>gene476 complete\n",
      "gorilla_>gene477 complete\n",
      "gorilla_>gene478 complete\n",
      "gorilla_>gene479 complete\n",
      "gorilla_>gene480 complete\n",
      "gorilla_>gene481 complete\n",
      "gorilla_>gene482 complete\n",
      "gorilla_>gene483 complete\n",
      "gorilla_>gene484 complete\n",
      "gorilla_>gene485 complete\n",
      "gorilla_>gene486 complete\n",
      "gorilla_>gene487 complete\n",
      "gorilla_>gene488 complete\n",
      "gorilla_>gene489 complete\n",
      "gorilla_>gene490 complete\n",
      "gorilla_>gene491 complete\n",
      "gorilla_>gene492 complete\n",
      "gorilla_>gene493 complete\n",
      "gorilla_>gene494 complete\n",
      "gorilla_>gene495 complete\n",
      "gorilla_>gene496 complete\n",
      "gorilla_>gene497 complete\n",
      "gorilla_>gene498 complete\n",
      "gorilla_>gene499 complete\n",
      "gorilla_>gene500 complete\n",
      "fox_>gene1 complete\n",
      "fox_>gene2 complete\n",
      "fox_>gene3 complete\n",
      "fox_>gene4 complete\n",
      "fox_>gene5 complete\n",
      "fox_>gene6 complete\n",
      "fox_>gene7 complete\n",
      "fox_>gene8 complete\n",
      "fox_>gene9 complete\n",
      "fox_>gene10 complete\n",
      "fox_>gene11 complete\n",
      "fox_>gene12 complete\n",
      "fox_>gene13 complete\n",
      "fox_>gene14 complete\n",
      "fox_>gene15 complete\n",
      "fox_>gene16 complete\n",
      "fox_>gene17 complete\n",
      "fox_>gene18 complete\n",
      "fox_>gene19 complete\n",
      "fox_>gene20 complete\n",
      "fox_>gene21 complete\n",
      "fox_>gene22 complete\n",
      "fox_>gene23 complete\n",
      "fox_>gene24 complete\n",
      "fox_>gene25 complete\n",
      "fox_>gene26 complete\n",
      "fox_>gene27 complete\n",
      "fox_>gene28 complete\n",
      "fox_>gene29 complete\n",
      "fox_>gene30 complete\n",
      "fox_>gene31 complete\n",
      "fox_>gene32 complete\n",
      "fox_>gene33 complete\n",
      "fox_>gene34 complete\n",
      "fox_>gene35 complete\n",
      "fox_>gene36 complete\n",
      "fox_>gene37 complete\n",
      "fox_>gene38 complete\n",
      "fox_>gene39 complete\n",
      "fox_>gene40 complete\n",
      "fox_>gene41 complete\n",
      "fox_>gene42 complete\n",
      "fox_>gene43 complete\n",
      "fox_>gene44 complete\n",
      "fox_>gene45 complete\n",
      "fox_>gene46 complete\n",
      "fox_>gene47 complete\n",
      "fox_>gene48 complete\n",
      "fox_>gene49 complete\n",
      "fox_>gene50 complete\n",
      "fox_>gene51 complete\n",
      "fox_>gene52 complete\n",
      "fox_>gene53 complete\n",
      "fox_>gene54 complete\n",
      "fox_>gene55 complete\n",
      "fox_>gene56 complete\n",
      "fox_>gene57 complete\n",
      "fox_>gene58 complete\n",
      "fox_>gene59 complete\n",
      "fox_>gene60 complete\n",
      "fox_>gene61 complete\n",
      "fox_>gene62 complete\n",
      "fox_>gene63 complete\n",
      "fox_>gene64 complete\n",
      "fox_>gene65 complete\n",
      "fox_>gene66 complete\n",
      "fox_>gene67 complete\n",
      "fox_>gene68 complete\n",
      "fox_>gene69 complete\n",
      "fox_>gene70 complete\n",
      "fox_>gene71 complete\n",
      "fox_>gene72 complete\n",
      "fox_>gene73 complete\n",
      "fox_>gene74 complete\n",
      "fox_>gene75 complete\n",
      "fox_>gene76 complete\n",
      "fox_>gene77 complete\n",
      "fox_>gene78 complete\n",
      "fox_>gene79 complete\n",
      "fox_>gene80 complete\n",
      "fox_>gene81 complete\n",
      "fox_>gene82 complete\n",
      "fox_>gene83 complete\n",
      "fox_>gene84 complete\n",
      "fox_>gene85 complete\n",
      "fox_>gene86 complete\n",
      "fox_>gene87 complete\n",
      "fox_>gene88 complete\n",
      "fox_>gene89 complete\n",
      "fox_>gene90 complete\n",
      "fox_>gene91 complete\n",
      "fox_>gene92 complete\n",
      "fox_>gene93 complete\n",
      "fox_>gene94 complete\n",
      "fox_>gene95 complete\n",
      "fox_>gene96 complete\n",
      "fox_>gene97 complete\n",
      "fox_>gene98 complete\n",
      "fox_>gene99 complete\n",
      "fox_>gene100 complete\n",
      "fox_>gene101 complete\n",
      "fox_>gene102 complete\n",
      "fox_>gene103 complete\n",
      "fox_>gene104 complete\n",
      "fox_>gene105 complete\n",
      "fox_>gene106 complete\n",
      "fox_>gene107 complete\n",
      "fox_>gene108 complete\n",
      "fox_>gene109 complete\n",
      "fox_>gene110 complete\n",
      "fox_>gene111 complete\n",
      "fox_>gene112 complete\n",
      "fox_>gene113 complete\n",
      "fox_>gene114 complete\n",
      "fox_>gene115 complete\n",
      "fox_>gene116 complete\n",
      "fox_>gene117 complete\n",
      "fox_>gene118 complete\n",
      "fox_>gene119 complete\n",
      "fox_>gene120 complete\n",
      "fox_>gene121 complete\n",
      "fox_>gene122 complete\n",
      "fox_>gene123 complete\n",
      "fox_>gene124 complete\n",
      "fox_>gene125 complete\n",
      "fox_>gene126 complete\n",
      "fox_>gene127 complete\n",
      "fox_>gene128 complete\n",
      "fox_>gene129 complete\n",
      "fox_>gene130 complete\n",
      "fox_>gene131 complete\n",
      "fox_>gene132 complete\n",
      "fox_>gene133 complete\n",
      "fox_>gene134 complete\n",
      "fox_>gene135 complete\n",
      "fox_>gene136 complete\n",
      "fox_>gene137 complete\n",
      "fox_>gene138 complete\n",
      "fox_>gene139 complete\n",
      "fox_>gene140 complete\n",
      "fox_>gene141 complete\n",
      "fox_>gene142 complete\n",
      "fox_>gene143 complete\n",
      "fox_>gene144 complete\n",
      "fox_>gene145 complete\n",
      "fox_>gene146 complete\n",
      "fox_>gene147 complete\n",
      "fox_>gene148 complete\n",
      "fox_>gene149 complete\n",
      "fox_>gene150 complete\n",
      "fox_>gene151 complete\n",
      "fox_>gene152 complete\n",
      "fox_>gene153 complete\n",
      "fox_>gene154 complete\n",
      "fox_>gene155 complete\n",
      "fox_>gene156 complete\n",
      "fox_>gene157 complete\n",
      "fox_>gene158 complete\n",
      "fox_>gene159 complete\n",
      "fox_>gene160 complete\n",
      "fox_>gene161 complete\n",
      "fox_>gene162 complete\n",
      "fox_>gene163 complete\n",
      "fox_>gene164 complete\n",
      "fox_>gene165 complete\n",
      "fox_>gene166 complete\n",
      "fox_>gene167 complete\n",
      "fox_>gene168 complete\n",
      "fox_>gene169 complete\n",
      "fox_>gene170 complete\n",
      "fox_>gene171 complete\n",
      "fox_>gene172 complete\n",
      "fox_>gene173 complete\n",
      "fox_>gene174 complete\n",
      "fox_>gene175 complete\n",
      "fox_>gene176 complete\n",
      "fox_>gene177 complete\n",
      "fox_>gene178 complete\n",
      "fox_>gene179 complete\n",
      "fox_>gene180 complete\n",
      "fox_>gene181 complete\n",
      "fox_>gene182 complete\n",
      "fox_>gene183 complete\n",
      "fox_>gene184 complete\n",
      "fox_>gene185 complete\n",
      "fox_>gene186 complete\n",
      "fox_>gene187 complete\n",
      "fox_>gene188 complete\n",
      "fox_>gene189 complete\n",
      "fox_>gene190 complete\n",
      "fox_>gene191 complete\n",
      "fox_>gene192 complete\n",
      "fox_>gene193 complete\n",
      "fox_>gene194 complete\n",
      "fox_>gene195 complete\n",
      "fox_>gene196 complete\n",
      "fox_>gene197 complete\n",
      "fox_>gene198 complete\n",
      "fox_>gene199 complete\n",
      "fox_>gene200 complete\n",
      "fox_>gene201 complete\n",
      "fox_>gene202 complete\n",
      "fox_>gene203 complete\n",
      "fox_>gene204 complete\n",
      "fox_>gene205 complete\n",
      "fox_>gene206 complete\n",
      "fox_>gene207 complete\n",
      "fox_>gene208 complete\n",
      "fox_>gene209 complete\n",
      "fox_>gene210 complete\n",
      "fox_>gene211 complete\n",
      "fox_>gene212 complete\n",
      "fox_>gene213 complete\n",
      "fox_>gene214 complete\n",
      "fox_>gene215 complete\n",
      "fox_>gene216 complete\n",
      "fox_>gene217 complete\n",
      "fox_>gene218 complete\n",
      "fox_>gene219 complete\n",
      "fox_>gene220 complete\n",
      "fox_>gene221 complete\n",
      "fox_>gene222 complete\n",
      "fox_>gene223 complete\n",
      "fox_>gene224 complete\n",
      "fox_>gene225 complete\n",
      "fox_>gene226 complete\n",
      "fox_>gene227 complete\n",
      "fox_>gene228 complete\n",
      "fox_>gene229 complete\n",
      "fox_>gene230 complete\n",
      "fox_>gene231 complete\n",
      "fox_>gene232 complete\n",
      "fox_>gene233 complete\n",
      "fox_>gene234 complete\n",
      "fox_>gene235 complete\n",
      "fox_>gene236 complete\n",
      "fox_>gene237 complete\n",
      "fox_>gene238 complete\n",
      "fox_>gene239 complete\n",
      "fox_>gene240 complete\n",
      "fox_>gene241 complete\n",
      "fox_>gene242 complete\n",
      "fox_>gene243 complete\n",
      "fox_>gene244 complete\n",
      "fox_>gene245 complete\n",
      "fox_>gene246 complete\n",
      "fox_>gene247 complete\n",
      "fox_>gene248 complete\n",
      "fox_>gene249 complete\n",
      "fox_>gene250 complete\n",
      "fox_>gene251 complete\n",
      "fox_>gene252 complete\n",
      "fox_>gene253 complete\n",
      "fox_>gene254 complete\n",
      "fox_>gene255 complete\n",
      "fox_>gene256 complete\n",
      "fox_>gene257 complete\n",
      "fox_>gene258 complete\n",
      "fox_>gene259 complete\n",
      "fox_>gene260 complete\n",
      "fox_>gene261 complete\n",
      "fox_>gene262 complete\n",
      "fox_>gene263 complete\n",
      "fox_>gene264 complete\n",
      "fox_>gene265 complete\n",
      "fox_>gene266 complete\n",
      "fox_>gene267 complete\n",
      "fox_>gene268 complete\n",
      "fox_>gene269 complete\n",
      "fox_>gene270 complete\n",
      "fox_>gene271 complete\n",
      "fox_>gene272 complete\n",
      "fox_>gene273 complete\n",
      "fox_>gene274 complete\n",
      "fox_>gene275 complete\n",
      "fox_>gene276 complete\n",
      "fox_>gene277 complete\n",
      "fox_>gene278 complete\n",
      "fox_>gene279 complete\n",
      "fox_>gene280 complete\n",
      "fox_>gene281 complete\n",
      "fox_>gene282 complete\n",
      "fox_>gene283 complete\n",
      "fox_>gene284 complete\n",
      "fox_>gene285 complete\n",
      "fox_>gene286 complete\n",
      "fox_>gene287 complete\n",
      "fox_>gene288 complete\n",
      "fox_>gene289 complete\n",
      "fox_>gene290 complete\n",
      "fox_>gene291 complete\n",
      "fox_>gene292 complete\n",
      "fox_>gene293 complete\n",
      "fox_>gene294 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fox_>gene295 complete\n",
      "fox_>gene296 complete\n",
      "fox_>gene297 complete\n",
      "fox_>gene298 complete\n",
      "fox_>gene299 complete\n",
      "fox_>gene300 complete\n",
      "fox_>gene301 complete\n",
      "fox_>gene302 complete\n",
      "fox_>gene303 complete\n",
      "fox_>gene304 complete\n",
      "fox_>gene305 complete\n",
      "fox_>gene306 complete\n",
      "fox_>gene307 complete\n",
      "fox_>gene308 complete\n",
      "fox_>gene309 complete\n",
      "fox_>gene310 complete\n",
      "fox_>gene311 complete\n",
      "fox_>gene312 complete\n",
      "fox_>gene313 complete\n",
      "fox_>gene314 complete\n",
      "fox_>gene315 complete\n",
      "fox_>gene316 complete\n",
      "fox_>gene317 complete\n",
      "fox_>gene318 complete\n",
      "fox_>gene319 complete\n",
      "fox_>gene320 complete\n",
      "fox_>gene321 complete\n",
      "fox_>gene322 complete\n",
      "fox_>gene323 complete\n",
      "fox_>gene324 complete\n",
      "fox_>gene325 complete\n",
      "fox_>gene326 complete\n",
      "fox_>gene327 complete\n",
      "fox_>gene328 complete\n",
      "fox_>gene329 complete\n",
      "fox_>gene330 complete\n",
      "fox_>gene331 complete\n",
      "fox_>gene332 complete\n",
      "fox_>gene333 complete\n",
      "fox_>gene334 complete\n",
      "fox_>gene335 complete\n",
      "fox_>gene336 complete\n",
      "fox_>gene337 complete\n",
      "fox_>gene338 complete\n",
      "fox_>gene339 complete\n",
      "fox_>gene340 complete\n",
      "fox_>gene341 complete\n",
      "fox_>gene342 complete\n",
      "fox_>gene343 complete\n",
      "fox_>gene344 complete\n",
      "fox_>gene345 complete\n",
      "fox_>gene346 complete\n",
      "fox_>gene347 complete\n",
      "fox_>gene348 complete\n",
      "fox_>gene349 complete\n",
      "fox_>gene350 complete\n",
      "fox_>gene351 complete\n",
      "fox_>gene352 complete\n",
      "fox_>gene353 complete\n",
      "fox_>gene354 complete\n",
      "fox_>gene355 complete\n",
      "fox_>gene356 complete\n",
      "fox_>gene357 complete\n",
      "fox_>gene358 complete\n",
      "fox_>gene359 complete\n",
      "fox_>gene360 complete\n",
      "fox_>gene361 complete\n",
      "fox_>gene362 complete\n",
      "fox_>gene363 complete\n",
      "fox_>gene364 complete\n",
      "fox_>gene365 complete\n",
      "fox_>gene366 complete\n",
      "fox_>gene367 complete\n",
      "fox_>gene368 complete\n",
      "fox_>gene369 complete\n",
      "fox_>gene370 complete\n",
      "fox_>gene371 complete\n",
      "fox_>gene372 complete\n",
      "fox_>gene373 complete\n",
      "fox_>gene374 complete\n",
      "fox_>gene375 complete\n",
      "fox_>gene376 complete\n",
      "fox_>gene377 complete\n",
      "fox_>gene378 complete\n",
      "fox_>gene379 complete\n",
      "fox_>gene380 complete\n",
      "fox_>gene381 complete\n",
      "fox_>gene382 complete\n",
      "fox_>gene383 complete\n",
      "fox_>gene384 complete\n",
      "fox_>gene385 complete\n",
      "fox_>gene386 complete\n",
      "fox_>gene387 complete\n",
      "fox_>gene388 complete\n",
      "fox_>gene389 complete\n",
      "fox_>gene390 complete\n",
      "fox_>gene391 complete\n",
      "fox_>gene392 complete\n",
      "fox_>gene393 complete\n",
      "fox_>gene394 complete\n",
      "fox_>gene395 complete\n",
      "fox_>gene396 complete\n",
      "fox_>gene397 complete\n",
      "fox_>gene398 complete\n",
      "fox_>gene399 complete\n",
      "fox_>gene400 complete\n",
      "fox_>gene401 complete\n",
      "fox_>gene402 complete\n",
      "fox_>gene403 complete\n",
      "fox_>gene404 complete\n",
      "fox_>gene405 complete\n",
      "fox_>gene406 complete\n",
      "fox_>gene407 complete\n",
      "fox_>gene408 complete\n",
      "fox_>gene409 complete\n",
      "fox_>gene410 complete\n",
      "fox_>gene411 complete\n",
      "fox_>gene412 complete\n",
      "fox_>gene413 complete\n",
      "fox_>gene414 complete\n",
      "fox_>gene415 complete\n",
      "fox_>gene416 complete\n",
      "fox_>gene417 complete\n",
      "fox_>gene418 complete\n",
      "fox_>gene419 complete\n",
      "fox_>gene420 complete\n",
      "fox_>gene421 complete\n",
      "fox_>gene422 complete\n",
      "fox_>gene423 complete\n",
      "fox_>gene424 complete\n",
      "fox_>gene425 complete\n",
      "fox_>gene426 complete\n",
      "fox_>gene427 complete\n",
      "fox_>gene428 complete\n",
      "fox_>gene429 complete\n",
      "fox_>gene430 complete\n",
      "fox_>gene431 complete\n",
      "fox_>gene432 complete\n",
      "fox_>gene433 complete\n",
      "fox_>gene434 complete\n",
      "fox_>gene435 complete\n",
      "fox_>gene436 complete\n",
      "fox_>gene437 complete\n",
      "fox_>gene438 complete\n",
      "fox_>gene439 complete\n",
      "fox_>gene440 complete\n",
      "fox_>gene441 complete\n",
      "fox_>gene442 complete\n",
      "fox_>gene443 complete\n",
      "fox_>gene444 complete\n",
      "fox_>gene445 complete\n",
      "fox_>gene446 complete\n",
      "fox_>gene447 complete\n",
      "fox_>gene448 complete\n",
      "fox_>gene449 complete\n",
      "fox_>gene450 complete\n",
      "fox_>gene451 complete\n",
      "fox_>gene452 complete\n",
      "fox_>gene453 complete\n",
      "fox_>gene454 complete\n",
      "fox_>gene455 complete\n",
      "fox_>gene456 complete\n",
      "fox_>gene457 complete\n",
      "fox_>gene458 complete\n",
      "fox_>gene459 complete\n",
      "fox_>gene460 complete\n",
      "fox_>gene461 complete\n",
      "fox_>gene462 complete\n",
      "fox_>gene463 complete\n",
      "fox_>gene464 complete\n",
      "fox_>gene465 complete\n",
      "fox_>gene466 complete\n",
      "fox_>gene467 complete\n",
      "fox_>gene468 complete\n",
      "fox_>gene469 complete\n",
      "fox_>gene470 complete\n",
      "fox_>gene471 complete\n",
      "fox_>gene472 complete\n",
      "fox_>gene473 complete\n",
      "fox_>gene474 complete\n",
      "fox_>gene475 complete\n",
      "fox_>gene476 complete\n",
      "fox_>gene477 complete\n",
      "fox_>gene478 complete\n",
      "fox_>gene479 complete\n",
      "fox_>gene480 complete\n",
      "fox_>gene481 complete\n",
      "fox_>gene482 complete\n",
      "fox_>gene483 complete\n",
      "fox_>gene484 complete\n",
      "fox_>gene485 complete\n",
      "fox_>gene486 complete\n",
      "fox_>gene487 complete\n",
      "fox_>gene488 complete\n",
      "fox_>gene489 complete\n",
      "fox_>gene490 complete\n",
      "fox_>gene491 complete\n",
      "fox_>gene492 complete\n",
      "fox_>gene493 complete\n",
      "fox_>gene494 complete\n",
      "fox_>gene495 complete\n",
      "fox_>gene496 complete\n",
      "fox_>gene497 complete\n",
      "fox_>gene498 complete\n",
      "fox_>gene499 complete\n",
      "fox_>gene500 complete\n",
      "chicken_>gene1 complete\n",
      "chicken_>gene2 complete\n",
      "chicken_>gene3 complete\n",
      "chicken_>gene4 complete\n",
      "chicken_>gene5 complete\n",
      "chicken_>gene6 complete\n",
      "chicken_>gene7 complete\n",
      "chicken_>gene8 complete\n",
      "chicken_>gene9 complete\n",
      "chicken_>gene10 complete\n",
      "chicken_>gene11 complete\n",
      "chicken_>gene12 complete\n",
      "chicken_>gene13 complete\n",
      "chicken_>gene14 complete\n",
      "chicken_>gene15 complete\n",
      "chicken_>gene16 complete\n",
      "chicken_>gene17 complete\n",
      "chicken_>gene18 complete\n",
      "chicken_>gene19 complete\n",
      "chicken_>gene20 complete\n",
      "chicken_>gene21 complete\n",
      "chicken_>gene22 complete\n",
      "chicken_>gene23 complete\n",
      "chicken_>gene24 complete\n",
      "chicken_>gene25 complete\n",
      "chicken_>gene26 complete\n",
      "chicken_>gene27 complete\n",
      "chicken_>gene28 complete\n",
      "chicken_>gene29 complete\n",
      "chicken_>gene30 complete\n",
      "chicken_>gene31 complete\n",
      "chicken_>gene32 complete\n",
      "chicken_>gene33 complete\n",
      "chicken_>gene34 complete\n",
      "chicken_>gene35 complete\n",
      "chicken_>gene36 complete\n",
      "chicken_>gene37 complete\n",
      "chicken_>gene38 complete\n",
      "chicken_>gene39 complete\n",
      "chicken_>gene40 complete\n",
      "chicken_>gene41 complete\n",
      "chicken_>gene42 complete\n",
      "chicken_>gene43 complete\n",
      "chicken_>gene44 complete\n",
      "chicken_>gene45 complete\n",
      "chicken_>gene46 complete\n",
      "chicken_>gene47 complete\n",
      "chicken_>gene48 complete\n",
      "chicken_>gene49 complete\n",
      "chicken_>gene50 complete\n",
      "chicken_>gene51 complete\n",
      "chicken_>gene52 complete\n",
      "chicken_>gene53 complete\n",
      "chicken_>gene54 complete\n",
      "chicken_>gene55 complete\n",
      "chicken_>gene56 complete\n",
      "chicken_>gene57 complete\n",
      "chicken_>gene58 complete\n",
      "chicken_>gene59 complete\n",
      "chicken_>gene60 complete\n",
      "chicken_>gene61 complete\n",
      "chicken_>gene62 complete\n",
      "chicken_>gene63 complete\n",
      "chicken_>gene64 complete\n",
      "chicken_>gene65 complete\n",
      "chicken_>gene66 complete\n",
      "chicken_>gene67 complete\n",
      "chicken_>gene68 complete\n",
      "chicken_>gene69 complete\n",
      "chicken_>gene70 complete\n",
      "chicken_>gene71 complete\n",
      "chicken_>gene72 complete\n",
      "chicken_>gene73 complete\n",
      "chicken_>gene74 complete\n",
      "chicken_>gene75 complete\n",
      "chicken_>gene76 complete\n",
      "chicken_>gene77 complete\n",
      "chicken_>gene78 complete\n",
      "chicken_>gene79 complete\n",
      "chicken_>gene80 complete\n",
      "chicken_>gene81 complete\n",
      "chicken_>gene82 complete\n",
      "chicken_>gene83 complete\n",
      "chicken_>gene84 complete\n",
      "chicken_>gene85 complete\n",
      "chicken_>gene86 complete\n",
      "chicken_>gene87 complete\n",
      "chicken_>gene88 complete\n",
      "chicken_>gene89 complete\n",
      "chicken_>gene90 complete\n",
      "chicken_>gene91 complete\n",
      "chicken_>gene92 complete\n",
      "chicken_>gene93 complete\n",
      "chicken_>gene94 complete\n",
      "chicken_>gene95 complete\n",
      "chicken_>gene96 complete\n",
      "chicken_>gene97 complete\n",
      "chicken_>gene98 complete\n",
      "chicken_>gene99 complete\n",
      "chicken_>gene100 complete\n",
      "chicken_>gene101 complete\n",
      "chicken_>gene102 complete\n",
      "chicken_>gene103 complete\n",
      "chicken_>gene104 complete\n",
      "chicken_>gene105 complete\n",
      "chicken_>gene106 complete\n",
      "chicken_>gene107 complete\n",
      "chicken_>gene108 complete\n",
      "chicken_>gene109 complete\n",
      "chicken_>gene110 complete\n",
      "chicken_>gene111 complete\n",
      "chicken_>gene112 complete\n",
      "chicken_>gene113 complete\n",
      "chicken_>gene114 complete\n",
      "chicken_>gene115 complete\n",
      "chicken_>gene116 complete\n",
      "chicken_>gene117 complete\n",
      "chicken_>gene118 complete\n",
      "chicken_>gene119 complete\n",
      "chicken_>gene120 complete\n",
      "chicken_>gene121 complete\n",
      "chicken_>gene122 complete\n",
      "chicken_>gene123 complete\n",
      "chicken_>gene124 complete\n",
      "chicken_>gene125 complete\n",
      "chicken_>gene126 complete\n",
      "chicken_>gene127 complete\n",
      "chicken_>gene128 complete\n",
      "chicken_>gene129 complete\n",
      "chicken_>gene130 complete\n",
      "chicken_>gene131 complete\n",
      "chicken_>gene132 complete\n",
      "chicken_>gene133 complete\n",
      "chicken_>gene134 complete\n",
      "chicken_>gene135 complete\n",
      "chicken_>gene136 complete\n",
      "chicken_>gene137 complete\n",
      "chicken_>gene138 complete\n",
      "chicken_>gene139 complete\n",
      "chicken_>gene140 complete\n",
      "chicken_>gene141 complete\n",
      "chicken_>gene142 complete\n",
      "chicken_>gene143 complete\n",
      "chicken_>gene144 complete\n",
      "chicken_>gene145 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chicken_>gene146 complete\n",
      "chicken_>gene147 complete\n",
      "chicken_>gene148 complete\n",
      "chicken_>gene149 complete\n",
      "chicken_>gene150 complete\n",
      "chicken_>gene151 complete\n",
      "chicken_>gene152 complete\n",
      "chicken_>gene153 complete\n",
      "chicken_>gene154 complete\n",
      "chicken_>gene155 complete\n",
      "chicken_>gene156 complete\n",
      "chicken_>gene157 complete\n",
      "chicken_>gene158 complete\n",
      "chicken_>gene159 complete\n",
      "chicken_>gene160 complete\n",
      "chicken_>gene161 complete\n",
      "chicken_>gene162 complete\n",
      "chicken_>gene163 complete\n",
      "chicken_>gene164 complete\n",
      "chicken_>gene165 complete\n",
      "chicken_>gene166 complete\n",
      "chicken_>gene167 complete\n",
      "chicken_>gene168 complete\n",
      "chicken_>gene169 complete\n",
      "chicken_>gene170 complete\n",
      "chicken_>gene171 complete\n",
      "chicken_>gene172 complete\n",
      "chicken_>gene173 complete\n",
      "chicken_>gene174 complete\n",
      "chicken_>gene175 complete\n",
      "chicken_>gene176 complete\n",
      "chicken_>gene177 complete\n",
      "chicken_>gene178 complete\n",
      "chicken_>gene179 complete\n",
      "chicken_>gene180 complete\n",
      "chicken_>gene181 complete\n",
      "chicken_>gene182 complete\n",
      "chicken_>gene183 complete\n",
      "chicken_>gene184 complete\n",
      "chicken_>gene185 complete\n",
      "chicken_>gene186 complete\n",
      "chicken_>gene187 complete\n",
      "chicken_>gene188 complete\n",
      "chicken_>gene189 complete\n",
      "chicken_>gene190 complete\n",
      "chicken_>gene191 complete\n",
      "chicken_>gene192 complete\n",
      "chicken_>gene193 complete\n",
      "chicken_>gene194 complete\n",
      "chicken_>gene195 complete\n",
      "chicken_>gene196 complete\n",
      "chicken_>gene197 complete\n",
      "chicken_>gene198 complete\n",
      "chicken_>gene199 complete\n",
      "chicken_>gene200 complete\n",
      "chicken_>gene201 complete\n",
      "chicken_>gene202 complete\n",
      "chicken_>gene203 complete\n",
      "chicken_>gene204 complete\n",
      "chicken_>gene205 complete\n",
      "chicken_>gene206 complete\n",
      "chicken_>gene207 complete\n",
      "chicken_>gene208 complete\n",
      "chicken_>gene209 complete\n",
      "chicken_>gene210 complete\n",
      "chicken_>gene211 complete\n",
      "chicken_>gene212 complete\n",
      "chicken_>gene213 complete\n",
      "chicken_>gene214 complete\n",
      "chicken_>gene215 complete\n",
      "chicken_>gene216 complete\n",
      "chicken_>gene217 complete\n",
      "chicken_>gene218 complete\n",
      "chicken_>gene219 complete\n",
      "chicken_>gene220 complete\n",
      "chicken_>gene221 complete\n",
      "chicken_>gene222 complete\n",
      "chicken_>gene223 complete\n",
      "chicken_>gene224 complete\n",
      "chicken_>gene225 complete\n",
      "chicken_>gene226 complete\n",
      "chicken_>gene227 complete\n",
      "chicken_>gene228 complete\n",
      "chicken_>gene229 complete\n",
      "chicken_>gene230 complete\n",
      "chicken_>gene231 complete\n",
      "chicken_>gene232 complete\n",
      "chicken_>gene233 complete\n",
      "chicken_>gene234 complete\n",
      "chicken_>gene235 complete\n",
      "chicken_>gene236 complete\n",
      "chicken_>gene237 complete\n",
      "chicken_>gene238 complete\n",
      "chicken_>gene239 complete\n",
      "chicken_>gene240 complete\n",
      "chicken_>gene241 complete\n",
      "chicken_>gene242 complete\n",
      "chicken_>gene243 complete\n",
      "chicken_>gene244 complete\n",
      "chicken_>gene245 complete\n",
      "chicken_>gene246 complete\n",
      "chicken_>gene247 complete\n",
      "chicken_>gene248 complete\n",
      "chicken_>gene249 complete\n",
      "chicken_>gene250 complete\n",
      "chicken_>gene251 complete\n",
      "chicken_>gene252 complete\n",
      "chicken_>gene253 complete\n",
      "chicken_>gene254 complete\n",
      "chicken_>gene255 complete\n",
      "chicken_>gene256 complete\n",
      "chicken_>gene257 complete\n",
      "chicken_>gene258 complete\n",
      "chicken_>gene259 complete\n",
      "chicken_>gene260 complete\n",
      "chicken_>gene261 complete\n",
      "chicken_>gene262 complete\n",
      "chicken_>gene263 complete\n",
      "chicken_>gene264 complete\n",
      "chicken_>gene265 complete\n",
      "chicken_>gene266 complete\n",
      "chicken_>gene267 complete\n",
      "chicken_>gene268 complete\n",
      "chicken_>gene269 complete\n",
      "chicken_>gene270 complete\n",
      "chicken_>gene271 complete\n",
      "chicken_>gene272 complete\n",
      "chicken_>gene273 complete\n",
      "chicken_>gene274 complete\n",
      "chicken_>gene275 complete\n",
      "chicken_>gene276 complete\n",
      "chicken_>gene277 complete\n",
      "chicken_>gene278 complete\n",
      "chicken_>gene279 complete\n",
      "chicken_>gene280 complete\n",
      "chicken_>gene281 complete\n",
      "chicken_>gene282 complete\n",
      "chicken_>gene283 complete\n",
      "chicken_>gene284 complete\n",
      "chicken_>gene285 complete\n",
      "chicken_>gene286 complete\n",
      "chicken_>gene287 complete\n",
      "chicken_>gene288 complete\n",
      "chicken_>gene289 complete\n",
      "chicken_>gene290 complete\n",
      "chicken_>gene291 complete\n",
      "chicken_>gene292 complete\n",
      "chicken_>gene293 complete\n",
      "chicken_>gene294 complete\n",
      "chicken_>gene295 complete\n",
      "chicken_>gene296 complete\n",
      "chicken_>gene297 complete\n",
      "chicken_>gene298 complete\n",
      "chicken_>gene299 complete\n",
      "chicken_>gene300 complete\n",
      "chicken_>gene301 complete\n",
      "chicken_>gene302 complete\n",
      "chicken_>gene303 complete\n",
      "chicken_>gene304 complete\n",
      "chicken_>gene305 complete\n",
      "chicken_>gene306 complete\n",
      "chicken_>gene307 complete\n",
      "chicken_>gene308 complete\n",
      "chicken_>gene309 complete\n",
      "chicken_>gene310 complete\n",
      "chicken_>gene311 complete\n",
      "chicken_>gene312 complete\n",
      "chicken_>gene313 complete\n",
      "chicken_>gene314 complete\n",
      "chicken_>gene315 complete\n",
      "chicken_>gene316 complete\n",
      "chicken_>gene317 complete\n",
      "chicken_>gene318 complete\n",
      "chicken_>gene319 complete\n",
      "chicken_>gene320 complete\n",
      "chicken_>gene321 complete\n",
      "chicken_>gene322 complete\n",
      "chicken_>gene323 complete\n",
      "chicken_>gene324 complete\n",
      "chicken_>gene325 complete\n",
      "chicken_>gene326 complete\n",
      "chicken_>gene327 complete\n",
      "chicken_>gene328 complete\n",
      "chicken_>gene329 complete\n",
      "chicken_>gene330 complete\n",
      "chicken_>gene331 complete\n",
      "chicken_>gene332 complete\n",
      "chicken_>gene333 complete\n",
      "chicken_>gene334 complete\n",
      "chicken_>gene335 complete\n",
      "chicken_>gene336 complete\n",
      "chicken_>gene337 complete\n",
      "chicken_>gene338 complete\n",
      "chicken_>gene339 complete\n",
      "chicken_>gene340 complete\n",
      "chicken_>gene341 complete\n",
      "chicken_>gene342 complete\n",
      "chicken_>gene343 complete\n",
      "chicken_>gene344 complete\n",
      "chicken_>gene345 complete\n",
      "chicken_>gene346 complete\n",
      "chicken_>gene347 complete\n",
      "chicken_>gene348 complete\n",
      "chicken_>gene349 complete\n",
      "chicken_>gene350 complete\n",
      "chicken_>gene351 complete\n",
      "chicken_>gene352 complete\n",
      "chicken_>gene353 complete\n",
      "chicken_>gene354 complete\n",
      "chicken_>gene355 complete\n",
      "chicken_>gene356 complete\n",
      "chicken_>gene357 complete\n",
      "chicken_>gene358 complete\n",
      "chicken_>gene359 complete\n",
      "chicken_>gene360 complete\n",
      "chicken_>gene361 complete\n",
      "chicken_>gene362 complete\n",
      "chicken_>gene363 complete\n",
      "chicken_>gene364 complete\n",
      "chicken_>gene365 complete\n",
      "chicken_>gene366 complete\n",
      "chicken_>gene367 complete\n",
      "chicken_>gene368 complete\n",
      "chicken_>gene369 complete\n",
      "chicken_>gene370 complete\n",
      "chicken_>gene371 complete\n",
      "chicken_>gene372 complete\n",
      "chicken_>gene373 complete\n",
      "chicken_>gene374 complete\n",
      "chicken_>gene375 complete\n",
      "chicken_>gene376 complete\n",
      "chicken_>gene377 complete\n",
      "chicken_>gene378 complete\n",
      "chicken_>gene379 complete\n",
      "chicken_>gene380 complete\n",
      "chicken_>gene381 complete\n",
      "chicken_>gene382 complete\n",
      "chicken_>gene383 complete\n",
      "chicken_>gene384 complete\n",
      "chicken_>gene385 complete\n",
      "chicken_>gene386 complete\n",
      "chicken_>gene387 complete\n",
      "chicken_>gene388 complete\n",
      "chicken_>gene389 complete\n",
      "chicken_>gene390 complete\n",
      "chicken_>gene391 complete\n",
      "chicken_>gene392 complete\n",
      "chicken_>gene393 complete\n",
      "chicken_>gene394 complete\n",
      "chicken_>gene395 complete\n",
      "chicken_>gene396 complete\n",
      "chicken_>gene397 complete\n",
      "chicken_>gene398 complete\n",
      "chicken_>gene399 complete\n",
      "chicken_>gene400 complete\n",
      "chicken_>gene401 complete\n",
      "chicken_>gene402 complete\n",
      "chicken_>gene403 complete\n",
      "chicken_>gene404 complete\n",
      "chicken_>gene405 complete\n",
      "chicken_>gene406 complete\n",
      "chicken_>gene407 complete\n",
      "chicken_>gene408 complete\n",
      "chicken_>gene409 complete\n",
      "chicken_>gene410 complete\n",
      "chicken_>gene411 complete\n",
      "chicken_>gene412 complete\n",
      "chicken_>gene413 complete\n",
      "chicken_>gene414 complete\n",
      "chicken_>gene415 complete\n",
      "chicken_>gene416 complete\n",
      "chicken_>gene417 complete\n",
      "chicken_>gene418 complete\n",
      "chicken_>gene419 complete\n",
      "chicken_>gene420 complete\n",
      "chicken_>gene421 complete\n",
      "chicken_>gene422 complete\n",
      "chicken_>gene423 complete\n",
      "chicken_>gene424 complete\n",
      "chicken_>gene425 complete\n",
      "chicken_>gene426 complete\n",
      "chicken_>gene427 complete\n",
      "chicken_>gene428 complete\n",
      "chicken_>gene429 complete\n",
      "chicken_>gene430 complete\n",
      "chicken_>gene431 complete\n",
      "chicken_>gene432 complete\n",
      "chicken_>gene433 complete\n",
      "chicken_>gene434 complete\n",
      "chicken_>gene435 complete\n",
      "chicken_>gene436 complete\n",
      "chicken_>gene437 complete\n",
      "chicken_>gene438 complete\n",
      "chicken_>gene439 complete\n",
      "chicken_>gene440 complete\n",
      "chicken_>gene441 complete\n",
      "chicken_>gene442 complete\n",
      "chicken_>gene443 complete\n",
      "chicken_>gene444 complete\n",
      "chicken_>gene445 complete\n",
      "chicken_>gene446 complete\n",
      "chicken_>gene447 complete\n",
      "chicken_>gene448 complete\n",
      "chicken_>gene449 complete\n",
      "chicken_>gene450 complete\n",
      "chicken_>gene451 complete\n",
      "chicken_>gene452 complete\n",
      "chicken_>gene453 complete\n",
      "chicken_>gene454 complete\n",
      "chicken_>gene455 complete\n",
      "chicken_>gene456 complete\n",
      "chicken_>gene457 complete\n",
      "chicken_>gene458 complete\n",
      "chicken_>gene459 complete\n",
      "chicken_>gene460 complete\n",
      "chicken_>gene461 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chicken_>gene462 complete\n",
      "chicken_>gene463 complete\n",
      "chicken_>gene464 complete\n",
      "chicken_>gene465 complete\n",
      "chicken_>gene466 complete\n",
      "chicken_>gene467 complete\n",
      "chicken_>gene468 complete\n",
      "chicken_>gene469 complete\n",
      "chicken_>gene470 complete\n",
      "chicken_>gene471 complete\n",
      "chicken_>gene472 complete\n",
      "chicken_>gene473 complete\n",
      "chicken_>gene474 complete\n",
      "chicken_>gene475 complete\n",
      "chicken_>gene476 complete\n",
      "chicken_>gene477 complete\n",
      "chicken_>gene478 complete\n",
      "chicken_>gene479 complete\n",
      "chicken_>gene480 complete\n",
      "chicken_>gene481 complete\n",
      "chicken_>gene482 complete\n",
      "chicken_>gene483 complete\n",
      "chicken_>gene484 complete\n",
      "chicken_>gene485 complete\n",
      "chicken_>gene486 complete\n",
      "chicken_>gene487 complete\n",
      "chicken_>gene488 complete\n",
      "chicken_>gene489 complete\n",
      "chicken_>gene490 complete\n",
      "chicken_>gene491 complete\n",
      "chicken_>gene492 complete\n",
      "chicken_>gene493 complete\n",
      "chicken_>gene494 complete\n",
      "chicken_>gene495 complete\n",
      "chicken_>gene496 complete\n",
      "chicken_>gene497 complete\n",
      "chicken_>gene498 complete\n",
      "chicken_>gene499 complete\n",
      "chicken_>gene500 complete\n",
      "goat_>gene1 complete\n",
      "goat_>gene2 complete\n",
      "goat_>gene3 complete\n",
      "goat_>gene4 complete\n",
      "goat_>gene5 complete\n",
      "goat_>gene6 complete\n",
      "goat_>gene7 complete\n",
      "goat_>gene8 complete\n",
      "goat_>gene9 complete\n",
      "goat_>gene10 complete\n",
      "goat_>gene11 complete\n",
      "goat_>gene12 complete\n",
      "goat_>gene13 complete\n",
      "goat_>gene14 complete\n",
      "goat_>gene15 complete\n",
      "goat_>gene16 complete\n",
      "goat_>gene17 complete\n",
      "goat_>gene18 complete\n",
      "goat_>gene19 complete\n",
      "goat_>gene20 complete\n",
      "goat_>gene21 complete\n",
      "goat_>gene22 complete\n",
      "goat_>gene23 complete\n",
      "goat_>gene24 complete\n",
      "goat_>gene25 complete\n",
      "goat_>gene26 complete\n",
      "goat_>gene27 complete\n",
      "goat_>gene28 complete\n",
      "goat_>gene29 complete\n",
      "goat_>gene30 complete\n",
      "goat_>gene31 complete\n",
      "goat_>gene32 complete\n",
      "goat_>gene33 complete\n",
      "goat_>gene34 complete\n",
      "goat_>gene35 complete\n",
      "goat_>gene36 complete\n",
      "goat_>gene37 complete\n",
      "goat_>gene38 complete\n",
      "goat_>gene39 complete\n",
      "goat_>gene40 complete\n",
      "goat_>gene41 complete\n",
      "goat_>gene42 complete\n",
      "goat_>gene43 complete\n",
      "goat_>gene44 complete\n",
      "goat_>gene45 complete\n",
      "goat_>gene46 complete\n",
      "goat_>gene47 complete\n",
      "goat_>gene48 complete\n",
      "goat_>gene49 complete\n",
      "goat_>gene50 complete\n",
      "goat_>gene51 complete\n",
      "goat_>gene52 complete\n",
      "goat_>gene53 complete\n",
      "goat_>gene54 complete\n",
      "goat_>gene55 complete\n",
      "goat_>gene56 complete\n",
      "goat_>gene57 complete\n",
      "goat_>gene58 complete\n",
      "goat_>gene59 complete\n",
      "goat_>gene60 complete\n",
      "goat_>gene61 complete\n",
      "goat_>gene62 complete\n",
      "goat_>gene63 complete\n",
      "goat_>gene64 complete\n",
      "goat_>gene65 complete\n",
      "goat_>gene66 complete\n",
      "goat_>gene67 complete\n",
      "goat_>gene68 complete\n",
      "goat_>gene69 complete\n",
      "goat_>gene70 complete\n",
      "goat_>gene71 complete\n",
      "goat_>gene72 complete\n",
      "goat_>gene73 complete\n",
      "goat_>gene74 complete\n",
      "goat_>gene75 complete\n",
      "goat_>gene76 complete\n",
      "goat_>gene77 complete\n",
      "goat_>gene78 complete\n",
      "goat_>gene79 complete\n",
      "goat_>gene80 complete\n",
      "goat_>gene81 complete\n",
      "goat_>gene82 complete\n",
      "goat_>gene83 complete\n",
      "goat_>gene84 complete\n",
      "goat_>gene85 complete\n",
      "goat_>gene86 complete\n",
      "goat_>gene87 complete\n",
      "goat_>gene88 complete\n",
      "goat_>gene89 complete\n",
      "goat_>gene90 complete\n",
      "goat_>gene91 complete\n",
      "goat_>gene92 complete\n",
      "goat_>gene93 complete\n",
      "goat_>gene94 complete\n",
      "goat_>gene95 complete\n",
      "goat_>gene96 complete\n",
      "goat_>gene97 complete\n",
      "goat_>gene98 complete\n",
      "goat_>gene99 complete\n",
      "goat_>gene100 complete\n",
      "goat_>gene101 complete\n",
      "goat_>gene102 complete\n",
      "goat_>gene103 complete\n",
      "goat_>gene104 complete\n",
      "goat_>gene105 complete\n",
      "goat_>gene106 complete\n",
      "goat_>gene107 complete\n",
      "goat_>gene108 complete\n",
      "goat_>gene109 complete\n",
      "goat_>gene110 complete\n",
      "goat_>gene111 complete\n",
      "goat_>gene112 complete\n",
      "goat_>gene113 complete\n",
      "goat_>gene114 complete\n",
      "goat_>gene115 complete\n",
      "goat_>gene116 complete\n",
      "goat_>gene117 complete\n",
      "goat_>gene118 complete\n",
      "goat_>gene119 complete\n",
      "goat_>gene120 complete\n",
      "goat_>gene121 complete\n",
      "goat_>gene122 complete\n",
      "goat_>gene123 complete\n",
      "goat_>gene124 complete\n",
      "goat_>gene125 complete\n",
      "goat_>gene126 complete\n",
      "goat_>gene127 complete\n",
      "goat_>gene128 complete\n",
      "goat_>gene129 complete\n",
      "goat_>gene130 complete\n",
      "goat_>gene131 complete\n",
      "goat_>gene132 complete\n",
      "goat_>gene133 complete\n",
      "goat_>gene134 complete\n",
      "goat_>gene135 complete\n",
      "goat_>gene136 complete\n",
      "goat_>gene137 complete\n",
      "goat_>gene138 complete\n",
      "goat_>gene139 complete\n",
      "goat_>gene140 complete\n",
      "goat_>gene141 complete\n",
      "goat_>gene142 complete\n",
      "goat_>gene143 complete\n",
      "goat_>gene144 complete\n",
      "goat_>gene145 complete\n",
      "goat_>gene146 complete\n",
      "goat_>gene147 complete\n",
      "goat_>gene148 complete\n",
      "goat_>gene149 complete\n",
      "goat_>gene150 complete\n",
      "goat_>gene151 complete\n",
      "goat_>gene152 complete\n",
      "goat_>gene153 complete\n",
      "goat_>gene154 complete\n",
      "goat_>gene155 complete\n",
      "goat_>gene156 complete\n",
      "goat_>gene157 complete\n",
      "goat_>gene158 complete\n",
      "goat_>gene159 complete\n",
      "goat_>gene160 complete\n",
      "goat_>gene161 complete\n",
      "goat_>gene162 complete\n",
      "goat_>gene163 complete\n",
      "goat_>gene164 complete\n",
      "goat_>gene165 complete\n",
      "goat_>gene166 complete\n",
      "goat_>gene167 complete\n",
      "goat_>gene168 complete\n",
      "goat_>gene169 complete\n",
      "goat_>gene170 complete\n",
      "goat_>gene171 complete\n",
      "goat_>gene172 complete\n",
      "goat_>gene173 complete\n",
      "goat_>gene174 complete\n",
      "goat_>gene175 complete\n",
      "goat_>gene176 complete\n",
      "goat_>gene177 complete\n",
      "goat_>gene178 complete\n",
      "goat_>gene179 complete\n",
      "goat_>gene180 complete\n",
      "goat_>gene181 complete\n",
      "goat_>gene182 complete\n",
      "goat_>gene183 complete\n",
      "goat_>gene184 complete\n",
      "goat_>gene185 complete\n",
      "goat_>gene186 complete\n",
      "goat_>gene187 complete\n",
      "goat_>gene188 complete\n",
      "goat_>gene189 complete\n",
      "goat_>gene190 complete\n",
      "goat_>gene191 complete\n",
      "goat_>gene192 complete\n",
      "goat_>gene193 complete\n",
      "goat_>gene194 complete\n",
      "goat_>gene195 complete\n",
      "goat_>gene196 complete\n",
      "goat_>gene197 complete\n",
      "goat_>gene198 complete\n",
      "goat_>gene199 complete\n",
      "goat_>gene200 complete\n",
      "goat_>gene201 complete\n",
      "goat_>gene202 complete\n",
      "goat_>gene203 complete\n",
      "goat_>gene204 complete\n",
      "goat_>gene205 complete\n",
      "goat_>gene206 complete\n",
      "goat_>gene207 complete\n",
      "goat_>gene208 complete\n",
      "goat_>gene209 complete\n",
      "goat_>gene210 complete\n",
      "goat_>gene211 complete\n",
      "goat_>gene212 complete\n",
      "goat_>gene213 complete\n",
      "goat_>gene214 complete\n",
      "goat_>gene215 complete\n",
      "goat_>gene216 complete\n",
      "goat_>gene217 complete\n",
      "goat_>gene218 complete\n",
      "goat_>gene219 complete\n",
      "goat_>gene220 complete\n",
      "goat_>gene221 complete\n",
      "goat_>gene222 complete\n",
      "goat_>gene223 complete\n",
      "goat_>gene224 complete\n",
      "goat_>gene225 complete\n",
      "goat_>gene226 complete\n",
      "goat_>gene227 complete\n",
      "goat_>gene228 complete\n",
      "goat_>gene229 complete\n",
      "goat_>gene230 complete\n",
      "goat_>gene231 complete\n",
      "goat_>gene232 complete\n",
      "goat_>gene233 complete\n",
      "goat_>gene234 complete\n",
      "goat_>gene235 complete\n",
      "goat_>gene236 complete\n",
      "goat_>gene237 complete\n",
      "goat_>gene238 complete\n",
      "goat_>gene239 complete\n",
      "goat_>gene240 complete\n",
      "goat_>gene241 complete\n",
      "goat_>gene242 complete\n",
      "goat_>gene243 complete\n",
      "goat_>gene244 complete\n",
      "goat_>gene245 complete\n",
      "goat_>gene246 complete\n",
      "goat_>gene247 complete\n",
      "goat_>gene248 complete\n",
      "goat_>gene249 complete\n",
      "goat_>gene250 complete\n",
      "goat_>gene251 complete\n",
      "goat_>gene252 complete\n",
      "goat_>gene253 complete\n",
      "goat_>gene254 complete\n",
      "goat_>gene255 complete\n",
      "goat_>gene256 complete\n",
      "goat_>gene257 complete\n",
      "goat_>gene258 complete\n",
      "goat_>gene259 complete\n",
      "goat_>gene260 complete\n",
      "goat_>gene261 complete\n",
      "goat_>gene262 complete\n",
      "goat_>gene263 complete\n",
      "goat_>gene264 complete\n",
      "goat_>gene265 complete\n",
      "goat_>gene266 complete\n",
      "goat_>gene267 complete\n",
      "goat_>gene268 complete\n",
      "goat_>gene269 complete\n",
      "goat_>gene270 complete\n",
      "goat_>gene271 complete\n",
      "goat_>gene272 complete\n",
      "goat_>gene273 complete\n",
      "goat_>gene274 complete\n",
      "goat_>gene275 complete\n",
      "goat_>gene276 complete\n",
      "goat_>gene277 complete\n",
      "goat_>gene278 complete\n",
      "goat_>gene279 complete\n",
      "goat_>gene280 complete\n",
      "goat_>gene281 complete\n",
      "goat_>gene282 complete\n",
      "goat_>gene283 complete\n",
      "goat_>gene284 complete\n",
      "goat_>gene285 complete\n",
      "goat_>gene286 complete\n",
      "goat_>gene287 complete\n",
      "goat_>gene288 complete\n",
      "goat_>gene289 complete\n",
      "goat_>gene290 complete\n",
      "goat_>gene291 complete\n",
      "goat_>gene292 complete\n",
      "goat_>gene293 complete\n",
      "goat_>gene294 complete\n",
      "goat_>gene295 complete\n",
      "goat_>gene296 complete\n",
      "goat_>gene297 complete\n",
      "goat_>gene298 complete\n",
      "goat_>gene299 complete\n",
      "goat_>gene300 complete\n",
      "goat_>gene301 complete\n",
      "goat_>gene302 complete\n",
      "goat_>gene303 complete\n",
      "goat_>gene304 complete\n",
      "goat_>gene305 complete\n",
      "goat_>gene306 complete\n",
      "goat_>gene307 complete\n",
      "goat_>gene308 complete\n",
      "goat_>gene309 complete\n",
      "goat_>gene310 complete\n",
      "goat_>gene311 complete\n",
      "goat_>gene312 complete\n",
      "goat_>gene313 complete\n",
      "goat_>gene314 complete\n",
      "goat_>gene315 complete\n",
      "goat_>gene316 complete\n",
      "goat_>gene317 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "goat_>gene318 complete\n",
      "goat_>gene319 complete\n",
      "goat_>gene320 complete\n",
      "goat_>gene321 complete\n",
      "goat_>gene322 complete\n",
      "goat_>gene323 complete\n",
      "goat_>gene324 complete\n",
      "goat_>gene325 complete\n",
      "goat_>gene326 complete\n",
      "goat_>gene327 complete\n",
      "goat_>gene328 complete\n",
      "goat_>gene329 complete\n",
      "goat_>gene330 complete\n",
      "goat_>gene331 complete\n",
      "goat_>gene332 complete\n",
      "goat_>gene333 complete\n",
      "goat_>gene334 complete\n",
      "goat_>gene335 complete\n",
      "goat_>gene336 complete\n",
      "goat_>gene337 complete\n",
      "goat_>gene338 complete\n",
      "goat_>gene339 complete\n",
      "goat_>gene340 complete\n",
      "goat_>gene341 complete\n",
      "goat_>gene342 complete\n",
      "goat_>gene343 complete\n",
      "goat_>gene344 complete\n",
      "goat_>gene345 complete\n",
      "goat_>gene346 complete\n",
      "goat_>gene347 complete\n",
      "goat_>gene348 complete\n",
      "goat_>gene349 complete\n",
      "goat_>gene350 complete\n",
      "goat_>gene351 complete\n",
      "goat_>gene352 complete\n",
      "goat_>gene353 complete\n",
      "goat_>gene354 complete\n",
      "goat_>gene355 complete\n",
      "goat_>gene356 complete\n",
      "goat_>gene357 complete\n",
      "goat_>gene358 complete\n",
      "goat_>gene359 complete\n",
      "goat_>gene360 complete\n",
      "goat_>gene361 complete\n",
      "goat_>gene362 complete\n",
      "goat_>gene363 complete\n",
      "goat_>gene364 complete\n",
      "goat_>gene365 complete\n",
      "goat_>gene366 complete\n",
      "goat_>gene367 complete\n",
      "goat_>gene368 complete\n",
      "goat_>gene369 complete\n",
      "goat_>gene370 complete\n",
      "goat_>gene371 complete\n",
      "goat_>gene372 complete\n",
      "goat_>gene373 complete\n",
      "goat_>gene374 complete\n",
      "goat_>gene375 complete\n",
      "goat_>gene376 complete\n",
      "goat_>gene377 complete\n",
      "goat_>gene378 complete\n",
      "goat_>gene379 complete\n",
      "goat_>gene380 complete\n",
      "goat_>gene381 complete\n",
      "goat_>gene382 complete\n",
      "goat_>gene383 complete\n",
      "goat_>gene384 complete\n",
      "goat_>gene385 complete\n",
      "goat_>gene386 complete\n",
      "goat_>gene387 complete\n",
      "goat_>gene388 complete\n",
      "goat_>gene389 complete\n",
      "goat_>gene390 complete\n",
      "goat_>gene391 complete\n",
      "goat_>gene392 complete\n",
      "goat_>gene393 complete\n",
      "goat_>gene394 complete\n",
      "goat_>gene395 complete\n",
      "goat_>gene396 complete\n",
      "goat_>gene397 complete\n",
      "goat_>gene398 complete\n",
      "goat_>gene399 complete\n",
      "goat_>gene400 complete\n",
      "goat_>gene401 complete\n",
      "goat_>gene402 complete\n",
      "goat_>gene403 complete\n",
      "goat_>gene404 complete\n",
      "goat_>gene405 complete\n",
      "goat_>gene406 complete\n",
      "goat_>gene407 complete\n",
      "goat_>gene408 complete\n",
      "goat_>gene409 complete\n",
      "goat_>gene410 complete\n",
      "goat_>gene411 complete\n",
      "goat_>gene412 complete\n",
      "goat_>gene413 complete\n",
      "goat_>gene414 complete\n",
      "goat_>gene415 complete\n",
      "goat_>gene416 complete\n",
      "goat_>gene417 complete\n",
      "goat_>gene418 complete\n",
      "goat_>gene419 complete\n",
      "goat_>gene420 complete\n",
      "goat_>gene421 complete\n",
      "goat_>gene422 complete\n",
      "goat_>gene423 complete\n",
      "goat_>gene424 complete\n",
      "goat_>gene425 complete\n",
      "goat_>gene426 complete\n",
      "goat_>gene427 complete\n",
      "goat_>gene428 complete\n",
      "goat_>gene429 complete\n",
      "goat_>gene430 complete\n",
      "goat_>gene431 complete\n",
      "goat_>gene432 complete\n",
      "goat_>gene433 complete\n",
      "goat_>gene434 complete\n",
      "goat_>gene435 complete\n",
      "goat_>gene436 complete\n",
      "goat_>gene437 complete\n",
      "goat_>gene438 complete\n",
      "goat_>gene439 complete\n",
      "goat_>gene440 complete\n",
      "goat_>gene441 complete\n",
      "goat_>gene442 complete\n",
      "goat_>gene443 complete\n",
      "goat_>gene444 complete\n",
      "goat_>gene445 complete\n",
      "goat_>gene446 complete\n",
      "goat_>gene447 complete\n",
      "goat_>gene448 complete\n",
      "goat_>gene449 complete\n",
      "goat_>gene450 complete\n",
      "goat_>gene451 complete\n",
      "goat_>gene452 complete\n",
      "goat_>gene453 complete\n",
      "goat_>gene454 complete\n",
      "goat_>gene455 complete\n",
      "goat_>gene456 complete\n",
      "goat_>gene457 complete\n",
      "goat_>gene458 complete\n",
      "goat_>gene459 complete\n",
      "goat_>gene460 complete\n",
      "goat_>gene461 complete\n",
      "goat_>gene462 complete\n",
      "goat_>gene463 complete\n",
      "goat_>gene464 complete\n",
      "goat_>gene465 complete\n",
      "goat_>gene466 complete\n",
      "goat_>gene467 complete\n",
      "goat_>gene468 complete\n",
      "goat_>gene469 complete\n",
      "goat_>gene470 complete\n",
      "goat_>gene471 complete\n",
      "goat_>gene472 complete\n",
      "goat_>gene473 complete\n",
      "goat_>gene474 complete\n",
      "goat_>gene475 complete\n",
      "goat_>gene476 complete\n",
      "goat_>gene477 complete\n",
      "goat_>gene478 complete\n",
      "goat_>gene479 complete\n",
      "goat_>gene480 complete\n",
      "goat_>gene481 complete\n",
      "goat_>gene482 complete\n",
      "goat_>gene483 complete\n",
      "goat_>gene484 complete\n",
      "goat_>gene485 complete\n",
      "goat_>gene486 complete\n",
      "goat_>gene487 complete\n",
      "goat_>gene488 complete\n",
      "goat_>gene489 complete\n",
      "goat_>gene490 complete\n",
      "goat_>gene491 complete\n",
      "goat_>gene492 complete\n",
      "goat_>gene493 complete\n",
      "goat_>gene494 complete\n",
      "goat_>gene495 complete\n",
      "goat_>gene496 complete\n",
      "goat_>gene497 complete\n",
      "goat_>gene498 complete\n",
      "goat_>gene499 complete\n",
      "goat_>gene500 complete\n",
      "gopher_>gene1 complete\n",
      "gopher_>gene2 complete\n",
      "gopher_>gene3 complete\n",
      "gopher_>gene4 complete\n",
      "gopher_>gene5 complete\n",
      "gopher_>gene6 complete\n",
      "gopher_>gene7 complete\n",
      "gopher_>gene8 complete\n",
      "gopher_>gene9 complete\n",
      "gopher_>gene10 complete\n",
      "gopher_>gene11 complete\n",
      "gopher_>gene12 complete\n",
      "gopher_>gene13 complete\n",
      "gopher_>gene14 complete\n",
      "gopher_>gene15 complete\n",
      "gopher_>gene16 complete\n",
      "gopher_>gene17 complete\n",
      "gopher_>gene18 complete\n",
      "gopher_>gene19 complete\n",
      "gopher_>gene20 complete\n",
      "gopher_>gene21 complete\n",
      "gopher_>gene22 complete\n",
      "gopher_>gene23 complete\n",
      "gopher_>gene24 complete\n",
      "gopher_>gene25 complete\n",
      "gopher_>gene26 complete\n",
      "gopher_>gene27 complete\n",
      "gopher_>gene28 complete\n",
      "gopher_>gene29 complete\n",
      "gopher_>gene30 complete\n",
      "gopher_>gene31 complete\n",
      "gopher_>gene32 complete\n",
      "gopher_>gene33 complete\n",
      "gopher_>gene34 complete\n",
      "gopher_>gene35 complete\n",
      "gopher_>gene36 complete\n",
      "gopher_>gene37 complete\n",
      "gopher_>gene38 complete\n",
      "gopher_>gene39 complete\n",
      "gopher_>gene40 complete\n",
      "gopher_>gene41 complete\n",
      "gopher_>gene42 complete\n",
      "gopher_>gene43 complete\n",
      "gopher_>gene44 complete\n",
      "gopher_>gene45 complete\n",
      "gopher_>gene46 complete\n",
      "gopher_>gene47 complete\n",
      "gopher_>gene48 complete\n",
      "gopher_>gene49 complete\n",
      "gopher_>gene50 complete\n",
      "gopher_>gene51 complete\n",
      "gopher_>gene52 complete\n",
      "gopher_>gene53 complete\n",
      "gopher_>gene54 complete\n",
      "gopher_>gene55 complete\n",
      "gopher_>gene56 complete\n",
      "gopher_>gene57 complete\n",
      "gopher_>gene58 complete\n",
      "gopher_>gene59 complete\n",
      "gopher_>gene60 complete\n",
      "gopher_>gene61 complete\n",
      "gopher_>gene62 complete\n",
      "gopher_>gene63 complete\n",
      "gopher_>gene64 complete\n",
      "gopher_>gene65 complete\n",
      "gopher_>gene66 complete\n",
      "gopher_>gene67 complete\n",
      "gopher_>gene68 complete\n",
      "gopher_>gene69 complete\n",
      "gopher_>gene70 complete\n",
      "gopher_>gene71 complete\n",
      "gopher_>gene72 complete\n",
      "gopher_>gene73 complete\n",
      "gopher_>gene74 complete\n",
      "gopher_>gene75 complete\n",
      "gopher_>gene76 complete\n",
      "gopher_>gene77 complete\n",
      "gopher_>gene78 complete\n",
      "gopher_>gene79 complete\n",
      "gopher_>gene80 complete\n",
      "gopher_>gene81 complete\n",
      "gopher_>gene82 complete\n",
      "gopher_>gene83 complete\n",
      "gopher_>gene84 complete\n",
      "gopher_>gene85 complete\n",
      "gopher_>gene86 complete\n",
      "gopher_>gene87 complete\n",
      "gopher_>gene88 complete\n",
      "gopher_>gene89 complete\n",
      "gopher_>gene90 complete\n",
      "gopher_>gene91 complete\n",
      "gopher_>gene92 complete\n",
      "gopher_>gene93 complete\n",
      "gopher_>gene94 complete\n",
      "gopher_>gene95 complete\n",
      "gopher_>gene96 complete\n",
      "gopher_>gene97 complete\n",
      "gopher_>gene98 complete\n",
      "gopher_>gene99 complete\n",
      "gopher_>gene100 complete\n",
      "gopher_>gene101 complete\n",
      "gopher_>gene102 complete\n",
      "gopher_>gene103 complete\n",
      "gopher_>gene104 complete\n",
      "gopher_>gene105 complete\n",
      "gopher_>gene106 complete\n",
      "gopher_>gene107 complete\n",
      "gopher_>gene108 complete\n",
      "gopher_>gene109 complete\n",
      "gopher_>gene110 complete\n",
      "gopher_>gene111 complete\n",
      "gopher_>gene112 complete\n",
      "gopher_>gene113 complete\n",
      "gopher_>gene114 complete\n",
      "gopher_>gene115 complete\n",
      "gopher_>gene116 complete\n",
      "gopher_>gene117 complete\n",
      "gopher_>gene118 complete\n",
      "gopher_>gene119 complete\n",
      "gopher_>gene120 complete\n",
      "gopher_>gene121 complete\n",
      "gopher_>gene122 complete\n",
      "gopher_>gene123 complete\n",
      "gopher_>gene124 complete\n",
      "gopher_>gene125 complete\n",
      "gopher_>gene126 complete\n",
      "gopher_>gene127 complete\n",
      "gopher_>gene128 complete\n",
      "gopher_>gene129 complete\n",
      "gopher_>gene130 complete\n",
      "gopher_>gene131 complete\n",
      "gopher_>gene132 complete\n",
      "gopher_>gene133 complete\n",
      "gopher_>gene134 complete\n",
      "gopher_>gene135 complete\n",
      "gopher_>gene136 complete\n",
      "gopher_>gene137 complete\n",
      "gopher_>gene138 complete\n",
      "gopher_>gene139 complete\n",
      "gopher_>gene140 complete\n",
      "gopher_>gene141 complete\n",
      "gopher_>gene142 complete\n",
      "gopher_>gene143 complete\n",
      "gopher_>gene144 complete\n",
      "gopher_>gene145 complete\n",
      "gopher_>gene146 complete\n",
      "gopher_>gene147 complete\n",
      "gopher_>gene148 complete\n",
      "gopher_>gene149 complete\n",
      "gopher_>gene150 complete\n",
      "gopher_>gene151 complete\n",
      "gopher_>gene152 complete\n",
      "gopher_>gene153 complete\n",
      "gopher_>gene154 complete\n",
      "gopher_>gene155 complete\n",
      "gopher_>gene156 complete\n",
      "gopher_>gene157 complete\n",
      "gopher_>gene158 complete\n",
      "gopher_>gene159 complete\n",
      "gopher_>gene160 complete\n",
      "gopher_>gene161 complete\n",
      "gopher_>gene162 complete\n",
      "gopher_>gene163 complete\n",
      "gopher_>gene164 complete\n",
      "gopher_>gene165 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gopher_>gene166 complete\n",
      "gopher_>gene167 complete\n",
      "gopher_>gene168 complete\n",
      "gopher_>gene169 complete\n",
      "gopher_>gene170 complete\n",
      "gopher_>gene171 complete\n",
      "gopher_>gene172 complete\n",
      "gopher_>gene173 complete\n",
      "gopher_>gene174 complete\n",
      "gopher_>gene175 complete\n",
      "gopher_>gene176 complete\n",
      "gopher_>gene177 complete\n",
      "gopher_>gene178 complete\n",
      "gopher_>gene179 complete\n",
      "gopher_>gene180 complete\n",
      "gopher_>gene181 complete\n",
      "gopher_>gene182 complete\n",
      "gopher_>gene183 complete\n",
      "gopher_>gene184 complete\n",
      "gopher_>gene185 complete\n",
      "gopher_>gene186 complete\n",
      "gopher_>gene187 complete\n",
      "gopher_>gene188 complete\n",
      "gopher_>gene189 complete\n",
      "gopher_>gene190 complete\n",
      "gopher_>gene191 complete\n",
      "gopher_>gene192 complete\n",
      "gopher_>gene193 complete\n",
      "gopher_>gene194 complete\n",
      "gopher_>gene195 complete\n",
      "gopher_>gene196 complete\n",
      "gopher_>gene197 complete\n",
      "gopher_>gene198 complete\n",
      "gopher_>gene199 complete\n",
      "gopher_>gene200 complete\n",
      "gopher_>gene201 complete\n",
      "gopher_>gene202 complete\n",
      "gopher_>gene203 complete\n",
      "gopher_>gene204 complete\n",
      "gopher_>gene205 complete\n",
      "gopher_>gene206 complete\n",
      "gopher_>gene207 complete\n",
      "gopher_>gene208 complete\n",
      "gopher_>gene209 complete\n",
      "gopher_>gene210 complete\n",
      "gopher_>gene211 complete\n",
      "gopher_>gene212 complete\n",
      "gopher_>gene213 complete\n",
      "gopher_>gene214 complete\n",
      "gopher_>gene215 complete\n",
      "gopher_>gene216 complete\n",
      "gopher_>gene217 complete\n",
      "gopher_>gene218 complete\n",
      "gopher_>gene219 complete\n",
      "gopher_>gene220 complete\n",
      "gopher_>gene221 complete\n",
      "gopher_>gene222 complete\n",
      "gopher_>gene223 complete\n",
      "gopher_>gene224 complete\n",
      "gopher_>gene225 complete\n",
      "gopher_>gene226 complete\n",
      "gopher_>gene227 complete\n",
      "gopher_>gene228 complete\n",
      "gopher_>gene229 complete\n",
      "gopher_>gene230 complete\n",
      "gopher_>gene231 complete\n",
      "gopher_>gene232 complete\n",
      "gopher_>gene233 complete\n",
      "gopher_>gene234 complete\n",
      "gopher_>gene235 complete\n",
      "gopher_>gene236 complete\n",
      "gopher_>gene237 complete\n",
      "gopher_>gene238 complete\n",
      "gopher_>gene239 complete\n",
      "gopher_>gene240 complete\n",
      "gopher_>gene241 complete\n",
      "gopher_>gene242 complete\n",
      "gopher_>gene243 complete\n",
      "gopher_>gene244 complete\n",
      "gopher_>gene245 complete\n",
      "gopher_>gene246 complete\n",
      "gopher_>gene247 complete\n",
      "gopher_>gene248 complete\n",
      "gopher_>gene249 complete\n",
      "gopher_>gene250 complete\n",
      "gopher_>gene251 complete\n",
      "gopher_>gene252 complete\n",
      "gopher_>gene253 complete\n",
      "gopher_>gene254 complete\n",
      "gopher_>gene255 complete\n",
      "gopher_>gene256 complete\n",
      "gopher_>gene257 complete\n",
      "gopher_>gene258 complete\n",
      "gopher_>gene259 complete\n",
      "gopher_>gene260 complete\n",
      "gopher_>gene261 complete\n",
      "gopher_>gene262 complete\n",
      "gopher_>gene263 complete\n",
      "gopher_>gene264 complete\n",
      "gopher_>gene265 complete\n",
      "gopher_>gene266 complete\n",
      "gopher_>gene267 complete\n",
      "gopher_>gene268 complete\n",
      "gopher_>gene269 complete\n",
      "gopher_>gene270 complete\n",
      "gopher_>gene271 complete\n",
      "gopher_>gene272 complete\n",
      "gopher_>gene273 complete\n",
      "gopher_>gene274 complete\n",
      "gopher_>gene275 complete\n",
      "gopher_>gene276 complete\n",
      "gopher_>gene277 complete\n",
      "gopher_>gene278 complete\n",
      "gopher_>gene279 complete\n",
      "gopher_>gene280 complete\n",
      "gopher_>gene281 complete\n",
      "gopher_>gene282 complete\n",
      "gopher_>gene283 complete\n",
      "gopher_>gene284 complete\n",
      "gopher_>gene285 complete\n",
      "gopher_>gene286 complete\n",
      "gopher_>gene287 complete\n",
      "gopher_>gene288 complete\n",
      "gopher_>gene289 complete\n",
      "gopher_>gene290 complete\n",
      "gopher_>gene291 complete\n",
      "gopher_>gene292 complete\n",
      "gopher_>gene293 complete\n",
      "gopher_>gene294 complete\n",
      "gopher_>gene295 complete\n",
      "gopher_>gene296 complete\n",
      "gopher_>gene297 complete\n",
      "gopher_>gene298 complete\n",
      "gopher_>gene299 complete\n",
      "gopher_>gene300 complete\n",
      "gopher_>gene301 complete\n",
      "gopher_>gene302 complete\n",
      "gopher_>gene303 complete\n",
      "gopher_>gene304 complete\n",
      "gopher_>gene305 complete\n",
      "gopher_>gene306 complete\n",
      "gopher_>gene307 complete\n",
      "gopher_>gene308 complete\n",
      "gopher_>gene309 complete\n",
      "gopher_>gene310 complete\n",
      "gopher_>gene311 complete\n",
      "gopher_>gene312 complete\n",
      "gopher_>gene313 complete\n",
      "gopher_>gene314 complete\n",
      "gopher_>gene315 complete\n",
      "gopher_>gene316 complete\n",
      "gopher_>gene317 complete\n",
      "gopher_>gene318 complete\n",
      "gopher_>gene319 complete\n",
      "gopher_>gene320 complete\n",
      "gopher_>gene321 complete\n",
      "gopher_>gene322 complete\n",
      "gopher_>gene323 complete\n",
      "gopher_>gene324 complete\n",
      "gopher_>gene325 complete\n",
      "gopher_>gene326 complete\n",
      "gopher_>gene327 complete\n",
      "gopher_>gene328 complete\n",
      "gopher_>gene329 complete\n",
      "gopher_>gene330 complete\n",
      "gopher_>gene331 complete\n",
      "gopher_>gene332 complete\n",
      "gopher_>gene333 complete\n",
      "gopher_>gene334 complete\n",
      "gopher_>gene335 complete\n",
      "gopher_>gene336 complete\n",
      "gopher_>gene337 complete\n",
      "gopher_>gene338 complete\n",
      "gopher_>gene339 complete\n",
      "gopher_>gene340 complete\n",
      "gopher_>gene341 complete\n",
      "gopher_>gene342 complete\n",
      "gopher_>gene343 complete\n",
      "gopher_>gene344 complete\n",
      "gopher_>gene345 complete\n",
      "gopher_>gene346 complete\n",
      "gopher_>gene347 complete\n",
      "gopher_>gene348 complete\n",
      "gopher_>gene349 complete\n",
      "gopher_>gene350 complete\n",
      "gopher_>gene351 complete\n",
      "gopher_>gene352 complete\n",
      "gopher_>gene353 complete\n",
      "gopher_>gene354 complete\n",
      "gopher_>gene355 complete\n",
      "gopher_>gene356 complete\n",
      "gopher_>gene357 complete\n",
      "gopher_>gene358 complete\n",
      "gopher_>gene359 complete\n",
      "gopher_>gene360 complete\n",
      "gopher_>gene361 complete\n",
      "gopher_>gene362 complete\n",
      "gopher_>gene363 complete\n",
      "gopher_>gene364 complete\n",
      "gopher_>gene365 complete\n",
      "gopher_>gene366 complete\n",
      "gopher_>gene367 complete\n",
      "gopher_>gene368 complete\n",
      "gopher_>gene369 complete\n",
      "gopher_>gene370 complete\n",
      "gopher_>gene371 complete\n",
      "gopher_>gene372 complete\n",
      "gopher_>gene373 complete\n",
      "gopher_>gene374 complete\n",
      "gopher_>gene375 complete\n",
      "gopher_>gene376 complete\n",
      "gopher_>gene377 complete\n",
      "gopher_>gene378 complete\n",
      "gopher_>gene379 complete\n",
      "gopher_>gene380 complete\n",
      "gopher_>gene381 complete\n",
      "gopher_>gene382 complete\n",
      "gopher_>gene383 complete\n",
      "gopher_>gene384 complete\n",
      "gopher_>gene385 complete\n",
      "gopher_>gene386 complete\n",
      "gopher_>gene387 complete\n",
      "gopher_>gene388 complete\n",
      "gopher_>gene389 complete\n",
      "gopher_>gene390 complete\n",
      "gopher_>gene391 complete\n",
      "gopher_>gene392 complete\n",
      "gopher_>gene393 complete\n",
      "gopher_>gene394 complete\n",
      "gopher_>gene395 complete\n",
      "gopher_>gene396 complete\n",
      "gopher_>gene397 complete\n",
      "gopher_>gene398 complete\n",
      "gopher_>gene399 complete\n",
      "gopher_>gene400 complete\n",
      "gopher_>gene401 complete\n",
      "gopher_>gene402 complete\n",
      "gopher_>gene403 complete\n",
      "gopher_>gene404 complete\n",
      "gopher_>gene405 complete\n",
      "gopher_>gene406 complete\n",
      "gopher_>gene407 complete\n",
      "gopher_>gene408 complete\n",
      "gopher_>gene409 complete\n",
      "gopher_>gene410 complete\n",
      "gopher_>gene411 complete\n",
      "gopher_>gene412 complete\n",
      "gopher_>gene413 complete\n",
      "gopher_>gene414 complete\n",
      "gopher_>gene415 complete\n",
      "gopher_>gene416 complete\n",
      "gopher_>gene417 complete\n",
      "gopher_>gene418 complete\n",
      "gopher_>gene419 complete\n",
      "gopher_>gene420 complete\n",
      "gopher_>gene421 complete\n",
      "gopher_>gene422 complete\n",
      "gopher_>gene423 complete\n",
      "gopher_>gene424 complete\n",
      "gopher_>gene425 complete\n",
      "gopher_>gene426 complete\n",
      "gopher_>gene427 complete\n",
      "gopher_>gene428 complete\n",
      "gopher_>gene429 complete\n",
      "gopher_>gene430 complete\n",
      "gopher_>gene431 complete\n",
      "gopher_>gene432 complete\n",
      "gopher_>gene433 complete\n",
      "gopher_>gene434 complete\n",
      "gopher_>gene435 complete\n",
      "gopher_>gene436 complete\n",
      "gopher_>gene437 complete\n",
      "gopher_>gene438 complete\n",
      "gopher_>gene439 complete\n",
      "gopher_>gene440 complete\n",
      "gopher_>gene441 complete\n",
      "gopher_>gene442 complete\n",
      "gopher_>gene443 complete\n",
      "gopher_>gene444 complete\n",
      "gopher_>gene445 complete\n",
      "gopher_>gene446 complete\n",
      "gopher_>gene447 complete\n",
      "gopher_>gene448 complete\n",
      "gopher_>gene449 complete\n",
      "gopher_>gene450 complete\n",
      "gopher_>gene451 complete\n",
      "gopher_>gene452 complete\n",
      "gopher_>gene453 complete\n",
      "gopher_>gene454 complete\n",
      "gopher_>gene455 complete\n",
      "gopher_>gene456 complete\n",
      "gopher_>gene457 complete\n",
      "gopher_>gene458 complete\n",
      "gopher_>gene459 complete\n",
      "gopher_>gene460 complete\n",
      "gopher_>gene461 complete\n",
      "gopher_>gene462 complete\n",
      "gopher_>gene463 complete\n",
      "gopher_>gene464 complete\n",
      "gopher_>gene465 complete\n",
      "gopher_>gene466 complete\n",
      "gopher_>gene467 complete\n",
      "gopher_>gene468 complete\n",
      "gopher_>gene469 complete\n",
      "gopher_>gene470 complete\n",
      "gopher_>gene471 complete\n",
      "gopher_>gene472 complete\n",
      "gopher_>gene473 complete\n",
      "gopher_>gene474 complete\n",
      "gopher_>gene475 complete\n",
      "gopher_>gene476 complete\n",
      "gopher_>gene477 complete\n",
      "gopher_>gene478 complete\n",
      "gopher_>gene479 complete\n",
      "gopher_>gene480 complete\n",
      "gopher_>gene481 complete\n",
      "gopher_>gene482 complete\n",
      "gopher_>gene483 complete\n",
      "gopher_>gene484 complete\n",
      "gopher_>gene485 complete\n",
      "gopher_>gene486 complete\n",
      "gopher_>gene487 complete\n",
      "gopher_>gene488 complete\n",
      "gopher_>gene489 complete\n",
      "gopher_>gene490 complete\n",
      "gopher_>gene491 complete\n",
      "gopher_>gene492 complete\n",
      "gopher_>gene493 complete\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gopher_>gene494 complete\n",
      "gopher_>gene495 complete\n",
      "gopher_>gene496 complete\n",
      "gopher_>gene497 complete\n",
      "gopher_>gene498 complete\n",
      "gopher_>gene499 complete\n",
      "gopher_>gene500 complete\n"
     ]
    }
   ],
   "source": [
    "# This is very inefficent, I sloppily ran this, \n",
    "# should have triangulaged first\n",
    "\n",
    "import json\n",
    "\n",
    "animal_class = \"\"\n",
    "square_dict = {}\n",
    "for animal_1 in distance_dict:\n",
    "    animal = animal_1.split(\"_\")[0]\n",
    "    if animal_class != animal:\n",
    "        animal_class = animal\n",
    "        with open('distance_{}.json'.format(animal), 'w') as fp:\n",
    "            json.dump(square_dict, fp)\n",
    "        square_dict = {}\n",
    "    square_dict[animal_1] = {}\n",
    "    for animal_2 in distance_dict:\n",
    "        square_dict[animal_1][animal_2] = lev.distance(distance_dict[animal_1], \n",
    "                                                       distance_dict[animal_2])\n",
    "    print(animal_1+\" complete\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "square_dict.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "square_dict = {}\n",
    "for animal_1 in distance_dict:\n",
    "    square_dict[animal_1] = {}\n",
    "    for animal_2 in distance_dict:\n",
    "        square_dict[animal_1][animal_2] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "square_dict.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d[9] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
